---
title: Computational Prediction of Boundaries of 3D Genomic Domains in Class Imbalance
  Settings
csl: styles.ref/genomebiology.csl
output:
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
  pdf_document: default
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT

**Background.** Chromosome conformation capture sequencing technologies have shown that the three-dimensional (3D) structure of the genome is folded into distinct compartments, known as topologically associated domains (TADs) -- units of coordinated gene expression. The location of TAD boundaries has been shown to be highly conserved. Additionally, the distinct patterns of epigenomic marks, binding peaks of different factors, and actively transcribed genes prompt the use of computational approaches to identify the epigenomic features that are most predictive of TAD boundaries.

Current methods rely on the use of classification algorithms to draw associations between the spatial relationship of functional genomic elements and regions that include TAD boundaries. However, many methods ignore key characteristics of domain data, such as the impact of heavily imbalanced classes of boundary vs. non-boundary regions and the use of distances ($log_{2}$ transformed) as the best representation of the association between TAD boundaries and genomic annotations.


**Methods.** Using replicate Hi-C data for the GM12878 and K562 cell lines, we proposed a novel ensemble framework that addresses these data irregularities and characteristics. Multiple random forest classification models built on regularized sets of genomic annotations, in combination with variations in data resolution, feature engineering, and re-sampling techniques. The composite metric Matthew's Correlation Coefficient (MCC) was used to compare models derived from the framework to determine a consensus approach to predicting TAD boundary formation.

**Results.** We discovered that models built on SMOTE re-sampled data out performed models built simply on  random under- and over-sampled data. Additionally, our results indicated that $log_{2}$ distances between annotations and TAD boundaries were more informative than other predictor-types used to build models. Furthermore, we found that performances tended to degrade as data resolution decreased. Lastly, the most predictive annotations in the best performing models included known architectural proteins such as CTCF, SMC3, RAD21, and ZNF143, as well as lesser known transcription factors including P300 and NFKB. 

**Conclusions.** The experiments derived from our ensemble framework provide a useful guide toward the exploration of chromatin organization, while highlighting the impact of both class imbalance and feature engineering in TAD boundary prediction. 

# INTRODUCTION

The advent of various genome-wide sequencing technologies, such as high-throughput conformation capture (notably Hi-C), have revealed how the spatial organization of the human genome may affect genomic regulation [@lieberman2009comprehensive]. Analyses have shown that the genome is folded into distinct compartments termed topologically associating domains (TADs) [@dixon2012topological]. Approximately 60-80% of TADs remain stable across cell types [@Schmitt:2016aa], suggesting that the boundaries of TADs may play a role in restricting the function of certain genomic elements such as enhancers, thereby impacting the transcription of genes. Furthermore, studies have shown that changing the 3D structure of the DNA, causing disruptions in these domains, can lead to adverse outcomes and diseases like cancer [@hnisz2016activation; @flavahan2016insulator; @song2017ctcf]. Therefore, it has become increasingly important to be able to identify the key molecular drivers of the formation of TAD boundaries in order to further our understanding of the human genome.

The distinct patterns of epigenomic marks, binding peaks of different factors, and actively transcribed genes prompt the use of computational approaches to identify which epigenomic features are most predictive of TAD boundaries. Evidence suggests that regulatory elements and genes tend to interact more frequently within the same TAD [@symmons2014functional]. Likewise, emerging research has shown that particular binding sites are enriched at TAD boundaries. It has been established that the zinc finger protein, CTCF, is associated with TAD boundaries and is an architectural protein responsible for the organization of the genome within the nucleus [@ghirlando2016ctcf; @ong2014ctcf]. CTCF along with other transcription factors, including SMC1, SMC3, and RAD21 have been shown to colocalize together near TAD boundaries, leading to the recruitment of a cohesin ring [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Together, these components are hypothesized to be what causes loops in the DNA, contributing to higher order chromatin organization. Other binding proteins have been shown to mediate enhancer-promoter interactions, including the transcription factors YY1 and ZNF143, thereby influencing the delineation of topological domains [@weintraub2017yy1; @gonzalez2017znf143; @heidari2014genome]. Even more so, loci within a domain have been shown to interact more frequently with one another, exhibiting similar patterns of chromatin modifications, and long-range contact patterns [rao20143d]. However, the growing number of annotated genomic features and the methods used to identify the ones assocaited with TAD boundaries still leaves much to be desired.

Current methods rely on the use of classification algorithms to draw associations between the spatial relationship of functional genomic elements and regions that include TAD boundaries. In doing so, many methods ignore key characteristics of domain data, such as the impact of heavily imbalanced number of boundaries vs. nonboundary regions. This is because, when constructing the data for modeling, a common methodology used to determine the class of the response variable is to bin the linear genome and then create a method used to determine if each bin "includes" a TAD boundary or not. Given the sparcity of TADs throughout the genome though, this will inevitably lead to severly imbalanced classes.

In a classification setting, when one is attempting to discriminate between regions with (minority class) and without (majority class) TAD boundaries, the class imbalance issue will contribute to small imbalance ratios. The imbalance ratio is defined as the ratio of minority to majority classes [@Orriols-Puig2008]. Such heavily imbalanced datasets are known to adversely impact the performance of the classifiers as the learned model is biased towards the majority class to minimize the overall error rate [@lusa2010class; @chen2013influence; @jeni2013facing]. Consequently, careful consideration must be taken when deciding which performance metric(s) to use for evaluation of predictive models, especially in the case of imbalanced data. 

Commonly used measures of model performance for binary classifiers include threshold-free metrics like the area under the receiver operating curve (AUROC), and other threshold-specific metrics like accuracy, sensitivity, and specificity. The ROC curve is created by plotting the true positive rate (TPR) against the false positive rate (FPR) at various threshold settings. The subsequent AUROC, measures how well the model is capable of distinguishing between the two classes [@park2004receiver]. Accuracy is defined as the ratio of correctly identified cases (either from the positive or negative class) over the total number of cases. It measures the overall proportion of correctly classified instances. Sensitivity, also referred to as recall or TPR, is the ratio of true positives over the total number of positive classes. Specificity, on the other hand, also referred to as TNR, is the ratio of true negatives over the total number of negative classes. Each measures the proportion of correctly identified positive and negative classes respectively [@yerushalmy1947statistical]. Problems arise when using such measures in the event of imbalanced classes. For AUROC, when the prevalence of an event is low--that is, a defined genomic region truly containing a TAD boundary--mean probabilities are biased towards the majority class. This results in the area of interest under an ROC curve to be compressed toward a small corner in the ROC space [@prati2011survey]. While not directly affecting the value of the AUROC, this can have severe consequences on the interpretability of the model's performance. A better threshold-free measure to use when dealing with imbalanced data is the area under the precision-recall curve (AUPRC).  The interpretation of the AUPRC of a model is similar to that of the AUROC. However, where as the baseline model performance according to the AUROC is given by 0.50, the baseline model performance according to the AUPRC is given by the ratio of true positives over the total number of cases. This enables a more realistic benchmark to evaluate a model on as this ratio will be lower than 0.5 for a test data set with imbalanced classes. Likewise, threshold-specific metrics like accuracy and specificity are also problematic when evaluating model performance on imbalance data due to being highly influenced by the correct classification of the majority class. A better measure to consider is a composite metric involving all quadrants of the confusion matirix, such as Matthew's Correlation Coefficient (MCC) [@jeni2013facing; @chicco2017ten]. The MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications [@brodersen2010balanced]. A value of positive 1 represents a perfect prediction, while a value of negative 1 represents total disagreement between prediction and observation. A value of 0 indicates no better than random prediction. 

<!--
precision and recall together because they soften the influence of a model's ability to correctly classify majority classes and yield more dependable and interpretable results. Therefore, measures like the F1-Score and Matthew's Correlation Coefficient (MCC) offer better interpretation, especially in the instance of imbalanced classes [@jeni2013facing]. The F1-Score can be interpreted as a weighted average of the precision and recall values, while the MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications [@brodersen2010balanced]. Each metric uses all 4 quadrants of the confusion matrix, thus providing more balanced measures in the case of imbalanced data [@chicco2017ten].
-->

<!--
, and balanced accuracy offer better interpretation, especially in the instance of imbalanced classes [@jeni2013facing]. The F1-Score can be interpreted as a weighted average of the precision and recall values, balanced accuracy measures the average accuracy obtained on either class, and MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications [@brodersen2010balanced]. Each metric uses all 4 quadrants of the confusion matrix, thus providing more balanced measures in the case of imbalanced data [@chicco2017ten].
-->

<!--
Certain composite threshold-specific metrics such as balanced accuracy and Youden's Index should be preferred because they soften the influence of a model's ability to correctly classify majority classes and yield more dependable and interpretable results [@jeni2013facing]. Here, balanced accuracy measures the average accuracy obtained on either class [@brodersen2010balanced]. Informedness, also referred to as Youden's Index, measures the probability of the model making an informed decision and corresponds to giving equal weight per class, rather than to each instance like Accuracy [@youden1950index;@powers2011evaluation].
-->

<!--
Better measures to consider are composite metrics involving both precision and recall together. Therefore, measures like the F1-Score, Youden's Index (YI), and Matthew's Correlation Coefficient (MCC) offer better interpretation, especially in the instance of imbalanced classes [@jeni2013facing]. The F1-Score can be interpreted as a weighted average of the precision and recall values, while Youden's index is interpreted as the arithmetic mean between sensitivity (recall) and specificity. MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications. Each metric uses all 4 quadrants of the confusion matrix, thus providing more balanced measures in the case of imbalanced data [@chicco2017ten].
-->

In addition to employing suitable metrics for evaluating models validated on imbalanced data, there also exist several algorithmic or data level solutions. Algorithmic level solutions involve modifyng existing classifiers in order to curtail the bias induced by imbalanced classes [@estabrooks2004multiple; @chawla2003smoteboost; @chen2004using; @khoshgoftaar2007empirical]. Data level solutions, on the other hand, mostly constitute as resampling techniques, in which random sampling is used to either replicate or remove data points in order to create a more balanced dataset [@chawla2004special; @chawla2009data]. It has been shown that quite simple resampling techniques can drastically improve model performance in the face of imbalanced data [@japkowicz2002class; @dubey2014analysis]. For the purposes of this study, we focus our attention toward data level solutions so that all results may be classifier-independent, and, thus, more easily comparable. 

Another underexplored aspect of modeling the association between TAD boundaries and genomic annotations is the various types of relationships between them. Typical approaches include overlap counts or percents [@mourad2016computational; @hong2017computational; @huang2015predicting]. Counts refer to the number of times a genomic annotation directly overlaps with a TAD boundary region, e.g., a TAD boundary may overlap with several ChIP-seq peaks for a given transcription factor. Percents refers to the percent of overlap of a genomic annotation with a TAD boundary region, e. g., a 60% of a TAD boundary may overlap genomic region annotated as heterochromatin. 

While direct overlaps represent the most straightforward measure of functional associations, they do not account for indirect relationships shown to be one of the major driving force of genome regulation [??? REFS]. The most prominent example being the enhancer-promoter relationships, where enhancers are frequently located hundreds or thousands of bases away from their target promoters [@whalen2016enhancer]. The use of distances between TAD boundaries and genomic annotations has largely been ignored. However, these distance features have the potential to offer a more accurate spatial representation of which genomic features are associated with TAD boundaries.

Additionally, model performance is heavily contingent upon the resolution of the contact matrix used to identify the location of TADs throughout the genome. It is often unclear from previous studies what resolution contact matrices were used when retrieving TAD boundaries, and how other resolutions would affect their results. A lower resolution contact matrix will result in the identification of larger-sized TADs, and subsequently, less TAD boundary points. This will contribute to smaller sample sizes and could effect predictive performances. 

Therefore, in this study, we aimed to developed an ensemble framework used to build random forest classifiers composed of different combinations of data driven parameters used to address the issues presented above. We used our framework to systematically compare model performances using a composite performance metric, MCC. Our goal was to determine a suitable combination of data re-sampling and feature engineering techniques that both corrected the inherent imbalanced classes and provided the best spatial representation linking genomic annotations to TAD boundaries. Additionally, we used our framework to directly compare the influence of Hi-C contact matrix resolution on model performances. We evaluated Hi-C data for both the GM12878 and K562 cell line obtained at 10 kb, 25 kb, 50 kb, and 100 kb resolution. For each resolution, we examined the inclusion of three different predictor types, including distance as the measure of association between genomic elements. We then tested four different re-sampling techniques aimed at creating balanced classes. Noninformative features were removed using a feature reduction algorithm and random forest classifications were performed to assess model performance, and investigate which genomic annotations were most associated with the TAD boundaries for a given system from the complete ensemble.



# METHODS

## DOMAIN DATA

Replicate Hi-C data for the GM12878 and K562 cell lines were downloaded from GEO GSE63525 [@rao20143d] (Supplementary Table 1). The autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries were obtained from data at 10 kb, 25 kb, 50 kb, and 100 kb resolution using the Arrowhead algorithm [@durand2016juicebox]. For each of the 4 data resolutions, a total of 8008, 4869, 2449, and 1075 unique TADs were extracted from the GM12878 cell line, and a total of 5828, 4097, 2181, and 986 unique TADs were extracted from the K562 cell line. The TADs were distributed throughput the autosomal chromosomes, with larger chromosomes containing more TADs (Supplementary Figure 14; Supplementary Table 8). The start and end coordinates of all TADs were concatenated, sorted, and unique TAD boundary points were obtained. 

## GENOMIC ANNOTATION DATA

Genomic annotation data in the form of cell type specific chromatin states, histone marks, and transcription factor binding sites (GRCh37/hg19 human genome assembly) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 2).

## MODEL CONSTRUCTION

To establish the data for modeling, the genome was binned into equally sized intervals defined by the data resolution for the corresponding contact matrix (Figure 1). As an example, for domain data obtained from a 10 kb resolution contact matrix, the genome was binned into 10 kb intervals. For the boundary data, in order to accommodate for the uncertainty in exact boundary location, the boundary points were flanked on both sides by the size of the resolution of the corresponding contact matrix. That is, for 10 kb resolution data, each uniquely identified TAD boundary was flanked on either side by 10 kilobases for a total width around each boundary point of 20 kb. A genomic bin was then labeled as having a TAD boundary within it (Y=1) if it overlapped with a particular flanked boundary. Otherwise, the bin was labeled as not containing a boundary (Y=0). The process was repeated for each subsequent resolution.

### FEATURE ENGINEERING 

Using the sets of genomic annotations, we created three types of predictors that were used as features for subsequent models to analyze their respective spatial relationship with TAD boundaries (Figure 3):

1. **Overlap Counts (OC)**: For each genomic bin, the total number of instances where a genomic bin overlapped with regions assigned to genomic annotations (features) was calculated. In the case of no overlaps, the value was set to 0.
   
2. **Overlap Percent (OP)**: For each genomic bin, the percent of overlap between the width of the region assigned to a genomic annotation and the total width of the bin was calculated. We defined the feature width as the number of bases that overlapped with a particular genomic bin. The percent was then calculated by dividing the feature width by the bin width. If multiple overlaps existed, the feature width was defined as the sum of the total number of bases that overlapped with a particular bin. Note that the bin width remained constant for a given resolution (either 10 kb, 25 kb, 50 kb, or 100 kb). Values for bins with no overlaps were set to be 0.

3. **Log2 Distance**: For each genomic bin, the $log_{2}$ of the distance in bases (refered to hereafter simply as $\textit{distance}$), from the center of the bin to the center of the nearest region defined by a specific genomic annotation was calculated. The log transformation was used as a normalization technique in order to account for the skewness present in the distance densities (Supplementary Figures 1).

### DATA RE-SAMPLING

Class imbalance was characterized by each dataset's imalance ratio, defined as the ratio of minority to majority classes given by the genomic bins. Minority classes were the set of genomic bins that overlapped with flanked TAD boundaries, while majority classes were the set of genomic bins that did not overlap with a flanked TAD boundary.

In order to assess the influence of imbalanced classes, we then evaluated four re-sampling techniques (in addition to no re-sampling), composed of the genomic bin class and data from each predictor type. The re-sampling techniques considered are described in detail below.

1. **No Sampling**: All of the data points from the majority and minority class of the training set were used.

2. **Random Under-Sampling (RUS)**: All of the minority classes from the training set were used. Sampling without replacement was used to obtain the same number of the majority classes. 

3. **Random Over-Sampling (ROS)**: All of the majority classes from the training set were used. Sampling with replacement was used to obtain the same number of the minority classes. 

4. **SMOTE** (Synthetic Minority Over-Sampling Technique): This method incorporates both random under- and over-sampling. Under-sampling is performed without replacement from the majority class, while over-sampling is performed by creating new synthetic observations using the minority class [@chawla2002smote]. The `DMwR` R package was used to perform SMOTE (version 0.4.1). 

<!--
5. **ROSE** (Random Over-Sampling Examples): Similar to SMOTE, ROSE combines techniques of over-sampling and under-sampling. However, contrary to SMOTE, all of the data generated by ROSE is synthetic. The data is created by sampling from one of the two classes, then generating a new example in its neighborhood using a smoothed bootstrap approach. The shape of the neighborhood is determined by the kernel density estimate of the two classes [@lunardon2013r]. Another important distinction between ROSE and SMOTE is that, for ROSE, the creation of perfectly balanced classes is unlikely due to the randomness in sampling between the two classes. The `ROSE` R package was used to perform ROSE (version 0.0.3).
-->

### FEATURE REDUCTION

A regularization technique using a combination of the $l_{1}$ and $l_{2}$ penalties, known as the elastic-net, was used to reduce the feature space. The elastic-net was chosen in order to suitably reduce the feature space while also maintaining the inclusion of multiple correlated predictors. For each elastic-net regularization, 10-fold cross-validation was used to tune the parameters that controlled the penalty terms. The parameters that maximized the mean AUPRC across each cross-fold was chosen. All elastic-net algorithms were implemented in R using the `caret` package version 6.0 [@kuhn2012caret]. 

<!--
Recursive feature elimination (RFE) was then used to determine the optimal number of features for models with the best performing predictor type. For each set of predictor type, random forest models were performed recursively on subsets of the feature space. The specific subsets of features were chosen as powers of 2 for computational efficiency. Model performance for each subset was evaluated using AUPRC. The optimal set of featues to include in downstream analyses was determined to be the subset in which the AUPRC leveled off. Again, to confirm that results were consistent, the reduction process was repeated for each combination of re-sampling technique, resolution, and cell line. 
-->

### FINAL MODEL ASSESSMENT AND PERFORMANCE

Once the appropriate predictor type and set of features were determined, a final set of random forest classification models were performed to compare model performances between re-sampling techniques and predictor types. <!--To reduce bias due to random dataset generation, 10-fold cross validation was used to tune the number of features to consider at each node and the number of ensemble trees to aggregate-->. The default number of features to consider at each node of the random forest algorithm was set as the square root of the number of features in the model. Likewise, the number of ensemble trees to aggregate was set at 500. All random forest algorithms were implemented in R using the `caret` package version 6.0 [@kuhn2012caret]. Once the model was implemented, it was validated on the test set. 

The performance of each ensemble system in the framework was compared using the composite metric, Matthew's Correlation Coefficient. The MCC is defined as follows:

<!--
$$F1 - Score = \frac{2TP}{2TP + FP + FN}$$
-->

$$MCC = \frac{TP \times TN - FP \times FN}{\sqrt{ (TP + FP)(TP+FN)(TN+FP)(TN+FN) }}$$
<!--
$$Balanced \quad Accuracy = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$
-->

<!--
$$Informedness = \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} - 1$$
-->

Here, TP refers to the number of bins correctly identified as containing a TAD boundary (true positives), FP refers to the number of bins incorrectly identified as containing a TAD boundary (false positives), TN refers to the number of bins correctly identified as not containing a TAD boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a TAD boundary. Each of these quantities are obtained from the confusion matrix created by validating the model on the testing set. Additional performance metrics were reported in the supplementary materials.

<!--Additional threshold-free metrics where also reported including the AUROC and AUPRC.--> 

<!--
Lastly, for the task of identifying the genomic annotations that were most predictive of the formation of TAD boundaries, we ranked the variable importances associated with each feature of the ensemble system-specific random forest model. To calculate variable importances for each model, the F1-Score on the out-of-bag portion of the data within each tree was recorded. Then the same was done after permuting each predictor variable. The difference between the two AUPRC values was then averaged over all trees, and normalized by the standard error of the differences. Annotations with low rankings correspond to functional epigenomic features that are either positively or negatively enriched near TAD boundaries, and thus, better allow the model to discriminate between genomic bins that overlap with flanked TAD boundaries.
-->

## ENSEMBLE FRAMEWORK

An ensemble framework was established for the purposes of systematically creating a variety of models that each included different sets of parameters described above (Figure 2). Each system in the framework is assumed to be composed of a model constructed from a particular data resolution, with a particular predictor type for the feature space. The data sets used for modeling were composed of the full set of genomic bins that either overlapped with a flanked TAD boundary or did not. These two classes made up the minority and majority sets respectively. The data was then split into a 7:3 ratio of training to testing sets. Each training and testing set was composed of a similar majority to minority class ratio as the full data set. For the task of performance evaluation and prediction, the test set remained constant for all ensemble systems to account for an unbiased comparison. Furthermore, this process was repeated, and performances were aggregated across each autosomal chromosome. Model results for Chromosome 9 was omitted from the analysis, for both cell lines, due to the sparcity of the contact matrix at the 10 kb resolution for the K562 cell line.

# RESULTS

## ENSEMBLE COMPARISONS

We compared the performances of random forest classification models implemented on TAD boundaries called using the arrowhead algorithm [@durand2016juicebox] from Hi-C data for 2 cell lines $L = \{\textit{GM12878}, \textit{K562}\}$, at four different resolutions $R = \{\textit{10 kb}, \textit{25 kb}, \textit{50 kb}, \textit{100 kb\}}$. [??? Bad spacing, for the whole paragraph. Why you are not using plain and simple Markdown?] For each classifier, we considered 3 separate predictor types as the feature space including $P = \{\textit{OC}, \textit{OP}, \textit{Distance}\}$. To assess the severity of class imbalance 4 re-sampling techniques were implemented prior to modelling, including $B = \{\textit{None}, \textit{RUS}, \textit{ROS}, \textit{SMOTE}\}$. Each combination of the above parameters was repeated 21 times (number of autosomal chromosomes, excluding chromosome 9), $C = \{ 1, 2, ..., 22 \}$. A system produced from the ensemble framework was defined as: $S = \{\textit{c}, \textit{r}, \textit{p}, \textit{b}\}$, where $l \in L$, $r \in R$, $p \in P$, $b \in B$, and $c \in C$. Letting $|X|$ denote the cardinality of each set, we evaluated $|L| \times |R| \times |P| \times |B|$ different ensemble systems, totaling $2016 (= 2 \times 4 \times 3 \times 4 \times 21)$ separate models. 

## IMBALANCE RATIOS REMAINED SEVERE AND CONSISTENT ACROSS RESOLUTIONS

Results showed that the imbalance ratios were well below acceptable thresholds for both cell lines, and across the 4 resolutions (Supplementary Table 9 ???SDs will come with more chromosomes). The low imbalance ratios indicated a wide disparity between the number of genomic bins that overlap with a flanked TAD boundary and the much larger number that did not. While resolution certainly played a role in determining the number of observations included in datasets used in downstream modeling, as it determines the width of genomic bins and larger widths (lower resolutions) created less bins, we found that there was very little variation in imbalance ratios across resolution (Supplementary Figure 16 ???variation will come with additional chrs). 

<!--
## HIGH CORRRELATIONS IMPLY COLOCALIZATION OF MANY ANNOTATIONS AROUND TAD BOUNDARIES

By isolating flanked TAD boundaries and, in turn, counting the instances in which annotations overlapped with boundary regions, we were able to assess the correlations of genomic annotations and determine which annotations were enriched together at boundaries. Clustered correlograms for known colocalized annotations showed that distinct groups of transcription factors were highly positively correlated in both cell lines, and across resolutions, including CTCF, ZNF143, RAD21, and SMC3ab (Supplementary Figures 2-9). Likewise, combinations of particular histone modifications and chromatin states were also shown to exhibit similar profiles near TAD boundaries including H3K4me1, H3K4me2, and H3K4me3, as well as strong and weak enhancers. Furthermore, particular chromatin segmentation states, inclduing heterochromatin, repressed chromatin, poised promotors, and repetitive copy number variations, were found to be highly positively inter-correlated, yet were negatively correlated with many transcription factor binding sites. Notably, correlations among annotations were found to increase as resolutions decreased for both cell lines. This indicated that TAD bounardies called, and subsequently flanked, at lower resolutions tended to encapsulate more of these epigenomic marks and contribute to higher levels of enrichment. 

The high correlations among many of the genomic annotations gave precedent for the use of an elastic-net regularization algorithm in order to suitably reduce the feature space while retaining possibly correlated annotations that could be predictive of TAD boundary formation. For each model created from the ensemble framework, the alpha and lambda were tuned over a grid of values using 10-fold cross validation (Supplementary Table 7).
-->

## CLASS IMBALANCE SEVERELY IMPACTS MODEL PERFORMANCE

In order to set a baseline in performance, we first evaluated the least granular model, that is one implemented using OC predictor types, without data re-sampling. We found that models built data with fully imbalanced classes exhibited poor MCC across all resolutions (Figure 5A). We then gradually increased the granularity by assessing models built on OP predictor types, while still not performing any data resampling. Interestingly, we found that MCCs were worse than when using OC predictor types for the GM12878 cell line (Figure 5B). When including distance-type predictors, models were found to exhibit slightly better MCCs than the other two types (Figure 5C). 
Importantly, when validating these 3 models built on fully imbalanced datasets, the MCC could not be determined at lower resolutions given by the model's complete inability to correctly classify any genomic bins that overlapped with flanked TAD boundaries in the test data, thereby resulting in 0 true positives. Furthermore, we found that evaluating models using conventional metrics, such as AUC or accuracy, did not depict the true severity in performance (Supplementary Table 3). Results were found to be consistently poor across resolutions for the K562 cell line (Supplementary Figure 10). Similarly, many of the performance metrics at higher resolutions were also unable to be reported due to the models inability to produce any true positive cases.

## DATA RE-SAMPLING COUPLED WITH THE INCLUSION OF A DISTANCE-TYPE FEATURE SPACE SUBSTANTIALLY IMPROVED MODEL PERFORMANCE

When models were built on randomly oversampled (ROS) data, we found a substantial increase in MCC for the GM12878 cell line (Figure 5 A-C). The biggest improvement in performance was seen at the 10 kb resolution in each of the 3 predictor types, with a downward trend as resolutions decreased. Although performances were low at lower resolutions, MCC values were able to be obtained, contrary to models built on completely unbalanced data as seen before. Distance-type predictors were found to offer the highest MCCs for ROS data, across all resolutions. Similar results were obtained when applying ROS to the K562 cell line (Supplementary Figure 10). Here, a much sharper negative trend was featured for MCCs as resolution decreased, than in the GM12878 cell line.

Further improvement in performance was found when models were implemented on both randomly undersampled (RUS) and SMOTE resampled data, for all 3 predictor types (Figure 5). Although the differences in performance between the two techniques were marginal, SMOTE resampled data was found to yield the best MCCs among all classes of models developed and compared in the ensemble framework. As with ROS, distance type predictors yielded the highest MCC values across resolutions, when applied to either RUS or SMOTE. SMOTE re-sampled data, with the inclusion of distance-type predictors was also found to be the best set of characteristics when applied to the K562 cell line (Supplementary Figure 10). Moreover, we found that there was a degredation in model performance as data resolution decreased. This was much more pronounced in the K562 cell line. On the GM12878 cell line, there were only marginal differences in performance between 10 kb and 25 kb, followed by a steep decline at 50 kb. Then, the MCCs showed a momentary increase in performance at 100 kb, the lowest resolution considered in the study. Performances from the K562 cell line degraded much faster as resolution decreased, starting with highest performance at 10 kb and ending with the worst performance at 100 kb. These results imply that a best performing model, one that is able to best discriminate between genomic bins that are near versus far away from TAD boundaries, is one that is built on distance-type predictors and incorporates both random under- and over-sampling.  

<!--
Additionally, when we performed and evaluated the variable importances of random forest models built on one genomic annotation at a time (including the 3 predictor types each), we found that the distance-type consistently out-ranked the OC and OP type predictors in terms of predictive importance, for a majority of the cell line specific annotations, across resolution and cell line (Supplementary Tables 4 & 5). Therefore, it appears that distance-type predictors provided the most informative spatial representation of epigenomic annotations associated with TAD boundaries.
-->

## MODELS IMPLEMENTED ON TRANSCRIPTION FACTOR BINDING SITES, USING OPTIMAL PARAMETERS, OUTPERFORM OTHER ANNOTATION-SPECIFIC MODELS

In order to assess the predictive capabilities of specific types of annotations, we further compared the performances of annotation-specific models. Using the optimal set of parameters, SMOTE re-sampling with distance-type predictors, random forest classifiers built on sets of histone modifications, chromatin state segmentations, and TFBS were compared to a model that included all three types together. For the GM12878 cell line, models built simply on TFBS were found to perform consistently just as well as models that included all 3 sets of annotations, across resolution (Figure 6). Similar results were found when evaluating the K562 cell line (supplementary Figure 11). For both resolutions, models built on only histone modifications were found to have the MCC, across resolutions, followed by models built on chromatin state segmentations. This suggests that TAD boundary formation is primarily carried out by transcription factor binding sites.

## PREDICTIVE IMPORTANCES CONFIRMED BIOLOGICAL FUNCTIONALITY OF KNOWN TRANSCRIPTION FACTOR BINDING SITES

As an external method of validation, we also investigated the predictive importance of the annotations that our top performing models were built on. We found that specific transcription factor binding sites (TFBS), including SMC3ab, ZNF143, CTCF, and RAD21, were given consistently high predictive importance in the models built on the GM12878 cell line, across all resolutions (Figure 8). These are known factors associated with the hypothesized loop extrusion model [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self] and were found to yield consistently high predictive importance on the K562 cell line as well, with similar profiles across resolutions (Supplementary Figure 15). 

## DISTANCE PROFILES HIGHLIGHT THE RELATIONSHIP BETWEEN KEY TFBS AND TAD BOUNDARIES

When evaluating the 10 kb data resolution, on GM12878, the distance profiles (on a $log_{2}$ scale) were found to be smaller for top predictive TFBS annotations in relation to genomic bins that overlapped with flanked TAD boundaries, when compared to bins that did not overlap with a flanked TAD boundary (Figure 7). At the same resolution and cell line, direct comparisons of quantiles obtained from the distribution of distances for the transcription factors SMC3a, ZNF143, CTCF, and RAD21, led to the the conclusion that these TFBS were more often located near TAD boundaries on the linear genome, as can be seen from the shifted quantile differences (Figure 7). Similar distance profiles for the same TFBS were obtained when assessed at lower resolutions, across both cell lines (Supplementary Figures 12-17 A-D).


# DISCUSSION

We developed an ensemble framework to systematically compare model performances for different combinations of re-sampling and feature engineering techniques aimed at determining a consensus approach to predicting TAD boundary formation using genomic annotations. Additionally, this study provides direct comparisons of 4 different Hi-C data resolutions, and their effects on performances of downstream models. Furthermore, we go on to show the importance of using adequate metric when assessing model performance, especially in the case of highly imbalanced classes.


A typical problem facing machine learning approaches toward predicting TAD boundary formation is the introduction of highly imbalanced classes. This is because, when constructing the data for modeling, a common methodology used to determine the class of the response variable is to bin the linear genome and then create a method used to determine if each bin "includes" a TAD boundary or not. Given the sparcity of TADs throughout the genome though, this will inevitably lead to severly imbalanced classes. A workaround to this employed by [@huang2015predicting; @hong2017computational; @bednarz2014supervised] was to randomly sample from the majority class to create balanced classes in the training set. It is unclear how these studies handled 




When constructing the data for modeling, a common methodology used to determine the response variable is to bin the linear genome into equally sized intervals and then create a method used to determine if each bin "includes" a TAD boundary or not. Given the sparcity of TADs throughout the genome, this will inevitably lead to severly imbalanced classes. 

Although we were not able to draw direct comparisons between the GM12878 and K562 cell lines, given that they did not contain the same list of cell type specific annotations, we did find that there was relative agreement in rankings of the top predictive predictors. However, there were notable exclusions from each cell line, in addition to inconsistencies across resolutions

## LIMITATIONS

Many elastic-net models using overlap count or percent predictors were reduced to ridge, or near-ridge regression models at higher data resolutions by having the $alpha$ penalization term tuned to 0 in the cross-validation step, thereby not reducing, or reducing very few of the feature space. The results of the variable reduction step imply that far more information relating annotions to TAD boundaries is encoded in the distance-type predictors. That is, the elastic-net algorithms were better able to separate noise and reduce the feature space when performed on distance-type predictors. Likewise, this reduction was most efficient at more refined levels of data resolution.


# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
