---
title: Computational Prediction of Boundaries of 3D Genomic Domains in Class Imbalance
  Settings
csl: styles.ref/genomebiology.csl
output:
  pdf_document: default
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT (Spiro version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance problem, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUCROC) curve, Precision-Recall curve (AUPRC), F1, Matthews Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were best type of genomic annotations for predicting TAD boundaries than histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# ABSTRACT (MD version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUROC) curve, Precision-Recall curve (AUPRC), F1, Matthews Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were the best type of genomic annotations for predicting TAD boundaries when directly compared with histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# INTRODUCTION

The advent of various genome-wide sequencing technologies, such as high-throughput conformation capture (notably Hi-C), have revealed the spatial organization of the human genome <!--may affect genomic regulation--> [@lieberman2009comprehensive]. Analyses have shown that the genome is folded into distinct compartments termed topologically associating domains (TADs) [@dixon2012topological]. Given their high degree of conservation between cell types and species, it has been proposed that TADs represent the fundamental building blocks of organization, giving the genome its 3D structure [@rocha2015breaking; @dixon2012topological]. Likewise, evidence suggests that the boundaries that delineate TADs may play a role in restricting the function of certain genomic elements such as enhancers, thereby impacting the transcription of genes [@symmons2014functional]. Even more so, loci within a domain have been shown to interact more frequently with one another, exhibiting similar patterns of chromatin modifications, and long-range contact patterns [@rao20143d; @symmons2014functional] [??? This sentence and reference belong to the previous paragraph, think and move; !!!combined and moved]. Furthermore, studies have shown that changing the 3D structure of the DNA, causing disruptions in these domains, can lead to adverse outcomes and diseases like cancer [@hnisz2016activation; @Flavahan2016; @song2017ctcf]. Therefore, it has become increasingly important to be able to identify the key molecular drivers of the formation of TAD boundaries in order to further our understanding of the human genome.

Distinct patterns in epigenomic marks, binding peaks of different factors, and actively transcribed genes prompt the use of computational approaches to identify which epigenomic features are most predictive of TAD boundaries [@dixon2015chromatin; @di2017novo] [??? REFS? Carefully select from my paper notes; !!!added refs]. Recent research has identified key functional genomic annotations enriched at boundaries. Namely, the CCCTC-binding factor  (CTCF), has been shown to be associated with TAD boundaries [@ghirlando2016ctcf]. Other transcription factors including SMC1, SMC3, and RAD21 have been shown to colocalize near TAD boundaries [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Together, these components are hypothesized to be what causes loops in the DNA, contributing to higher order chromatin organization. Additional binding proteins, known to mediate enhancer-promoter interactions including YY1 and ZNF143, have been shown to influence the delineation of topological domains [@weintraub2017yy1; @gonzalez2017znf143; @heidari2014genome]. <!--However, there is a growing number of publically available annotated genomic features, and it is unclear if current methods are robust enough to identify all of them. [??? This is very weak and undefined. You started with mentioning prediction - finish with unsolved problems with it. !!!editted]-->

Current methods used to predict TAD boundaries rely on the use of classification models to draw associations between functional annotations and genomic regions that include TAD boundaries. However, many key characteristics of 3D genomic data remain unexplored, potentially limiting performance of predictive modeling. Specifically, the minority of TAD boundaries as compared with the rest of the genome creates proportionally small TAD boundary regions versus non-TAD boundary regions. These TAD boundary regions make up the minority class, while the non-TAD boundary regions make up the remaining majority class. The large disparity is sizes between minority and majority classes creates imbalanced data, as can be characterized by the *proportion of imbalance* (PI). Here, PI is defined as the ratio of the set minority class observations to the sum of both sets of minority and majority class observations. A dataset is considered imbalanced if PI<0.5. <!--Specifically, the minority of TAD boundaries as compared with the rest of the genome creates heavily imbalanced classes, characterized by ratios of positive/negative classes used to train the models. Here, the imbalance ratio is defined as the ratio of majority classes (regions not including TAD boundaries) to minority classes (regions including TAD boundaries) [@Orriols-Puig2008].--> Classifiers implemented on heavily imbalanced datasets are known to exhibit poor performances as the learned model is biased towards the majority class [@lusa2010class; @chen2013influence; @jeni2013facing]. Furthermore, the use of different levels of association between TAD boundaries and genomic features (number, or proportion of overlaps, or distance) as predictors remain underexplored. Lastly, it remains unclear how increased resolution of Hi-C data might affect the performance of predictive models. 

Recently, an L1-regularized multiple logistic regression model, called *HiCfeat*, was developed to predict TAD boundaries using transcription factor binding sites on GM12878 [@mourad2016computational]. Based on a 1 kb binned genomic data matrix and 20 kb flanked TAD boundaries, the model was trained using proportions of overlaps between ChiP-seq peak defined functional element coordinates and genomic bins. However, several problems exist with this study. First, many transcription factors are known to colocalize, and therefore exhibit high correlations. Given a group of features that are highly correlated, L1-regularization tends to pick only one of them, while shrinking the others to zero [@zou2005regularization]. Here, ensemble learning techniques could offer a more robust method toward identifying  annotations associated with TAD boudaries. Likewise, it is unclear how other sets of epigenomic features affect prediction, including chromatin state segmentations and histone modifications. Secondly, the class imbalance issue was not addressed prior to modelling. Instead, model performance was evaluated using ROC (receiver operating characteristic) curves and subsequent AUC (area under the curve). It has been shown that AUC is insensitive to class imbalance, creating artificially inflated values inlfuenced by the majority class [@prati2011survey; @wei2013role]. It is unclear how simple data level solutions to class imbalance, such random under-sampling and random over-sampling affect TAD boundary prediction. Likewise, a more composite metric such as Matthews Correlation Coefficient (MCC), might provide more sensitive assessment of model performance. Thirdly, the relationship between annotations and boundaries was quantified using proportions of overlaps. This type of predictor ignores distances, which offers a much more spatially descriptive feature space and could improve predictive performance. Lastly, it is unclear how *HiCfeat* would perform on other cell lines and Hi-C data resolutions, giving *HiCfeat* a limited application in TAD boundary prediction.

<!--
[Paragraph about disconnect between Hi-C and epigenomic data] Despite the success of Hi-C technologies in revealing the high-order levels of genome organization (chromosome territories, A/B compartments, TADs, sub-TADs, chromatin loops), resolution of Hi-C data (~750kb [??? REF, Bonev 2017?]) remains well below resolution of epigenomic assays (hundredths of bases for ChIP-seq). One of the key missing pieces to connect 3D genome architecture to transcription has been the inability of Hi-C to attain the resolution necessary to discern gene-level features of chromatin. This resolution gap between 1D and 3D genomic maps, as well as a paucity of chromatin features below the level of TADs, has significantly limited our understanding of epigenomic regulation of gene expression and its potential link to chromatin architecture. Although computational [??? REFS?](Zhang et al., “HiCPlus.”, Carron et al., “Boost-HiC.”) and technological [@Hsieh:2019aa][@Ma:1998aa][@Belaghzal2017] methods for enhancing resolution of Hi-C data started to appear, they [??? remains to write about their pitfalls]. One aim of this study is to predict the location of TAD boundaries on a scale comparable to epigenomic assays. For this, we developed a model trained on a relatively coarse Hi-C data for predicting 3D features (TAD boundaries) from 1D data (epigenomic datasets).
-->

In this study, we developed a computational framework aimed at determining the optimal set of data level characteristics necessary for optimal prediction of TAD boundaries and identification of influential genomic annotations. We utilize random forest classification models built on HiC-data from two cell lines, including GM12878 and K562, obtained at four different resolutions, including 10 kb, 25 kb, 50 kb, and 100 kb. For each cell line and resolution, we compared three feature engineering techniques, including counts, percents, and distances. Additionally, we compared three re-sampling techniques, including random under-sampling, random over-sampling, and a weighted combination of both referred to as Synthetic Minority Over-sampling Technique (SMOTE). Model performances were evaluated and compared using Matthews Correlation Coefficient (MCC).

[???Include concluding paragraph]


# METHODS

## DOMAIN DATA

Biological replicate Hi-C data for the GM12878 and K562 cell lines were downloaded from GEO GSE63525 [@rao20143d] (Supplementary Table 1). The autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries were obtained from data at 10 kb, 25 kb, 50 kb, and 100 kb resolution using the Arrowhead algorithm [@durand2016juicebox]. TAD boundaries called on chromosome 9 were omitted from all downstream analyses, for all cell lines and resolutions, due to the sparsity of the contact matrix at the 10 kb resolution for the K562 cell line.

## GENOMIC ANNOTATION DATA

Genomic annotation data in the form of cell type specific chromatin states, histone marks, and transcription factor binding sites (GRCh37/hg19 human genome assembly) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 2). Functional annotations were represented by chromosome-specific regions on the linear genome obtained by ChIP-seq peak calling algorithms.

## CLASSIFICATION

To establish the data for classification, the genome was binned into equally sized intervals defined by the data resolution of the corresponding contact matrix. As an example, for domain data obtained from a 10 kb resolution contact matrix, the genome was binned into 10 kb intervals. In order to accommodate for the uncertainty in exact boundary location, the boundary points were flanked on both sides by the size of the resolution of the corresponding contact matrix. That is, for 10 kb resolution data, each uniquely identified TAD boundary was flanked on either side by 10 kilobases for a total width around each boundary point of 20 kb. A genomic bin was then labeled as having a TAD boundary within it (Y=1) if it overlapped with a particular flanked boundary. Otherwise, the bin was labeled as not containing a boundary (Y=0).

### FEATURE ENGINEERING 

Using the sets of cell line specific genomic annotations, we created three types of predictors that were used to describe the relationship between functional annotations and genomic bins. They are described in detail below:

1. **Overlap Counts (OC)**: For each genomic bin, the total number of instances where a bin overlapped with regions assigned to genomic annotations (features) was calculated. 
   
2. **Overlap Percent (OP)**: For each genomic bin, the percent of overlap between the width of the region assigned to a genomic annotation and the total width of the bin was calculated. We defined the feature width as the number of bases that overlapped with a particular genomic bin. If multiple overlaps existed, the feature width was defined as the sum of the total number of bases that overlapped with a particular bin.

3. **Log2 Distance**: For each genomic bin, the $log_{2}$ of the distance in bases (refered to hereafter simply as $\textit{distance}$), from the center of the bin to the center of the nearest region defined by a specific genomic annotation was calculated. 

### TRAINING AND TESTING

The data set used for modeling was composed of the response vector **Y**, and the predictor-type specific data matrix, **X** = {$X_1$, $X_2$, ..., $X_p$}, where $p$ was the complete set of functional annotations. Each composed data set was specific to cell line and resolution. We then split the data into a 7:3 ratio of training to testing sets. For the task of performance evaluation and prediction, the test set remained constant to account for an unbiased comparison between data level characteristics.

### DATA RE-SAMPLING

Class imbalance was characterized by each dataset's *proportion of imbalance* (PI), defined as the ratio of minority class observations to the sum of minority and majority class observations:

$$PI = \dfrac{C_{min}}{C_{min} + C_{maj}}$$

where $C_{min}$ denotes the complementary set of genomic bins that did overlap with a flanked TAD boundary and $C_{maj}$ denotes the set of genomic bins that did not overlap with a flanked TAD boundary. In order to assess the influence of imbalanced classes, we then evaluated four re-sampling techniques (in addition to no re-sampling). They are described in detail below.

1. **No Sampling**: All of the data points from the majority and minority classes of the training set were used.

2. **Random Under-Sampling (RUS)**: All of the data points from the minority class of the training set were used. Sampling without replacement was used to obtain the same number of data points from the majority class. 

3. **Random Over-Sampling (ROS)**: All of the data points from the majority class of the training set were used. Sampling with replacement was used to obtain the same number of data points from the minority class. 

4. **SMOTE** (Synthetic Minority Over-Sampling Technique): This method incorporates both random under- and over-sampling [@chawla2002smote]. Under-sampling is performed without replacement from the majority class. Over-sampling is performed by creating new synthetic observations using the minority class. The minority class is over-sampled by taking each minority class sample and introducing synthetic examples along the feature vectors joining any/all of the k=5 (default) minority class nearest neighbors. The SMOTE algorithm can have many variations in the weightinh of over- and under-sampling. In this study we restricted the algorithm to 100% over-sampling and 200% under-sampling. That is, 1 new synthetic observation was to be created for each existing minority class data point, while 2 data points were simultaneously randomly selected from the majority class. This combination results in perfectly balanced classes. The *DMwR* R package was used to perform SMOTE (version 0.4.1). 

<!--
### FEATURE REDUCTION

Architectural proteins tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To measure the pairwise correlation of each cell line specific annotation, we used the Jaccard Index. The Jaccard Index measures the ratio of the number of intersecting base pairs between two sets to the number of base pairs in the union, subtracted by the intersection [@favorov2012exploring]. The quantity ranges from 0.0 to 1.0, where 0.0 represents no overlap (no correlation) and 1.0 represent complete overlap (complete correlation). [?? Jaccard is a distance measure, I would avoid using the term correlation here ??]

In order to suitably reduce the feature space while also maintaining the inclusion of possibly colocalized annotations, we implemented an elastic-net regularization technique prior to modelling. The technique using a combination of the *l_{1}* and *l_{2}* penalties. For each elastic-net regularization, 10-fold cross-validation was used to tune the parameters that controlled the penalty terms. All elastic-net algorithms were implemented in R using the *caret* package version 6.0 [@kuhn2012caret]. 
-->

### FINAL MODEL ASSESSMENT AND PERFORMANCE

Random forest classification models were built to compare model performances between combinations of predictor types and re-sampling techniques. To reduce bias due to random dataset generation, 10-fold cross validation was used to tune the number of annotations, *p*, to consider at each node. The number of trees that were aggregated for each random forest model was set to the default, n=500. All random forest algorithms were implemented in R using the *caret* package version 6.0 [@kuhn2012caret]. Once the model was implemented, it was validated on the test set. The performance of each model was compared using the composite metric, Matthews Correlation Coefficient (MCC). The MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications [@brodersen2010balanced]. A value of 1 represents a perfect prediction, while a value of -1 represents total disagreement between prediction and observation. A value of 0 indicates no better than random prediction. The MCC is defined as

<!--
$$F1 - Score = \frac{2TP}{2TP + FP + FN}$$
-->

$$
MCC = \frac{TP \times TN - FP \times FN}{\sqrt{ (TP + FP)(TP+FN)(TN+FP)(TN+FN) }}
$$
<!--
$$Balanced \quad Accuracy = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$
-->

<!--
$$Informedness = \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} - 1$$
-->

where TP refers to the number of bins correctly identified as containing a TAD boundary (true positives), FP refers to the number of bins incorrectly identified as containing a TAD boundary (false positives), TN refers to the number of bins correctly identified as not containing a TAD boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a TAD boundary (false negatives). Each of these quantities are obtained from the confusion matrix created by validating the model on the testing set. Additional performance metrics were reported in the supplementary materials.

# RESULTS

## DEVELOPING A RESOLUTION-SPECIFIC APPROACH FOR DATA CONSTRUCTION

We analyzed publically available replicate Hi-C data sets from two cell lines including GM12878 and K562 (Supplementary Table 1). Topologically associating domains (TADs) were subsequently called using the ARROWHEAD pipeline [@durand2016juicer] at four resolutions: 10 kb, 25 kb, 50 kb, and 100 kb. To account for the varying levels of data resolution, we developed a resolution-specific approach toward constructing TAD versus non TAD boundary regions on the linear genome, labeled as a response vector **Y** (Figure 1). First we binned the linear genome according to the data resolution. We then flanked the boundaries on either side by the same number of basepairs as the previously constructed genomic bins. TAD boundary regions were defined as regions on the linear genome where a genomic bin overlapped with a flanked TAD boundary ($Y = 1$), while non-TAD boundary regions were defined as the regions that did not overlap with a flanked TAD boundary ($Y = 0$).  

<!--## TOTAL NUMBER OF TAD BOUNDARY REGIONS DECREASE AS Hi-C DATA RESOLUTION DECREASES-->
Expectedly, the total number of TAD boundary regions increased with the increased resolution of Hi-C data irrespectively of cell line (Figure 2A). At higher resolutions, GM12878 was found to contain more TADs, and thus more TAD boundary regions compared to K562. This can largely be attributed to the increased sequencing depth of the GM12878 library. However, this difference was found to dissipate as resolution decreased, resulting in much more similar numbers of TAD boundary regions between the two cell lines at the 50 kb and 100 kb resolutions respectively. 

![**An illustration of the proposed approach for resolution-specific data construction used for TAD boundary prediction.** Resolution-specific TADs were called and unique boundaries were identified, denoted by the blue dots. The yellow regions denote the flanked boundary regions according to the Hi-C data resolution (i.e. 100 kb resolution corresponded to a 100 kb flank for a total boundary region of 200 kb). The linear genome was subsequently binned into resolution-specific intervals. The response vector **Y** was defined as a variable that indicated whether a genomic bin overlapped with a flanked TAD boundary ($Y = 1$) or not ($Y = 0$).](Figures/figure1_model_construction_V7.png)

## TAD BOUNDARY REGIONS EXHIBIT CELL-LINE SPECIFICITY

We first sought to compare the cell-line specificity of TAD boundary regions. Between the two cell lines, a significant number of TAD boundary regions were specific to one of the two cell lines (Figure 2B). It was found that 45%, 38%, 37%, and 36% of TAD boundary regions were specific to the GM12878 cell line, across each resolution respectively. Similarly, for K562, it was found that 27%, 27%, 29%, and 33% of TAD boundary regions were cell line specific for each resolution specifically. 

## PROPORTIONAL IMBALANCE REMAINED SEVERE IRRESPECTIVELY OF Hi-C DATA RESOLUTION 

We then investigated the proportion of class imbalance, defined as the ratio of TAD boundary regions versus the sum of both TAD and non-TAD boundary regions. We compared the estimated the number of TAD boundary regions ($Y=1$ in the response vector) and the rest of genomic bins ($Y=0$) for each cell line, at each of the four resolutions of Hi-C data. The proportional imbalance was found to be invariant to resolution, remaining consistently low for both cell lines, with very little variability between chromosomes (Figure 3C). Such low proportions of imbalance gave precedence for the implementation of re-sampling prior to modeling. 

![**Cell-line specific boundary summaries.** (A) Barplots illustrating that the total number of TAD boundary regions decreases as Hi-C data resolution decreases. (B) Venn diagrams showing that TAD boundary regions exhibit cell-line specificity across resolution. (C) Boxplots indicating that the proportion of class imbalance remained severe irrespective of resolution.](Figures/bounds_and_pis.png)

<!--
**Table 1. Sparse TAD boundaries throughout the linear genome create high imbalance ratios irrespective of resolution.** 

| Resolution (Bin Width) | Average Number of Genomic Bins (SD) | Average Number of TAD Boundaries (SD) | Average Imbalance Ratio (SD) |
|------------------------|-------------------------------------|---------------------------------------|------------------------------|
| 10 kb                  | 12214.1 (6068.5)                    | 1254.6 (598.8)                        | 8.7 (1.2)                    |  
| 25 kb                  | 4892.7 (2431.2)                     | 714.0 (358.4)                         | 5.9 (0.5)                    |
| 50 kb	                 | 2450.9 (1217.6)                     | 364.4 (184.5)                         | 5.8 (0.5)                    |
| 100 kb                 | 1227.5 (609.8)                      | 165.3 (90.9)                          | 6.8 (1.1)                    |
-->

<!--**Table 2. Inter-Resolution ICCs.** For GM12878, two-way intraclass correlation coefficients between resolutions for the following quantities: Number of genomic bins, number of TADs called, number of unique TAD boundaries, and imbalance ratio. The ICC was computed across 21 of the autosomal chromosomes. P-values indicate the level of significance of the correlation between data resolution and each quantity.[??? What is ICC? !!!editted]

| Quantity                 | Inter-Resolution | 95% CI          | P-Value  |
|--------------------------|------------------|-----------------|----------|
| Number of Genomic Bins   | 0.464            | [0.244, 0.690]  | < 0.0001 |
| Number of TADs Called    | 0.627            | [0.426, 0.802]  | < 0.0001 |
| Number of TAD Boundaries | 0.611            | [0.407, 0.791]  | < 0.0001 |
| Imbalanced Ratio         | 0.114            | [-0.058, 0.372] | 0.109    |
-->

## DEVELOPING THREE TYPES OF PREDICTORS DESCRIBING THE ASSOCIATIONS BETWEEN TAD BOUNDARY REGIONS AND EPIGENOMIC FEATURES 

To establish the feature space used for classification, we constructed three different predictor types. Counts of overlaps (**overlap count**, OC) is the most commonly used type of association between genomic elements (Figure 3A). Percents of overlaps (**overlap percent**, OP) represents a more precise measure of association as it considers the proportion of a bin overlapped by a genomic annotation (Figure 3B). **Distance** between the centers of bins and genomic annotations quantifying the spatial relationship between genomic elements is currently unexplored (Figure 3C). Log2-transformation of distance was used as a normalization technique in order to account for the skewness present in the distributions of distances (Supplementary Figures 1-2). These types of genomic associations provide a granular way to investigate classification performance in class imbalance settings.

![**Three types of associations between TAD boundaries and genomic annotations.** (A) _Overlap Counts_ (OC) consider the total number of genomic features that overlapped with each genomic bin. (B) _Overlap Percent_ (OP) was calculated by dividing the sum of all feature widths within each specific bin by the total bin width. (C) _Distance_ ($log_{2}$ transformed) was calculated by measuring the distance (in bases) from the center of each genomic bin to the center of the nearest region defining a genomic annotation. ](Figures/figure2_predictor_types.png)

## ENSEMBLE FRAMEWORK & COMPARISONS

In order to identify the optimal set of parameters necessary for TAD boundary prediction, we created an ensemble framework to systematically compare performances of random forest classifiers built on different combinations of predictor types and re-sampling techniques (Figure 4). Each system in the framework was assumed to be composed of a random forest model constructed from a particular cell line ($L = \{GM12878, K562\}$). Within each cell line we evaluated 4 data resolutions ($R = \{10 kb, 25 kb, 50 kb, 100 kb\}$). We then considered 3 separate predictor types as the feature space ($P = \{OC, OP, Distance\}$. To assess the severity of class imbalance, 4 re-sampling techniques were implemented prior to modelling, including $B = \{None, RUS, ROS, SMOTE\}$). Each combination of the above parameters was implemented on chromosome specific sections of the linear genome. Performances were then aggregated across each of the 21 chromosomes (omitting chromosome 9). Letting $|X|$ denote the cardinality of each set, we evaluated $|L| \times |R| \times |P| \times |B|$ different ensemble systems, totaling 2016 ($2 \times 4 \times 3 \times 4 \times 21$) separate models. <!--We report our main results for the GM12878 cell line and have included results from the K562 cell line in the supplementary materials.-->

![**Ensemble Framework.** A diagram of the ensemble framework detailing the combination of inputs used in the model building pipeline. TAD boundary data was extracted from a specific Hi-C resolution and a variety of functional genomic annotations were used to establish the feature space. Each subsequent cell line specific random forest classification model was built on an isolated predictor type in conjunction with a particular data re-sampling technique to account for class imbalance. The model was then validated on a testing set and performances were evaluated using MCC.](Figures/figure3_ensemble_framework_V2.png)

## CLASS IMBALANCE SEVERELY IMPACTS MODEL PERFORMANCE

In order to set a baseline in performance, we first evaluated the least granular random forest classifier, that is one implemented using OC predictor types, without data re-sampling. We found that models built on data with fully imbalanced classes exhibited poor MCCs across all resolutions (Figure 4A). We gradually increased the complexity in data level characteristics by assessing models built on OP predictor types, while still not performing any data resampling. Here, the results were similar to that of OC types (Figure 4B). However, when including distance-type predictors, models were found to exhibit a marked improvement in MCC compared to the other two types (Figure 4C). Results were found to be consistently poor across resolutions and predictor types, when implemented on the K562 cell line (Supplementary Figure 3). Furthermore, we found that conventional metrics of performance (i.e. accuracy and AUC) were less sensitive to models built on imbalanced data, regardless of cell line, resolution, or predictor-type (Supplementary Tables 6-7). 

## SMOTE RE-SAMPLING COUPLED WITH THE INCLUSION OF A DISTANCE-TYPE FEATURE SPACE SUBSTANTIALLY IMPROVED MODEL PERFORMANCE

When assessing performances for random forest classifiers built on balanced data, we found that SMOTE re-sampling provided the optimal MCC consistently outperforming models built on ROS or RUS, across both predictor type and resolution (Figure 4). Specifically, the most improvement in model performance was seen when combining SMOTE re-sampling with distance-type predictors. This was also the optimal combination for the K562 cell line (Supplementary Figure 3). Interestingly, we found that there was a degredation in model performance as data resolution decreased. Chromosome specific models built on 10 kb data resolutions presented with the highest MCC values. Additionally, performances tended to be much more variable between chromosomes at lower resolutions. Moreover, this decreasing trend in performance was much more evident under the optimal combination of data level characteristics. These results imply that an optimally performing random forest classifier, one that is able to best discriminate between genomic bins that are near versus far away from TAD boundaries, is one that is built on 10 kb resolution data, with distance-type predictors, and incorporates both random under- and over-sampling.

![**Figure 4. SMOTE re-sampled models built on distance-type predictors yielded optimal MCCs across resolutions for GM12878.** Performances of random forest classification models built on each of the 3 predictor types: (A) OC, (B) OP, and (C) Distance. Within each predictor type, and across resolution, training data was re-sampled using different techniques including: None (red), ROS (green), RUS (blue), and SMOTE (purple). Performances were aggregated using means across each chromosome, with error bars representing 1 standard deviation away from the mean.](Figures/figure4_gm12878_model_performances.png)

## MODELS IMPLEMENTED ON TRANSCRIPTION FACTOR BINDING SITES, USING OPTIMAL PARAMETERS, OUTPERFORM OTHER ANNOTATION-SPECIFIC MODELS

After determining the optimal set of data level characteristics, we then aimed to compare the performances of annotation-specifc models. In doing so, we inspected the proportion of predictive importance that was explained by each of the 3 groups of annotations (chromatin states, histone modifications, and transcription factor binding sites), across resolution. We found that for random forest models implemented on distance-type predictors, using SMOTE re-sampling, the majority of the proportional predictive importance was obtained by the cell-line specific transcription factor binding sites, and was consistent across resolutions (Figure 5; Supplementary Figure 4). 

![**Figure 5. The proportional importance of optimally built predictive models was outweighed by transcription factor binding sites across resolutions for GM12878.** Stacked barplots illustrate the proportion of total predictive importance in models using optimal data level characteristics determined by the ensemble framework (SMOTE re-sampling; with distance-type predictors), for each resolution on GM12878. Proportions are grouped by the class of cell line specific annotations in the feature space including chromatin state segmentations (CSS), histone modifications (HM), and transcription factor binding sites (TFBS).](Figures/figure5_proportion_of_importances_gm12878.png)

We further assessed the predictive capabilities of the specific sets of annotations by comparing the performances of annotation-specific random forest models implemented on the optimal set of characteristics. We found that models built simply on cell line specific TFBS out performed other cell line and annotation-specific models at all resolutions (Figure 6; Supplementary Figure 5). Additionally, when compared to the optimal model identified from the ensemble framework, one that included all 3 sets of annotations, the TFBS-specific models performed similarly at the 10 kb and 25 kb resolutions. Interestingly though, at the 50 kb and 100 kb resolutions, the TFBS-specific models out performed models built on the combined sets of annotations, indicating that, at lower resolution, the inclusion of other types of annotations contributed to additional noise rather than signal. 

![**Figure 6. Optimally built TFBS-specific models outperformed other annotation-specific models across resolution for GM12878** Performances were compared for random forest models built using specific groups of annotations as the feature space including: histone modifications (HM; red), chromatin state segmentations (CSS; green), transcription factor binding sites (TFBS; blue), and a feature space containing all 3 types combined (combined; purple). Each bar represents a model ran through the ensemble framework, with re-sampling technique and predictor-type characteristics fixed to SMOTE and distance-types. The error bars represent $\pm$ 1 standard deviation from the mean MCC across each chromosome.](Figures/figure6_comparing_sets_of_annotations_gm12878.png)

## PREDICTIVE IMPORTANCES CONFIRMED BIOLOGICAL FUNCTIONALITY OF KNOWN TRANSCRIPTION FACTOR BINDING SITES

We then sought to determine which of the transcription factor binding sites were most influential in predicting TAD boundaries. We found that the cell line specific binding sites associated with SMC3, CTCF, RAD21, ZNF143, and YY1sc281 were given consistently high predictive importance, across all resolutions (Figure 7; Supplementary Figure 6). The SMC3, CTCF, and RAD21 binding sites are known factors associated with the cohesin protein complex, which is hypothesized to be what creates loops in the DNA  [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Likewise, ZNF143 and YY1 are known structural regulators of enhancer-promoter interactions that also contribute to looping of the linear genome [@weintraub2017yy1; @gonzalez2017znf143]. With the identification of these factors known to be enriched at TAD boundaries, we are able to confirm the validity of our model building process. Additionally, we were also able to identify some lesser known cell line specific transcription factors as predictive of TAD boundaries including the MYC-associated zing-finger protein (MAZ), jun-D proto-oncogene (JUND), and ELF1. These transcription factor binding sites tended to be given higher predictive importance at lower resolutions. In fact, at lower resolutions, the predictive importance was found to be much more evenly dispersed across the top transcription factor binding sites, than for higher resolutions. This is again suggestive of the fact that the spatial relationship between functional annotations, specifically transcription factor binding sites, and TAD boundaries changes at different Hi-C data resolutions.

![**Figure 7. Top 20 most predictive transcription factor binding sites sorted by 10 kb for GM12878.** A heatmap illustrating the importance of the key cell line specific transcription factors for optimally built random forest models across resolution. Importances were sorted according to the 10 kb resolution and rescaled to values between 0 and 100. Transcription factors with higher values of predictive importance are denoted in dark red.](Figures/figure7_variable_importance_heatmaps_tfbs_gm12878.png)

## INFLUENTIAL TFBS TENDED TO BE CLOSER TO BINS THAT OVERLAPPED WITH FLANKED TAD BOUNDARIES THAN THOSE THAT DID NOT

We then investigated the spatial patterns of the most predictive TFBS around TAD boudaries by characterizing the distance profiles between the classes of genomic bins. For GM12878, at 10 kb resolution, we found that SMC3, CTCF, RAD21, ZNF143, and YY1sc281 were all statistically significantly closer to bins that overlapped with flanked TAD boudaries than those that did not (Figure 8). Although still statistically significantly different, the distance profiles between genomic class were found to be much more similar at lower resolutions (Supplementary Figure 9). For the K562 cell line, we reported the distance profiles of the same transcription factor binding sites, excluding the YY1sc281, due to its unavailability in the UCSC Genome Browser. These cell line specific distance profiles were again found to be significantly different between genomic class, at all resolutions (Supplementary Figure 10). 

![**Figure 8. Top predictive transcription factor binding sites were significantly closer to bins that overlapped with 10 kb flanked TAD boundaries on GM12878** Boxplots illustrating the distribution of distances (on $log_{2}$ scale) around 10 kb regions on the linear genome that either overlapped with a flanked TAD boundary (Yes) or did not (No), for the most influential cell line specific TFBS including SMC3, CTCF, RAD21, ZNF143, and YY1sc281. P-values obtained from Wilcoxon Rank Sum tests are provided assessing the significance [$NS$ = p > 0.05; $*$ = 0.01 < p $\le$ 0.05; $**$ = 0.001 < p $\le$ 0.01; $***$ = 0.0001 < p $\le$ 0.001; $****$ = p $\le$ 0.00001].](Figures/figure8_distance_boxplots_key_tfbs_gm12878_10kb.png)

# DISCUSSION

In this study we developed an ensemble framework to systematically compare model performances for different combinations of re-sampling and feature engineering techniques aimed at determining a consensus approach to predicting TAD boundary formation using genomic annotations. We provide direct comparisons of cell line specific random forest models built on a variety of different data dependent characteristics, including 4 contact matrix resolutions ($\{\textit{10 kb}, \textit{25 kb}, \textit{50 kb}, \textit{100 kb\}}$), 3 predictor-types ($\{\textit{OC}, \textit{OP}, \textit{Distance}\}$), and 4 re-sampling techniques ($\{\textit{None}, \textit{RUS}, \textit{ROS}, \textit{SMOTE}\}$), for both the GM12878 and K562 cell lines. 

The experiments derived from our ensemble framework provide a useful guide toward the exploration of chromatin organization, while highlighting the impact of both class imbalance and feature engineering in TAD boundary prediction. Using the framework we demonstrated that random forest models built on imbalanced data exhibited poor performance, regardless of cell line, resolution, or predictor type. In our framework, we opted to leave the test set as a completely untouched holdout set, only performing the re-sampling, and subsequent learning, on the training set. This was done so as to incur the least amount of bias and so that all conclusions could be made soley between model characteristics. As a result, we showed that Matthews Correlation Coefficient was much more sensitive to the imbalance present in the test set, compared to more popular metrics like AUROC, which attain artificially high values by correctly predicting majority class samples [?? I may be mistaken but I don't see where this is proven ??]. When comparing re-sampling techniques, we found that SMOTE re-sampling outperformed models built on both simple ROS or RUS, implying that, in fact, a combination of both random under-sampling and random over-sampling is preferred in TAD boundary prediction. Furthermore, the most improvement [?? Over what ??] was seen when combining SMOTE resampling on distance-type features when building random forest classifiers. These two characteristics exhibited consistently better performance results across both the GM12878 and K562 cell lines. This confirmed our hypothesis that the most granular and spatially informative representation of the relationship between TAD boundaries and annotations resulted in the best performing model.

When applying SMOTE re-sampling, with distance-type predictors, on transcription factor binding sites alone, we found that random forest classifiers performed better when compared to other annotation-specific models, including histone modifications and chromatin state segmentations. Likewise the performance of TFBS specific models was comparable to models with all annotations combined, across both cell line and resolution. This suggests that the majority of the key molecular drivers aiding in the formation of TAD boundaryies are transctiption factors, and that modifications at histone tails such as acetylation or methylation play less of a role.

We also observed that, while the best set of data characteristics were consistent across resolutions, performances of classifiers built on said characteristics tended to decline as resolution decreased. For both cell lines, top performance was achieved at the 10 kb data resolution, while the worst performance was obtained at the 50 kb and 100 kb resolution for the GM12878 and K562 cell line respectively. This suggests that TAD boundary prediction is intimately related to the resolution that boundaries are called on. This result can be attributed to one of two factors in the scheme of our framework. First, at higher resolutions, by model construction genomic bins were smaller, thereby resulting in more accurate profiles for the genomic annotations [?? Don't understand this sentence ??]. Secondly, because we are mainly concerned with TAD boundaries, and not the hierarchical structure of domains as a whole, the sparsity of contact matrices at higher resolutions does not impair boundary prediction [?? Doesn't this contradict your results about low resolution causing worse boundary prediction ??]. Conversely, lower resolutions imply larger bin windows and lead to less specificity in boundary location on the linear genome, resulting in worse performances. Given the trend in performances across resolution, future research is necessary in determining the optimal resolution to use when predicting TAD boundaries [?? Might want to qualify the statement in results about 10kb data being best by saying that it only applies to your data ??]. 

Given the formulation of our framework, we were not able to make direct comparisons between cell lines. This was due to the fact that we opted to use cell line specific annotations, and there existed some annotations that were available on the K562 cell line that were not available to GM12878, and vice versa. However, we were able to confirm that final models built on the optimal set of data characteristics identified consistent and well known proteins related to the 3D structure of chromatin, for both cell lines. Our results indicated that the zinc finger protein CTCF, as well components of the cohesin complex, SMC3 and RAD21, were found to be the key molecular drivers most predictive of TAD boundary formation, consistent across cell line and resolution. In addition, models derived from our framework recognized known mediaters of long range enhancer-promoter interactions, the ZNF143 and YY1 transcription factors, as being highly predictive of TAD boundaries. 

In summary, we have shown that TAD boundary prediction is a multi-faceted problem composed of many data irregularities and characteristics that previous model have failed to address. We have shown that our proposed framework suitably addresses each of these components, while also providing the optimal considerations to make in an effort to build consensus classifiers used to predict TAD boundary formation [?? formation or location. Formation implies some temporal element ??] from functional genomic annotations. Given the stability of our results, this opens up the possibility of widening our framework to include a broader set of classifiers, cell lines, and Hi-C contact matix resolutions that will inevitably assist us in gaining a better understanding of the mechanisms associated with the architecture of the human genome.



# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
