---
title: Computational Prediction of Boundaries of 3D Genomic Domains in Class Imbalance Settings
csl: styles.ref/genomebiology.csl
output:
  pdf_document: default
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT (Spiro version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks that lead to the organization and regulation of the three-dimensional (3D) genome. The enrichment of several epigenetic marks at TAD boundaries suggests a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data including spatial associations between genomic annotations and boundaries, as well as the severe imbalance in TAD-boundary vs non-TAD boundary defined genomic regions. Furthermore, resolution of Hi-C data remains well below the resolution of epigenomic assays (hundredths of bases for ChIP-seq) affecting the ability of Hi-C sequencing in discerning gene-level features of chromatin. This resolution gap between 1D and 3D genomic maps, as well as a paucity of chromatin features below the level of TADs, has significantly limited our understanding of epigenomic regulation of gene expression and its potential link to chromatin architecture.

**Methods.** We first developed an ensemble framework for optimizing data level characteristics necessary for TAD boundary prediction including combinations of 5 Hi-C data resolutions (5 kb to 100 kb), 3 spatially defined predictor types, and 4 re-sampling techniques, across the GM12878 and K562 cell lines. We then aimed to predict TAD boundary location on a 1D genomic map by validating optimally built cell line specific models at basepair resolution. 

**Results.** We demonstrate with our ensemble framework that cell line specific random forest classifiers built on 10 kb resolution Hi-C data, with distance-type predictors, using SMOTE (*Synthetic Minority Oversampling Technique*) re-sampling yield optimal predictive performances, and outperform existing methods. Cell specific transcription factor binding sites (TFBS) were the best type of genomic annotations for predicting TAD boundaries when directly compared with histone modifications and chromatin states; we identified the well-known CTCF, SMC3, RAD21, and ZNF143 TFBS known to be enriched at TAD boundaries. Additionally, TAD boundary regions predicted at basepair resolution using our optimally built cell line specific models were found to be significantly closer to known enriched epigenomic marks indicating more accurate identification of TAD boundaries than existing Hi-C sequencing technologies.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.


# ABSTRACT (MD version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUROC) curve, Precision-Recall curve (AUPRC), F1, Matthews Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were the best type of genomic annotations for predicting TAD boundaries when directly compared with histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# Introduction

The advent of chromosome conformation capture (3C) sequencing technologies, and its successor Hi-C, have revealed important 3-dimensional (3D) spatial constructs within the human genome such as loops [@rao20143d], Topologically Associated Domains (TADs) [@nora2012spatial;@dixon2012topological], and A/B compartments [@lieberman2009comprehensive]. Among these, TADs represent regions on the linear genome, at megabase scale, that are highly self-interacting. Emerging evidence has linked TADs to critical roles in cell dynamics like gene regulation and cell differentiation, an example being the mechanism by which enhancers loop to come into contact with distal promoters to initiate transcription within these domains [@Lupianez:2015aa; @Franke:2016aa]. Likewise, disruption of TAD boundaries has been shown to lead to developmental disorders [@lupianez2016breaking], cancer [@hnisz2016activation; @taberlay2016three] and other disorders [@Sun:2018aa; @Meaburn:2007aa]. Therefore, identifying the location of TAD boundaries remains a top priority in our goal to understand the functionality of the human genome.

A key factor in locating TAD boundaries is "resolution". Due to the high-dimensional nature of Hi-C data, genome binning has been a common strategy for dealing with a low signal-to-noise ratio problem [@li2019comparison]. Resolution, therefore, determines the bin sizes of the linear genome used to enumerate contacts caused by chromatin folding. Lower resolutions contribute to larger bin sizes, which lead to variability in the location of TAD boundaries. Due to high sequencing costs, most available Hi-C datasets have relatively low resolution, ranging from 25 kb to 100 kb. Because increasing the resolution of Hi-C data requires a quadratic increase in the total sequencing depth, obtaining high resolution remains difficult [@schmitt2016genome]. While these low resolution Hi-C datasets can be used to probe higher order genomic patterns such as A/B compartments, more refined resolutions are necessary to accurately define sub-domains and the boundaries that demarcate them [@fudenberg2016formation]. Furthermore, even at its highest (750 bp [@bonev2017multiscale]), Hi-C sequencing resolution still remains well below the resolution of ChIP-Sequencing experiments used to annotate the genome, which ranges in the tens of base pairs [@mendoza2009genome]. This resolution gap between 1D and 3D genomic maps can make it difficult to disentangle the role of genomic factors related to TAD boundaries, thereby limiting our understanding of epigenomic regulation and its potential link to chromatin architecture. We hypothesize that, by leveraging genomic annotations--defined by high resolution ChIP-Seq data--known to be enriched at TAD boundaries, we may be able more to precisely predict the locations of TAD boundaries at higher resolutions.

Several methods have been developed to predict TAD boundary location using genomic annotations. However, most ignore key characteristics of 3D genomic data that are detrimental to both model performance and accurate boundary identification. A method developed by Mourad et al [@mourad2016computational], called *HiCFeat*, used the percentage of overlap between several ChIP-seq defined transcription factor binding site (TFBS) regions and 1 kb genomic bins to build an L1-regularized multiple logistic regression model to predict TAD boundary regions. However, such high level of resolution is likely to introduce heavily imbalanced classes created from the proportionally much smaller number of TAD boundary regions compared to non-TAD boundary regions throughout the linear genome. *HiCFeat* did not the address class imbalance. Instead, model performance was evaluated using area under the receiver operating characteristic curve (AUROC) which is known to be insensitive to class imbalance, creating artificially inflated values influenced by the majority class [@prati2011survey; @wei2013role]. <!--A more appropriate evaluation tool to consider would be a composite metric such as Matthews Correlation Coefficient (MCC), which penalizes for both false positives and false negatives.--> Furthermore, by only considering the percentage of overlap in defining the feature space, *HiCFeat* is limited in its granularity given that many regions on the linear genome will not be overlapped with any genomic annotations. <!--This, likely being further exacerbated by the class imbalance.--> Two additional studies were proposed in 2015 and 2017 respectively, one using histone modifications in a Bayesian Additive Regression Trees (BART) model [@huang2015predicting] and the other using combined sets of TFBS, DNase I hypersensitive sites, and histone modifications together in an L2-based regularized linear model [@hong2017computational]. The BART model was built on relatively high resolution Hi-C data at 5 kb, while the L2-regularized model used 200 kb resolution Hi-C data. Both methods addressed class imbalance by performing random under-sampling (RUS), and used the elemental read count that appeared in each bin to describe the relationship between ChIP-seq annotations and genomic regions. Firstly, given the large disparity between the number of TAD vs non-TAD boundary regions, performing random under-sampling alone is likely to be unstable and introduce bias. Also, it is unclear how other re-sampling solutions compare to simple random under-sampling such as random over-sampling or a weighted combination of both random under- and over-sampling together. Secondly, using read count overlap as the feature space suffers from similar limitations as the percentage of overlap. Moreover, it is unclear how either of these feature engineering steps compare to enumerating the distance in base pairs between ChIP-seq annotations and genomic bins in TAD boundary prediction, which offer a more spatial measure of association. 

Given the many pitfalls encountered by previous methods we set out to establish a unified approach toward TAD boundary prediction. In this study we first developed a computational framework aimed at determining the set of data level characteristics necessary for optimal TAD boundary prediction. We assessed the Matthews Correlation Coefficient (MCC) of 10-fold cross-validated random forest classifiers with different combinations of model inputs consisting of vary data resolutions, re-sampling techniques, and feature engineering steps. We showed that 10 kb resolution data, combined with both random over- and under-sampling, and a feature space composed of distances between ChIP-seq annotations and genomic bins yielded the best model performances for TAD boundary prediction. We showed that our optimally built model was able to identify known architectural proteins CTCF, RAD21, SMC3, and ZNF143 as exhibiting high predictive importance, confirming their biological functionality near TAD boundaries. We then aimed to use our optimally built model to more precisely predict the location of TAD boundaries at base pair resolution. We showed that, compared to TAD boundaries identified by a TAD caller [@durand2016juicebox], these top ranked transcription factors were much more enriched near the predicted TAD boundaries identified by our model indicating much more precise locations throughout the linear genome. This improved precision in TAD boundary location can provide insight into the role that epigenomics has on the spatial architecture of the human genome.

<!--
Extensive research using high resolution maps (ranging from 750 bp to 1 mb) have probed much of the aspects of 3D chromatin architecture identifying several epigenomic modifications associated with TAD boundaries [@rao20143d; @bonev2017multiscale]. For instance, it is known that the CCCTC-binding factor (CTCF) is enriched at TAD boundaries acting as an insulator interfering with enhancer-promotor interactions [@ghirlando2016ctcf; @song2017ctcf; @ong2014ctcf]. Additional transcription factors including SMC1, SMC3, and RAD21 have also been shown to colocalize near TAD boundaries forming a cohesin ring under the proposed loop extrusion model, whereby loops are formed during interphase [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. 
-->

# Methods

## Data sources

Within this study we used both TAD boundary and loop boundary data derived from the publically available tools ARROWHEAD [@durand2016juicebox] and PEAKACHU [@salameh2019supervised] respectively. ARROWHEAD is part of the Juicer suite of tools for Hi-C data analysis and visualization. To obtain TAD boundaries we applied ARROWHEAD using downloaded primary Hi-C data from Geo GSE63525 [@rao20143d] for the GM12878 and K562 cell lines (libraries HIC001-HIC018 and HIC069-HIC074 for GM12878 and K562 respectively) produced with the MboI restriction enzyme (Supplementary Table 1). The contact matrices were constructed from all read pairs that uniquely aligned to the genome (MAPQ>0). The default tools for ARROWHEAD were used when extracting the autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries at 5 kb, 10 kb, 25 kb, 50 kb, and 100 kb resolutions (Additional File 1). For chromatin loop boundaries, we used loops identified by PEAKACHU at 10 kb for the GM12878 and K562 cell lines [@salameh2019supervised]. Unique boundaries were considered as the midpoints within the coordinate of each chromatin loop. Chromosome 9 was excluded from all downstream analyses due to the sparsity of the contact matrix at the 5 kb and 10 kb resolutions for the K562 cell line. Cell line-specific genomic annotations were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] in the form of BroadHMM chromatin segmentations (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS) (GRCh37/hg19 human genome assembly, Supplementary Table 2). 

## Computational framework for boundary prediction

In Hi-C, each chromosome is binned into non-overlapping regions of the same length *r*, typically 5kb and above. The *r* parameter defines the resolution of Hi-C data. Consequently, the location of TAD boundaries is limited by the resolution *r*. We designed a strategy called _shifted binning_ that partitions the genome into regions of the same length *r*, but with middle points corresponding to boundaries defined by the original binning. 

To create shifted binning, the first chromosome-specific bin was set at half of the resolution *r*, and continued in congruent intervals of length *r* until the end of the chromosome (*mod r* + *r/2*). The shifted genomic bins, referred hereafted as bins for simplicity, were then defined as boundary regions (*Y = 1*) if they contained a TAD (or loop) boundary, and non-boundary regions (*Y = 0*) otherwise, thus establishing the binary response vector (**Y**) used for classification (Figure 1).

Cell type-specific genomic annotations were used to build the feature space for our boundary prediction framework. Bins were annotated by one of either the number of overlaps with an annotation (*Overlap Count (OC)*), the percent of overlap between the bin and the total width of genomic annotation regions overlapping it (*Overlap Percent (OP)*), the average signal strength of the corresponding annotation (*Peak Signal Strength*), or the distance in bases from the center of the bin to the center of the nearest genomic annotation region (*Distance*). A $log_2$-transformation of distance was used to account for the skewness present in the distributions of genomic distances (Supplementary Figure 1). It should be noted that models built using a *Peak Signal Strength* feature space were only composed of histone modifications and transcription factor binding sites. That is, they did not include BroadHMM chromatin segmentations.

Chromosome-specific datasets used for modeling were composed of two components: the response vector $Y_{i}$, an indicator variable distinguishing boundary and non-boundary regions, and the predictor-type specific data matrix, $\textbf{X}_{ij}$, where $i \in \{ 1,2, \cdots,N \}$ ($N$ being the number of chromosome-specific bins) and $j \in \{ 1,2,\cdots,P \}$ ($P$ being the number of genomic annotations). <!--Chromosome-specific datasets were concatenated to form a whole genome dataset. The dataset was then split into a 7:3 ratio for training and testing, respectively.[!!!only comes into play with preciseTAD]-->

## Class imbalance assessment

To assess the impact of class imbalance (CI), defined as the proportion of boundary regions to non-boundary regions, we evaluated three re-sampling techniques including *Random Over-Sampling (ROS)*, *Random Under-Sampling (RUS)*, and *Synthetic Minority Over-Sampling Technique (SMOTE)*. For ROS, the minority class was sampled with replacement to obtain the same number of data points in the majority class. For RUS, the majority class was sampled without replacement to obtain the same number of data points in the minority class. For SMOTE, under-sampling was performed without replacement from the majority class, while over-sampling was performed by creating new synthetic observations using the $k=5$ minority class nearest neighbors [@chawla2002smote] (implemented in the `DMwR` v.0.4.1 R package). We restricted the algorithm to 100% over-sampling and 200% under-sampling, which creates perfectly balanced classes. 

## Assessment of model performance

Random forest (RF) classification models (the `caret` v.6.0 R package [@kuhn2012caret]) were built to compare model performances between combinations of data resolutions, predictor types, and re-sampling techniques. Five-fold cross validation was used to tune the number of annotations, *P*, to consider at each node. The number of trees that were aggregated for each RF model was set to the default, $n=500$. The performance of each model was compared using Balanced Accuracy (BA), defined as the average of sensitivity and specificity:

<!--
$$F1 - Score = \frac{2TP}{2TP + FP + FN}$$
-->

<!--
$$
MCC = \frac{TP \times TN - FP \times FN}{\sqrt{ (TP + FP)(TP+FN)(TN+FP)(TN+FN) }}
$$
-->

$$BA = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$

<!--
$$Informedness = \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} - 1$$
-->

where TP refers to the number of bins correctly identified as containing a boundary (true positives), FP refers to the number of bins incorrectly identified as containing a boundary (false positives), TN refers to the number of bins correctly identified as not containing a boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a boundary (false negatives). Each of these quantities are obtained from the confusion matrix created by validating the model on the testing set. <!--We refer to the complete computational framework used to compare random forest model inputs as *TADRF*.-->

## Feature selection and predictive importance

Many genomic annotations, notably architectural proteins, tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To suitably reduce the feature space and improve computational efficiency, while maintaining optimal performance, we implemented recursive feature elimination. In doing so, we estimated the near optimal number of necessary features to include in the best fitting RF model. The best performing subset was identified, ranging from 2 to the maximum number of features by power of 2. We then aggregated the predictive importance of the union of the optimal set of features across chromosomes using the mean decrease in node impurity among permuted features in out-of-bag samples to determine the most common and top ranked annotations for predicting boundary regions. 

## Boundary prediction at bp resolution 

To investigate whether we could alleviate resolution limitations of conventional TAD and loop boundary calling algorithms, we applied our RF model to predict boundaries from base pair (bp) resolution data. [??? Write how your vector Y and annotation matrices change;!!!added ???You describe whole-genome model. It should have been described early on. Here, describe chromosome-specific bp-resolution model]. At this point, we opted to implement an optimally built whole genome RF model. That is, a random forest model built on the optimal combination of data resolution (binning), predictor type (feature space), re-sampling technique (class balancing correction), and annotations (most informative genomic elements). Here, $Y$ was of length $N$ where $N=n_1+n_2+\cdots+n_{22}$, the total number of bins across each chromosome. The data matrix $\textbf{X}$ had dimensions $N \times p_s$, where $p_s \subset P$, a subset of the most predictive genomic elements. Given that each chromosome is composed of tens, in some cases hundreds, of millions of base pairs, we used the same whole genome RF model to predict boundaries from chromosome-specific bp resolution data.

Naturally, when dealing with such fine resolution, there will be spatially close base pairs with similar predictor profiles (i.e. similar overlap counts, distances). To guard against inflating the number of predicted TAD boundaries we developed a technique that combines both density-based spatial clustering (DBSCAN) and partitioning techniques to identify precise TAD boundaries throughout the linear genome. 

First, we extracted the vector $p_{n_{i}}$ that denoted the probability of being a boundary point for each chromosome-specific base pair from our RF model, where $n_{i}$ was the length of chromosome $i$. We then applied *DBSCAN* (version 1.1-5) to the matrix of pairwise distances [??? Is that true? Matrix is 2-dimensional] between base pairs with $p_{n_{i}}=1$. The DBSCAN algorithm takes two parameters, *eps* ($\epsilon$) and *MinPts* [@hahsler2019dbscan]. The eps parameter was assigned as the respective data resolution [??? Meaning?]. The heuristic being that base pairs with genomic distances less than resolution should occupy the same designated cluster. The corresponding MinPts parameter was set to 3 [??? Meaning?]. The resulting clusters of density-reachable base pairs identified by DBSCAN were termed *predicted TAD boundary regions* (PTBR) [??? Use simple words: "boundary region" (many bases clustered by DBSCAN), then "boundary base" (selected by PAM)]. To precisely identify the most representative base pair among each PTBR we performed an extension of partitioning around medoids (PAM), known as CLARA (Clustering Large Applications) for better scaling and more computational efficiency [@reynolds2006clustering; @wei2000empirical]. Thusly, the corresponding cluster medoid was defined as a *preciseTAD boundary point* (PTBP), making it the most representative base pair within each clustered PTBR.


<!--
We then performed hierarchical clustering on the distance matrix (in base pairs) constructed from the midpoints of each PTBA using complete linkage clustering. To identify the optimal number of clusters, *k*, we calculated the mean silhouette score over all possible sizes of clusters *m*:

$$\bar{s}_{m} = \dfrac{\sum^{m}_{i} s_{i}}{m},$$

where $s_{i}$ is the cluster-specific silhouette score defined as,

$$s_{i} = \dfrac{b_{i} - a_{i}}{max[a_{i},b_{i}]}.$$

Here, $a_{i}$ is the mean distance between each cluster entry and all other observations in the same cluster, and $b_{i}$ is the mean distance between entries within the same cluster. Therefore number of clusters, *k*, that maximized the mean silhouette score, $\bar{s}_{m}$ were used to identify groupings of PTBA, which we defined as *predicted TAD boundary regions* (PTBR).
-->

<!--
Using optimally trained models, we established *predicted TAD boundary regions* as a collection of adjacent bases predicted as TAD boundary with the probability of 1. [??? Do not understand from here:] To guard against inflating the number of predicted TAD boundaries, we clustered the distances between the midpoints of the *predicted TAD boundary regions*. We then performed partitioning around mediods (PAM) to determine the *predicted TAD boundary region* that was most representative of the cluster. The midpoint, in this case a particular basepair, of the *predicted TAD boundary region* identified as the medoid was defined as a final *predicted TAD boundary*. [??? To here. REWRITE!]
-->

## Comparing called vs predicted boundaries

Next, we evaluated the biological significance of our predicted boundaries with respect to the enrichment of several known molecular drivers of 3D chromatin architecture including CTCF, RAD21, SMC3, and ZNF143 ChIP-Seq peaks. Enrichmed heatmaps were used to compare preciseTAD boundaries to both ARROWHEAD and PEAKACHU called boundaries using the `EnrichedHeatmap` R package (version 1.14.0) [@gu2018enrichedheatmap]. For each boundary (both called and predicted), a 5000 bp flank was created on either side. The flanking was broken up into 50 bp segments, for a total of 100 windows on both sides of a given boundary. A coverage value was generated for each window with respect to a given ChIP-seq annotation, as the mean signal averaged by the width of the window, denoted by,

$$v_{c} = \dfrac{\Sigma_{j}^{m} \Sigma_{i}^{n}x_{ij}w_{ij}}{L}$$
where $x_{ij}$ is the total number of base pairs of annotation $i$ in window $j$ and $w_{ij}$ is the number of base pairs that overlap between annotation $i$ and window $j$. The denominator, $L$, is the width of the windows (here, $L=50$). The rows of the heatmaps are ordered by enriched scores. The enriched scores are calculated as the sum of coverage values weighted by the distance to the flanked boundary, denoted by

$$s_{e} = \Sigma_{d=1}^{n_{1}}\left(\dfrac{x_{1d}* d}{n_{1}} \right) + \Sigma_{u=1}^{n_{2}}\left(\dfrac{x_{2u}* (n_{2} - u + 1)}{n_{2}} \right)$$
where $n_{1}$ and $n_{2}$ represent the number of downstream and upstream coverage values to sum over. Here, $n_{1}=n_{2}=100$. 

Additionally, we compared the $log_2$ distance between TAD boundaries and the same top predictive transcription factor binding sites using Wilcoxon Rank-Sum test. The alpha level of significance was set to 0.05.

Furthermore, under the hypothesis that TAD boundaries are highly concerved across cell types [@dixon2012topological; @nora2012spatial; @dixon2015chromatin; @rao20143d; @Schmitt2016aa], we assessed the conservation of our predicted TAD boundaries across cell types and prediction algorithms. Boundaries were flanked by resolution and overlaps were visualized using Venn diagrams from the `Vennerable` R package (version 3.1.0) and evaluated using the Jaccard index defined as

$$J_{(A,B)} = \dfrac{A \cap B}{A \cup B}$$

where A and B represent genomic regions created by flanking called and predicted boundaries. 

# Results

## Developing a computational framework for optimal TAD boundary prediction

We developed a computational framework<!--, refered to as *TADRF*,--> aimed at determining the optimal set of data level characteristics necessary for topologically associated domain (TAD) boundary prediction. We chose the random forest (RF) machine learning algorithm as our binary classification tool. The reason for choosing RF was two-fold: (1) In order to devise a tunable prediction rule in a supervised learning framework that is both robust to overfitting and able to handle multiple correlated predictors, and (2) to allow for an interpretable ranking of predictors with respect to their ability to predict boundary regions [@boulesteix2012overview]. We then leveraged our optimally built RF to precisely predict the genomic coordinates of both TAD and loop boundaries at base pair resolution. 

Our framework is resolution-specific and uses arbitrary called TAD and loop boundaries as ground truth to establish TAD(loop)-boundary and non-TAD(loop) boundary regions. To demonstrate, we applied our framework to publically available replicate Hi-C data sets from the GM12878 and K562 cell lines (Supplementary Table 1). TADs were subsequently called using the ARROWHEAD pipeline from the juicer toolbox [@durand2016juicer] at five data resolutions: 5 kb 10 kb, 25 kb, 50 kb, and 100 kb. For loop boundaries we opted to use boundaries obtained by the PEAKACHU tool for the GM12878 and K562 cell lines at 10 kb resolution [@salameh2019supervised]. [??? TAD and loop boundaries should be distinguished, to describe when we have all results]

Boundary regions were defined as regions containing a called boundary ($Y = 1$), while non-boundary regions were defined as genomic bins that did not contain a boundary ($Y = 0$) (Figure 1A, see Methods). The total number of TADs, TAD boundaries, and the number of genomic bins expectedly decreased with the decreased resolution of Hi-C data (Table 1, Supplementary Table 3). The number of non-boundary regions highly outnumbered boundary regions, and these results were consistent for loop boundaries (Supplementary Table 4). Such disproportional presence of examples in one class is known as a "class imbalance" problem that negatively affects model performance [@prati2011survey; @wei2013role]. To address class imbalance, we evaluated the effect of three re-sampling techniques on prediction performance. *Random over-sampling* (ROS) was defined as sampling with replacement from boundary regions. *Random under-sampling* (RUS) was defined as sampling with replacement from non-boundary regions. Lastly, we considered *Synthetic minority over-sampling technique* (SMOTE), which is a combination of both random over- and under-sampling [@chawla2002smote] (see Methods). These strategies allowed us to select the best performing model.

A total of 78 cell line specific functional genomic annotations were used to build the feature space of downstream random forest models. These annotations included BroadHMM shromatin segmentations, histone modifications (HM), and transcription factor binding sites (TFBS, Supplementary Table 2). Four feature engineering procedures were developed for each genomic annotation (Figure 1B, see Methods) with the goal to improve boundary predictions by accounting for signal strength association (Signal), direct (OC), proportional (OP), and spatial (Distance) relationships between TAD boundaries and genomic annotations.

![**Illustrating the resolution-specific data construction and feature engineering used for downstream random forest modeling.** (A) The linear genome was subsequently binned into resolution-specific intervals. The response vector **Y** was defined as a genomic bin that either overlapped with a TAD boundary ($Y = 1$) or did not ($Y = 0$). (B) Four types of associations between bins and genomic annotations (green shapes) were then used to build the feature space including _Peak Signal_ (Signal), _Overlap Counts_ (OC), _Overlap Percent_ (OP), and ($log_{2}$) _Distance_.](Figures/figure1_model_construction_and_predictor_types.png)

In summary, we considered combinations of data for a particular cell line *L = {GM12878, K562}*, at a given resolution *R = {5 kb, 10 kb, 25 kb, 50 kb, 100 kb}*, with a specific engineered feature space *P = {Signal, OC, OP, Distance}*, and re-sampling technique *S = {None, RUS, ROS, SMOTE}* (Figure 2). Once the model inputs were established, a random forest classifier was trained on chromosome-specific genomic bins (excluding chromosome 9), *C = {CHR1, ..., CHR22}*, using 5-fold cross-validation. Model inputs were evaluated by assessing the balanced accuracy (BA) of each combination. <!--The results presented are for the ARROWHEAD boundaries, with PEAKACHU boundaries presented in the Supplementary material. Letting $|X|$ denote the cardinality of each set, we evaluated $|L| \times |R| \times |P| \times |S| \times |C|$ different ensemble systems, totaling $2 \times 5 \times 4 \times 4 \times 21 (= 3360)$ models.--> [??? When we have all results, Peakachu results will likely need to be described in parallel. They are about loops, theoretically different from TAD boundaries. For discussion]

![**A diagram of the ensemble framework detailing the steps for combining different inputs used in the random forest model building pipeline.** Step 1 is to establish the chromosome-specific rowise (genomic bins) and columnwise (genomic annotations) dimensions of the data matrix, **A**, by combining a resolution (5 kb, 10 kb, 25 kb, 50 kb, or 100 kb) and feature engineering procedure (overlap count (OC), overlap percent (OP), $log_{2}$ distance (Distance)). Step 2 is to split **A** into a training and testing set according to a 7:3 ratio. Step 3 is to apply a re-sampling technique to the training set (no re-sampling, random over-sampling, random under-sampling, or SMOTE re-sampling). Step 4 is to implement the random forest algorithm and perform 5-fold cross-validation to tune for the *mtry* parameter. Finally, step 5 is to validate the model on the separate test data and evaluate model performance using balanced accuracy (BA).](Figures/figure2_ensemble_framework.png)


<!-- Using ARROWHEAD, we obtained a total of 8052, 7676, 4670, 2349, and 1031 called TADs  across each resolution respectively (Table 1). From the called TADs, a total of 15468, 14253, 8363, 4224, and 1883 unique TAD boundaries were obtained (Table 1). The resolution-specific approach we developed to create boundary vs non-boundary regions resulted in a total of 535363, 267682, 107073, 53537, and 26768 genomic bins. The class imbalance among the genomic bins was both severe and invariant toward resolution, resulting in 0.03, 0.05, 0.08, 0.08, and 0.07 for each resolution respectively (Table 1). Boundary data summaries were similar for K562 cell line, as well as for PEAKACHU defined boundaries (Supplementary Tables 3-4). -->

| Resolution | Total number of called TADs | Total number of unique TAD boundaries | Total number of genomic bins | Class imbalance |
|:----------:|:---------------------------:|:-------------------------------------:|:----------------------------:|:---------------:|
|5 kb        |8052                         | 15468                                 | 535363                       | 0.03            |
|10 kb	     |7676 	                       | 14253                                 | 267682                       | 0.05            |
|25 kb	     |4670                         | 8363                                  | 107073                       | 0.08            |
|50 kb	     |2349	                       | 4224                                  | 53537                        | 0.08            |
|100 kb	     |1031                         | 1883                                  | 26768                        | 0.07            |
Table: Domain data and class imbalance summaries across resolutions for GM12878.

## 5 kb resolution, random under-sampling, and distance-based predictors provided optimal performance for TAD boundary prediction


Expectedly, when using data with class imbalance present, that is, the data were used as-is, the models exhibited low balanced accuracies, with very little variability among different resolutions (Figure 3). Similarly, poor performances were found when using ROS as well. However, RUS and SMOTE re-sampling led to drastic improvement in performance, especially at higher resolutions. We found that RUS marginally outperformed SMOTE as the most optimal class balancing technique for all resolutions and predictor types. These results suggest that simple randomly under-sampling alone can drastically improve TAD boundary prediction. 

Additionally, we found that at higher resolutions, random forest classifiers built using a distance-type feature space yielded substantially higher balanced accuracies when compared to the peak signal stength, overlap count, and overlap percent predictor types. As with comparing class balancing techniques, this improvement was less evident at lower resolutions. These results were found to be consistent for the K562 cell line (Supplementary Figure 2). Random forest models built on PEAKACHU loop boundaries with SMOTE resampling marginally outperformed RUS for both GM12878 and K562 cell lines (Supplementary Figure 3). In summary, random under-sampling and SMOTE, distance-type predictors, and Hi-C data at 5kb resolution yields an optimally performing cell line-specific random forest classification model for TAD and loop boundary predictions. 

<!--
Interestingly, for the optimal re-sampling and predictor type, we found that performances peaked at 10 kb resolution and then degraded as data resolution decreased. The reason for the improved performance between 5 kb and 10 kb data resolutions can be attributed to the sparsity in the interaction frequency matrices at finer resolutions. This sparsity may lead to inaccurate TAD calling and, subsequently, to worse model performance. These results suggest that finer resolution of Hi-C data is not always optimal for accurate TAD calling and prediction, and sparsity of the data needs to be considered. Furthermore, performances tended to be much more variable between chromosomes at lower resolutions, evidenced by the wider error bars for the 50 kb and 100 kb resolutions. These results were found to be consistent for the K562 cell line (Supplementary Figure 2). In summary, our computational framework showed that using random under-sampling, distance-type predictors, and Hi-C data at 10kb resolution yields an optimally performing cell line-specific random forest classification model for TAD boundary prediction.
-->

![**Models built using distance-type predictors combined with random under-sampling at 5 kb yielded optimal predictive performance for GM12878.** Line plots comparing the performances of random forest classifiers built using different combinations of data level characteristics. Performances are compared across resolution, within each predictor-type: (A) Signal, (B) OC, (C) OP, and (D) Distance, and between each re-sampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Performances were evaluated using balanced accuracy.](Figures/figure3_performances.png)

## Transcription factor binding sites outperform HM- and BroadHMM-specific models

We hypothesized that the type of genomic annotations may affect the predictive performance. Using the aforementioned optimal settings, we built separate random forest models using BroadHMM chromatin segmentations, histone modifications (HM), and transcription factor binding sites (TFBS). We found that models built on TFBSs outperformed other annotation-specific models, and these results were consistent for loop boundaries (Figure 4A; Supplementary Figures 4A,5A,6A). These results suggest that TFBS are the primary drivers of TAD and loop boundary formation in both the GM12878 and K562 cell lines. 

## Predictive importances confirmed biological functionality of CTCF, RAD21, SMC3, and ZNF143 at TAD boundaries

We sought to further optimize our TAD prediction model by performing recursive feature elimination to avoid overfitting and select only the most influential TFBS across all chromosomes. We were able to obtain near-optimal performance using approximately eight TFBS (Figure 4B; Supplementary Figures 4B,5B,6B). However, given that we trained our model on chromosome-specific data, the most significant annotations were found to vary for each chromosome. To determine transcription factors most important for boundary predigtion, we clustered the predictive importances of top eight significant TFs across chromosomes. We found that, similar to other studies [@sanborn2015chromatin; @weintraub2017yy1; @gonzalez2017znf143], CTCF, RAD21, SMC3, and ZNF143 transcription factors were most predictive of TAD and loop boundaries for all chromosomes (Figure 4C; Supplementary Figures 4C,5C,6C). We further optimized our model by only considering these top informative TFBS when building the random forest, thereby decreasing computational burden while maintaining high predictive performance. In summary, our framework was able to yield the top molecular drivers of the 3D architecture of the human genome, components of the popular loop extrusion model [@fudenberg2016formation; @alipour2012self]. 

<!--
To summarize, the optimally built cell line-specific random forest model under our computational framework was found to be one built using 10 kb resolution data, leveraging distance-type predictors for the CTCF, RAD21, SMC3, and ZNF143 transcription factor binding sites, while implementing random under-sampling to balance TAD and non-TAD boundary defined genomic bins.
-->

![**Assessing top ranked transcription factors in TAD boundary prediction for GM12878.** (A) Barplots comparing performances of models using specific classes of functional genomic elements as the feature space including histone modifications (HM; red), cis-regulated elements (BroadHMM; green), transcription factor binding sites (TFBS; blue), in addition to a model containing all three classes (ALL; purple). (B) Recursive feature elimination (RFE) analysis used to reduce the number of TFBS to include in the optimally built random forest model. Error bars represent standard deviation from the cross-validated accuracy across each chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes as determined from performing RFE.](Figures/figure4_annots_specifc_rfe_importances.png)

## *preciseTAD* identifies precise and biologically relevant locations of TAD boundaries

After obtaining our optimally built random forest model, we investigated whether we could leverage our model to alleviate resolution limitations of Hi-C data by predicting TAD boundaries at base pair resolution. Our method, referred to as *preciseTAD*, leverages a user specified binary classifier with a combination of density-based hierarchical clustering and a scalable version of partitioning around mediods (PAM), known as clustering large applications (CLARA) to precisely identify the most representative and biologically meaningful predicted chromosome-specific domain boundaries (see Methods). 

Here, we opted to implement preciseTAD using an optimally built whole genome random forest model. That is, a random forest model built on 5 kb defined genomic bins, with random under-sampling, and distance-type predictors for the top informative TFBS (CTCF, RAD21, SMC3, and ZNF143). The reason for this was because of the improved performance that was exhibited for a whole genome random forest compared to aggregated chromosome-specific models (Supplementary Figure 7).

???stopped here; waiting to predicted tad boundaries results

<!--
We first used ROC analysis to obtain an optimal probability threshold for defining potential predicted TAD boundaries. We reported an average threshold of ??? (SD=???) among the 21 autosomal chromosomes that were analyzed (mean=???, SD=??? for K562). We referred to collections of chromosome-specific base pairs that exceeded this threshold when used to validate our *TADRF* model as *predicted TAD boundary areas* (PTBA), given that collections of adjacent base pairs tended to exhibit similar prediction profiles. We found that there were a total of ??? PTBA, containing an average of ??? base pairs each (??? PTBA for K562; mean base pairs=???, SD=???). *preciseTAD* then implements complete linkage hierarchical clustering on the dissimilarity matrix constructed from the pairwise distances (in base pairs) among the midpoints of each chromsome-specific PTBA. Using the silhouette score (see methods), we reported a total of k=??? chromosome-specific clusters (k=??? for K562). We referred to these *k* clusters as *predicted TAD boundary regions* (PTBR). That is, each PTBR contained the midpoints of multiple PTBA. Lastly, given each PTBR, *preciseTAD* uses CLARA partitioning to obtain the most representative midpoint. We referred to this as the *predicted TAD boundary point* (PTBP). These PTBP were identified as the most representative base pair among each PTBR. *preciseTAD* was able to identify a total of ??? PTBP across each chromosome. We reported less predicted TAD boundaries than ARROWHEAD at 10 kb (Table 1). This is attributed to ARROWHEAD's inherent inflation of called TAD boundaries, especially near fuzzy areas along the diagonal of Hi-C contact matrices (Figure 5A). Likewise, under the framework of *preciseTAD*, any two TADs are inherently demarcated by only one predicted TAD boundary, leading to comparatively less TAD boundaries. We also find that predicted TAD boundaries identified by *preciseTAD* were much more likely to demarcate high concentrations of inter-chromosomal contacts among Hi-C contact matrices (Figure 5A). [???include a result that demonstrates this; discuss in meeting]
-->

<!--We were able to identify regions where the model predicted the base pair to be a TAD boundary with probability equal to 1 (Figure 5A). [??? Where is this figure? It seemed important; !!!added] By clustering the regions (see Methods), we were able to predict a comparable number of TAD boundary regions as called by the ARROWHEAD TAD caller. However, only 24% and 29% of predicted and called (and 10 kb flanked) TAD boundaries overlapped for the GM12878 and K562 cell line, respectively, suggesting differing locations in TAD boundaries throughout the linear genome (Supplementary Figure 5). [??? Need further drill down, as we discussed; !!!added distance to annotations figure in next paragraph] 

For each predicted TAD boundary region, we were able to identify the summit locations of TAD boundary location by performing partitioning around medoids (see Methods). To compare the summit locations of predicted and called TAD boundaries, we evaluated the signal strength of ChIP-seq peak regions around a 10 kb flanking region. We found that the signal strength for the CTCF, RAD21, SMC3, and ZNF143 [??? Make all in this order!; !!!fixed] TFBS peaked at our predicted TAD boudaries, while was poorly defined around the called TAD boundaries (Figure 5B; Supplementary Figure 6).  Thus, our model is able to alleviate resolution limitations and precisely locate TAD boundaries that correspond to the highest signal strength of known TAD boundary-forming TFBSs. Furthermore, when compared to called TAD boundaries, the boundaries that we predicted with our model were found to be statistically significantly closer to the top ranked TFBS, suggesting stronger enrichment profiles near predicted TAD boundaries (Supplementary Figure 7).


![**Predicted TAD boundaries are enriched in known TAD boundary-forming TFBSs.** (A) Visualizing locations of predicted and called TAD boundaries (chr 22: 17389000-18011000; GM12878). The upper triangular contact map shows the normalized number of contacts within the specified window using the JUICER toolbox.The blue vertical lines represent called boundaries identified by ARROWHEAD, with the corresponding called TADs outlined in blue on the upper diagonal contact matrix above. The black line represents the predicted probability of each base pair being a TAD boundary when validated by *TADRF*. Collections of red vertical lines represent base pairs whose predictive probability exceed the ROC threshold, labeled as *predicted TAD boundary areas* (PTBA). The midpoints of spatially similar PTBA were clustered together into *predicted TAD boundary regions* (PTBR), denoted by the yellow regions along the x-axis. The green vertical lines represent the *predicted TAD boundary points* (PTBA) identified by *preciseTAD*, with the corresponding predicted TADs outlined above in green.](Figures/visualizing_ptbp_and_ctbp.png) 
-->

# Discussion

<!--Additional performance metrics were reported for both cell lines including positive predictive value (PPV; precision), F1-Score (F1), Accuracy, Area Under the Receiver Operating Characteristic Curve (AUC), Area Under the Precision-Recall Curve.-->

In this study we developed an ensemble framework to systematically compare the effects of different combinations of Hi-C resolution, data re-sampling, and feature engineering techniques on the performance of models aimed at TAD boundary prediction. By implementing our framework across multiple cell lines, we showed that there exists a consensus approach toward precise TAD boundary prediction. We showed that random forest classification models trained on domain data extracted from 10 kb resolution maps, using random over- and under-sampling to account for class imbalance, with a distance-type predictor space yielded optimal model performances for both the GM12878 and K562 cell lines. 

Previous studies tended to focus only on specific sets of genomic annotations including chromatin states, histone modifications, or transcription factor binding sites (TFBS) alone. Using our framework, we were able to conclude that simply building random forest classifiers only on cell line specific TFBS can yield near optimal model performance, thereby significantly reducing the feature space and computational complexity of the model. Specifically, we found that the CCCTC-binding factor (CTCF) and cohesin complexes (RAD21 and SMC3) were the most important features of TAD boundary prediction, confirming their enrichment at TAD boundaries as part of the loop extrusion model [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Likewise, we found the the zinc-finger protein ZNF143 was also given high predictive importance. This confirmed the known role of ZNF143 as a chromatin-looping factor that contributes to the structural foundation of the genome by providing sequence specificity at promoters connected with distal regulatory elements [@bailey2015znf143;@gonzalez2017znf143]. 

By leveraging the enrichment of these architectural proteins we were able to more precisely identify TAD boundaries at a much more refined resolution. To do so, we evaluated the ChIP-Seq signal strength between the CTCF, RAD21, SMC3, and ZNF143 transcription factors and a 10 kb flanked window of predicted TAD boundaries identified by our optimally built random forest model. We showed that the genomic signals for each TFBS was much more consolidated near the center of our flanked predicted TAD boundary then the flanked boundary identified by the ARROWHEAD TAD caller. We also evaluated the distal relationships between the same architectural proteins and found that they were statistically significantly closer in base pair distance to our predicted TAD boundaries, than to those identified by ARROWHEAD. 

<!-- Discuss advantages and disadvantages of Balanced Accuracy. The disadvantage is that we restrict TP to the originally detected TAD boundaries. While there are likely many more that couldn't be detected but are TP. We currently count them as FP -->


# Conclusion

In summary, we have shown that TAD boundary prediction is a multi-faceted problem composed of many data irregularities and characteristics that previous methods have failed to address. Through our ensemble framework, we provide a useful guide toward the exploration of chromatin organization, while highlighting the impact of data resolution, class imbalance, and feature engineering in TAD boundary prediction. Lastly, by building models using known enriched transcription factors, we were able to bridge the resolution gap between 1D ChIP-Seq annotations and 3D Hi-C sequencing for more precise TAD boundary prediction. This improved precision in TAD boundary location offers the potential for further exploration into the 3D organization of the human genome.



<!--
The two proteins differ in that YY1 preferentially
occupies interacting enhancers and promoters, whereas CTCF
preferentially occupies sites distal from these regulatory elements that tend to form larger loops and participate in insulation.
Deletion of YY1 binding sites or depletion of YY1 can disrupt
enhancer-promoter contacts and normal gene expression.
Thus, YY1-mediated structuring of enhancer-promoter loops is
analogous to CTCF-mediated structuring of TADs, CTCF contact domains, and insulated neighborhoods. This model of
YY1-mediated structuring of enhancer-promoter loops accounts
for diverse functions reported previously for YY1, including contributions to both gene activation and repression and to gene
dysregulation in cancer. ["YY1 Is a Structural Regulator
of Enhancer-Promoter Loops"]
-->





# Supplementary Legends

**Additional File 1**. An example script for applying the ARROWHEAD tool to in situ Hi-C data (HIC001-HIC018) for resolutions 5kb, 10kb, 25kb, 50kb, 100kb to obtain chromosome-specific TAD boundaries on the GM12878 cell line.

**Supplementary Table 1.** A complete list of the Hi-C data libraries obtained from GEO GSE63525 [@rao20143d] used to obtain the locations of topologically associating domains for ARROWHEAD, as well as the locations for the cell line specific .txt files of loop boundaries obtained by PEAKACHU. [@salameh2019supervised].

**Supplementary Table 2.** A complete list of the genomic annotations used to build the feature space for all downstream models. The GRCh37/hg19 human genome assembly was used. "Genomic Class" - broad category of genomic features, "Element" - names of genomic features, "Cell Type-Specific Source" - download URL specific to the cell line (not all annotations were provided by the same institutions).

**Supplementary Figure 1.** Density curves illustrate the normalization of distances before (red) and after (blue) performing a $log_{2}$ transformation across differing bin sizes including 10 kb, 25 kb, 50 kb, and 100 kb for both the (A) GM12878 and (B) K562 cell lines respectively. Distances are measured as the number of base pairs from the center of a genomic bin to the center of the nearest genomic annotation. Each density curve represents a particular genomic annotation.

**Supplementary Table 3.** Domain boundary and class imbalance summaries from SpectralTAD across resolution for both the
GM12878 and K562 cell lines respectively.

**Supplementary Table 4.** Loop boundary and class imbalance summaries from PEAKACHU 10 kb resolution for both the GM12878 and K562 cell lines respectively.

**Supplementary Figure 2. Models built using distance-type predictors combined with random under-sampling at 5 kb yielded optimal predictive performance for K562.** Line plots comparing the performances of random forest classifiers built using ARROWHEAD boundaries at different combinations of data level characteristics. Performances are compared across resolution, within each predictor-type: (A) Signal (B) OC, (C) OP, and (D) Distance, and between each re-sampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Performances were evaluated using balanced accuracy.

**Supplementary Figure 3. Models built using distance-type predictors combined with SMOTE re-sampling yielded optimal predictive performance for GM12878 K562.** Line plots comparing the performances of random forest classifiers built using using PEAKACHU boundaries at different combinations of data level characteristics for (A) GM12878 and (B) K562. Performances are compared across resampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple), and within each predictor-type: Signal, OC, OP, and Distance. 

**Supplementary Figure 4. Assessing top ranked transcription factors in TAD boundary prediction for K562.** (A) Barplots comparing performances of models built using ARROWHEAD boundaries on specific classes of functional genomic elements as the feature space including histone modifications (HM; red), cis-regulated elements (BroadHMM; green), transcription factor binding sites (TFBS; blue), in addition to a model containing all three classes (ALL; purple). (B) Recursive feature elimination (RFE) analysis used to reduce the number of TFBS to include in the optimally built random forest model. Error bars represent standard deviation from the cross-validated accuracy across each chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes as determined from performing RFE.

**Supplementary Figure 5. Assessing top ranked transcription factors in loop boundary prediction for GM12878.** (A) Barplots comparing performances of models built using PEAKACHU boundaries on specific classes of functional genomic elements as the feature space including histone modifications (HM; red), cis-regulated elements (BroadHMM; green), transcription factor binding sites (TFBS; blue), in addition to a model containing all three classes (ALL; purple). (B) Recursive feature elimination (RFE) analysis used to reduce the number of TFBS to include in the optimally built random forest model. Error bars represent standard deviation from the cross-validated accuracy across each chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes as determined from performing RFE.

**Supplementary Figure 6. Assessing top ranked transcription factors in loop boundary prediction for K562.** (A) Barplots comparing performances of models built using PEAKACHU boundaries on specific classes of functional genomic elements as the feature space including histone modifications (HM; red), cis-regulated elements (BroadHMM; green), transcription factor binding sites (TFBS; blue), in addition to a model containing all three classes (ALL; purple). (B) Recursive feature elimination (RFE) analysis used to reduce the number of TFBS to include in the optimally built random forest model. Error bars represent standard deviation from the cross-validated accuracy across each chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes as determined from performing RFE.

**Supplementary Figure 7. Whole genome random forest model outperform chromosome-specific random forest models across resolutions.** Random forest models built on ARROWHEAD boundaries using 5 kb defined genomic bins, with random under-sampling, and distance-type predictors for the top informative TFBS (CTCF, RAD21, SMC3, and ZNF143) applied to the entire genome (A) exhibit higher balanced accuracies than their chromosome specific counterparts (B) across each resolution respectively.


<!--
**Supplementary Figure 3. The proportional importance of optimally built cell line specific models was outweighed by transcription factor binding sites across resolutions.** Stacked barplots illustrate the proportion of total predictive importance in cell line specific random forest classifiers for (A) GM12878 and (B) K562 respectively. Model inputs included the optimal data level characteristics determined by the ensemble framework (SMOTE re-sampling; with distance-type predictors). Predictive importance was averaged for models built on chromosome specific regions of the linear genome. Proportional importance was then calculated according to the annotation group of the feature space including histone modifications (HM), chromatin state segmentations (ChromHMM), and transcription factor binding sites (TFBS).

**Supplementary Figure 5. Venn diagrams illustrating the amount of overlap between predicted and called TAD boundary regions.** For each of the GM12878 (A) and K562 (B), both predicted and called TAD boundaries were flanked other either side by 10 kb and the amount of overlap was enumerated.

**Supplementary Figure 6. Predicted TAD boundaries were centered closer to peak signal regions defined by known architectural transcription factors when compared to called TAD boundaries for K562.** Enrichment
heatmaps illustrate the average signal strength of K562-specific ChIP-seq peak regions defined by (A) SMC3, (B) ZNF143, (C) CTCF, and (D) RAD21 around the centers of 10 kb flanked predicted and called TAD boundaries.

**Supplementary Figure 7. Known architectural transcription factors were statistically significantly closer to predicted TAD boundaries then called TAD boundaries.** Barplots illustrate the log2 distance in basepairs from classes of TAD boundaries to transcription factors including CTCF, RAD21, SMC3, and ZNF143 for the GM12878 (A) and K562 (B) cell lines. TAD boundaries were classified into the 3 classes. Predicted but not called (PBNC) were the set of 10 kb flanked predicted TAD boundaries the did not overlap with 10 kb flanked called TAD boundaries. Called and predicted (CAP) were the set of 10 kb flanked called and predicted TAD boundaries that overlapped with each other. Called but not predicted (CBNP) were the set of 10 kb flanked called TAD boundaries that did not overlap with 10 kb flanked predicted TAD boundaries.
-->

# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
