---
title: "preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution"
csl: styles.ref/genomebiology.csl
output:
  pdf_document: 
    extra_dependencies: ["float"]
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
bibliography: References/references.bib
--- 

```{r echo=FALSE}
knitr::opts_chunk$set(fig.pos = "!H", out.extra = "")
```


Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# Abstract

High-throughput chromosome conformation capture technology (Hi-C) revealed extensive DNA looping and folding into discrete 3D domains. These include Topologically Associating Domains (TADs) and chromatin loops, the 3D domains critical for cellular processes like gene regulation and cell differentiation. The relatively low resolution of Hi-C data (regions of several kilobases in size) prevents precise mapping of domain boundaries. However, the high resolution of genomic annotations associated with boundaries, such as CTCF and members of cohesin complex, suggests they can inform the precise location of domain boundaries. Several methods attempted to leverage genome annotation data for predicting domain boundaries; however, they overlooked key characteristics of the data, such as spatial associations between an annotation and a boundary, and a much smaller number of boundaries than the rest of the genome (class imbalance). 

We developed *preciseTAD*, an optimized random forest model to improve the location of domain boundaries. Trained on high-resolution genome annotation data and boundaries from low-resolution Hi-C data, the model predicts the location of boundaries at base-level resolution. We investigated several feature engineering and resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) to select the most optimal data characteristics and address class imbalance. Density-based clustering (DBSCAN) and partitioning around medoids (PAM) techniques were used to identify the precise location of boundary regions and summit points. We benchmarked our method against the Arrowhead domain caller and a novel chromatin loop prediction algorithm, Peakachu, on the two most annotated cell lines. 

We found that spatial relationship (distance in the linear genome) between boundaries and annotations has the most predictive power. Random under-sampling significantly improved model performance. Transcription factor binding sites outperformed other genome annotation types. Boundaries predicted by *preciseTAD* were more enriched for CTCF, RAD21, SMC3, and ZNF143 signal and more conserved across cell lines, highlighting their higher biological significance. The model pre-trained in one cell line performs well in another cell line. Using cell line-specific genomic annotations, the pre-trained models enable the detection of domain boundaries in cells without Hi-C data.

Our study implements the method and the pre-trained models for precise domain boundary prediction using genome annotation data. The precise identification of domain boundaries will improve our understanding of how genomic regulators are shaping the 3D structure of the genome. <!-- *preciseTAD* R package is available on https://dozmorovlab.github.io/preciseTAD/ and Bioconductor (submitted). -->

# Introduction

The advent of chromosome conformation capture (3C) sequencing technologies, and its successor Hi-C, have revealed a hierarchy of the 3-dimensional (3D) structure of the human genome such as chromatin loops [@rao20143d], Topologically Associating Domains (TADs) [@nora2012spatial;@dixon2012topological], and A/B compartments [@lieberman2009comprehensive]. At the kilobase scale, chromatin loops connect gene promoters with distal enhancers, thereby regulating gene expression [@Lupianez:2015aa; @Franke:2016aa]. At the megabase scale, TADs represent regions on the linear genome, that are highly self-interacting. <!--Emerging evidence has linked TADs to critical roles in cell dynamics and cell differentiation. Studies have shown that TADs themselves are highly conserved across species and cell lines [@dixon2012topological; @krefting2018evolutionary; @nora2012spatial; @dixon2015chromatin; @rao20143d; @Schmitt2016aa].--> Disruption of boundaries demarcating TADs and loops promotes cancer [@hnisz2016activation; @taberlay2016three] and other disorders [@lupianez2016breaking;@Sun:2018aa;@Meaburn:2007aa]. Identifying the precise location of TAD and chromatin loop boundaries remains a top priority in our goal to fully understand the functionality of the human genome.

Several methods have been proposed to identify genomic coordinates of TADs (reviewed in [@Zufferey:2018aa; @forcato2017comparison]), and chromatin loops [@rao20143d; @Ay:2014aa; @Heinz:2010aa; @salameh2019supervised]. However, a key limitation of them is that they depend on Hi-C data *resolution*. Resolution refers to the size of genomic regions (bins) used to segment the linear genome [@li2019comparison]. Lower resolutions contribute to larger bin sizes, which leads to increased uncertainty in the location of boundaries. Due to high sequencing costs, most available Hi-C datasets have relatively low resolution, ranging from 25 kb to 100 kb. While these low-resolution Hi-C datasets can be used to probe higher-order genomic patterns such as A/B compartments, more refined resolutions are necessary to accurately define TAD and loop boundaries [@fudenberg2016formation]. Because increasing the resolution of Hi-C data requires a quadratic increase in the total sequencing depth [@schmitt2016genome], obtaining high resolution remains difficult. 

Another limitation among domain callers is that they disregard prior knowledge about functional genomic annotations associated with TAD and loop boundaries, referred hereafter as domain boundaries. The insulator binding protein, CTCF, and additional cofactors such as SMC3 and RAD21 have been identified as components of the loop extrusion model<!--, whereby DNA is extruded through cohesin rings forming chromatin loops--> [@hansen2018recent;@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self; @Mirny:2019aa; @Davidson:2019aa]. Furthermore, distinct patterns of histone modifications have also been shown to be present at boundaries<!--including H3K36me3--> [@lieberman2009comprehensive;@dixon2012topological;@libbrecht2015joint]. These genomic annotations are obtained using chromatin immunoprecipitation followed by high-throughput DNA sequencing (ChIP-seq). The resolution of ChIP-seq experiments is typically on the order of tens to hundreds of bases [@mendoza2009genome], well below the resolution of Hi-C data (tens of kilobases; 750 bp is the highest resolution of Hi-C data to date [@bonev2017multiscale]). Therefore, leveraging precisely mapped genomic annotations in a supervised machine learning framework enables more precise prediction of the locations of domain boundaries. 

We present *preciseTAD*, an optimally tuned machine learning framework for precise identification of domain boundaries using genome annotation data. Our method utilizes the random forest (RF) algorithm trained on high-resolution chromatin state (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS) data to predict low-resolution boundaries. We introduce a systematic pipeline for building the optimal boundary prediction classifier. We found that spatial associations (linear distance) between boundaries and annotations perform best, transcription factor binding sites improve prediction accuracy, and a simple random undersampling technique effectively addresses the negative effect of class imbalance. Translated from Hi-C data resolution level to base level (annotating each base and predicting its boundary probability), *preciseTAD* employs density-based clustering (DBSCAN) and partitioning around medoids (PAM) to detect genome annotation-guided boundary regions and points at a base-level resolution. This approach circumvents resolution restrictions of Hi-C data, allowing for the precise detection of biologically meaningful boundaries. We demonstrate that *preciseTAD* predictions are more enriched for known molecular drivers of 3D chromatin, including CTCF, RAD21, SMC3, and ZNF143. Further, we show that *preciseTAD*-detected boundaries are more conserved across cell lines, and the model trained in one cell line can predict boundaries in another cell line. This improved precision in the domain boundary location can provide insight into the association between genomic regulators and the 3D genome organization. The *preciseTAD* R package is freely available on https://dozmorovlab.github.io/preciseTAD/ and Bioconductor (submitted).

# Methods

## Data sources

TAD and loop boundaries called by Arrowhead [@durand2016juicebox] and Peakachu [@salameh2019supervised] tools were used for training. The autosomal genomic coordinates in the GRCh37/hg19 human genome assembly were considered. Arrowhead-defined TAD boundaries were called from the GM12878 and K562 cell line Hi-C data (MAPQ>0, 5 kb, 10 kb, 25 kb, 50 kb, and 100 kb resolutions) using the default parameters (Additional File 1). Peakachu chromatin loop boundaries called at 10 kb for the GM12878 and K562 cell lines were downloaded from the Yue lab website (Supplementary Table 1). Unique boundaries were considered as the midpoints within the coordinate of each chromatin loop anchor. Chromosome 9 was excluded from all downstream analyses due to the sparsity of contact matrices at 5 kb and 10 kb resolutions for the K562 cell line. Cell-line-specific genomic annotations (BroadHMM chromatin states (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS)) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 2). 

## Shifted-binning for binary classification

In Hi-C, each chromosome is binned into non-overlapping regions of length *r*, typically, 5 kb and above. The *r* parameter defines the resolution of Hi-C data. Here, we designed a strategy called _shifted binning_ that partitions the genome into regions of the same length *r*, but with middle points corresponding to boundaries defined by the original binning. 

To create shifted binning, the first shifted bin was set to start at half of the resolution *r* and continued in intervals of length *r* until the end of the chromosome (*mod r* + *r/2*). The shifted bins, referred hereafter as bins for simplicity, were then defined as boundary-containing regions (*Y = 1*) if they contained a TAD (or loop) boundary, and non-boundary regions (*Y = 0*) otherwise, thus establishing the binary response vector (**Y**) used for classification (Figure 1A).

## Feature engineering

Cell line-specific genomic annotations were used to build the predictor space. Bins were annotated by one of either the average signal strength of the corresponding annotation (*Peak Signal Strength*), the number of overlaps with an annotation (*Overlap Count (OC)*), the percent of overlap between the bin and the total width of genomic annotation regions overlapping it (*Overlap Percent (OP)*), or the distance in bases from the center of the bin to the center of the nearest genomic annotation region (*Distance*) (Figure 1B). A ($log_2 + 1$)-transformation of distance was used to account for the skewness of the distance distributions. Models built using a *Peak Signal Strength* predictor space were only composed of histone modifications and transcription factor binding sites because BroadHMM chromatin states lack signal values (Supplementary Figure 1).

## Addressing class imbalance

To assess the impact of class imbalance (CI), defined as the proportion of boundary regions to non-boundary regions, we evaluated three resampling techniques: *Random Over-Sampling (ROS)*, *Random Under-Sampling (RUS)*, and *Synthetic Minority Over-Sampling Technique (SMOTE)*. For ROS, the minority class was sampled with replacement to obtain the same number of data points in the majority class. For RUS, the majority class was sampled without replacement to obtain the same number of data points in the minority class. For SMOTE, under-sampling was performed without replacement from the majority class, while over-sampling was performed by creating new synthetic observations using the $k=5$ minority class nearest neighbors [@chawla2002smote] (implemented in the `DMwR` v.0.4.1 R package). We restricted the SMOTE algorithm to 100% over-sampling and 200% under-sampling, to create perfectly balanced classes. 

## Establishing optimal data level characteristics for TAD boundary region prediction

Random forest (RF) classification models (the `caret` v.6.0 R package [@kuhn2012caret]) were built to compare model performances between combinations of data resolutions, feature engineering procedures, and resampling techniques. Following recommendations to evaluate the model on unseen data [@schreiber2019pitfall], a *holdout chromosome* technique was used for estimating model's performance. The $i^{th}$ holdout chromosome was identified and a data matrix, $A_{N \times (p+1)}$, was constructed by combining the binned genome from the remaining chromosomes ($1,2,\cdots, i-1,i+1,\cdots,21,22$), where $N=[n_{1} n_{2} \cdots n_{21} n_{22}]'$ and $n_{k}$ is the length of chromosome *k* after being binned into non-overlapping regions of resolution *r*, such that $k \ne i$. $p$, the number of annotations, together with the response vector, **Y**, defined the column-wise dimension of the matrix $A$. Re-sampling was then performed on $A$, and an RF classifier was trained using 3-fold cross-validation to tune for the number of annotations to consider at each node (*mtry*). The number of trees (*ntree*) that were aggregated for each RF model was set to 500. The minimum number of observations per root node (*nodesize*) was set to 0.1% of the rows in the data. The binned data for the holdout chromosome $i$ was reserved for testing. Models were evaluated using Balanced Accuracy (BA), defined as the average of sensitivity and specificity:

$$BA = \dfrac{1}{2}\left( sensitivity + specificity \right) = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$

where TP refers to the number of bins correctly identified as containing a boundary (true positives), FP refers to the number of bins incorrectly identified as containing a boundary (false positives), TN refers to the number of bins correctly identified as not containing a boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a boundary (false negatives). Each of these quantities is obtained from the confusion matrix created by validating the model on the test data. The process was repeated for each $i^{th}$ holdout chromosome, and performances were aggregated using mean and standard deviation.

## Feature selection and predictive importance

Many genomic annotations, notably architectural proteins, tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To suitably reduce the predictor space and improve computational efficiency, while maintaining optimal performance, we implemented recursive feature elimination (RFE). We estimated the near-optimal number of necessary features, ranging from 2 to the maximum number of features incremented by the power of 2. We then aggregated the predictive importance of the union of the optimal set of features across holdout chromosomes using the mean decrease in node impurity among permuted features in out-of-bag samples to determine the most common and top-ranked annotations for predicting boundary regions. 

## Evaluating performance across cell lines

When evaluating performance across cell lines, we compared the strategy of training and testing binned and annotation data from the same cell line with training and testing across cell lines. To do so, we built a model using binned and annotation data from $n-1$ chromosomes of the other cell line, while maintaining same testing data from the same cell line strategy (i.e. data from the $i^{th}$ holdout chromosome). As before, this process was repeated for each holdout chromosome. To evaluate performance, we constructed receiver operating characteristic (ROC) curves composed of the average sensitivities and specificities at different cutoffs, across each holdout chromosome, and reported the corresponding average area under the curve (AUC).

## Boundary prediction at the base-level resolution using *preciseTAD*

To investigate whether we could alleviate the resolution limitations of conventional domain calling tools, we developed *preciseTAD*. This algorithm leverages a random forest model in conjunction with density-based and partitioning techniques to predict boundaries at base resolution (Supplementary Figure 2). 

A random forest model $(M)$ was built on the optimal combination of predictor type (predictor space), resampling technique (class balancing correction), and top-ranked annotations (most informative genomic elements) for a set of binned chromosomes $\{k | k \ne i \}$ <!--, to maximally accurately predict boundaries at the _Hi-C resolution level_-->. To precisely identify boundary locations, we first constructed a _base-level resolution_ predictor space for the chromosome *i*, $A_{n \times p}$, where $n$ is the length of chromosome $i$ and $p$ is the optimal number of annotations. We then evaluate $M$ on the base-level predictor space to extract the probability vector, $\pi_{n_{i}}$, denoting each base's probability of being a boundary. Here, $n_{i}$ represents the length of chromosome *i*. A threshold $t$ specifies the probability at which a base with $\pi_{n_{i}} \ge t$ is designated as a potential boundary (the default $t=1$). Next, density-based spatial clustering of applications with noise (DBSCAN; version 1.1-5) is applied to the matrix of pairwise genomic distances between boundary-annotated bases, $D$. The minimum and maximum coordinates of each cluster, $k$, of spatially colocalized bases were termed *preciseTAD boundary regions* (PTBR). To precisely identify a single base among each PTBR, *preciseTAD* implements partitioning around medoids (PAM) on the distance matrix, $D_{k}$, derived from each cluster. The corresponding cluster medoid was defined as a *preciseTAD boundary point* (PTBP), making it the most representative base coordinate within each clustered PTBR.

The DBSCAN algorithm has two parameters, *MinPts* and *eps* ($\epsilon$). The *MinPts* parameter was set to the recommended value of 3, representing $1+dim(data)$ [@hahsler2019dbscan]. To decide on the optimal value of $t$ and $\epsilon$ in *preciseTAD*, we considered the normalized enrichment $NE$ of flanked predicted boundaries, defined as

$$NE = \dfrac{1}{p} \left[ \Sigma_{s=1}^{p} \left[ \dfrac{1}{b} \Sigma_{k=1}^{b} e_{ks} \times c_{k} \right] \right]$$

where $e_{ks} = \textbf{I}\{ r_{s} \in (b_{k}-f, b_{k}+f)\}$ is the number of elemental regions $r$ of predictor $p$ that overlap with each flanked boundary. We evaluated $NE$ for combinations of $t=\{0.975, 0.99, 1.0 \}$ and $\epsilon = \{1000, 5000, 10000, 15000, 20000, 25000 \}$. The heuristic of $\epsilon$ is that density-reachable bases with genomic distances less than $\epsilon$ should occupy the same designated cluster. The default combination was set to $t=1.0$ and $\epsilon=10000$ based on our tests (Supplementary Figure 11).

## Predicting boundaries across cell lines

We aimed to extend the strategy of building and comparing boundary region prediction models using same cell line versus across cell line data to predict boundaries at base level resolution across cell lines. To do so, we applied *preciseTAD* on models built using cell line specific annotation data to predict boundaries on base level annotation data from another cell line. As before, this strategy was compared to applying *preciseTAD* on models and base level annotation data from the same cell line.

## Evaluating called and predicted boundary precision

We assessed the biological significance of our predicted boundaries by their association with the signal of CTCF, RAD21, SMC3, and ZNF143 using *deepTools* (version 2.0) [@ramirez2016deeptools] (*computeMatrix*, *plotProfile*, and *plotHeatmap* tools). Additionally, we compared the median $log_2$ genomic distances between TAD boundaries and the same top predictive ChIP-seq annotations using Wilcoxon Rank-Sum tests. Furthermore, we compared the overlap between predicted and called boundaries in GM12878 and K562 cell lines. Boundaries were first flanked by resolution, *r*, and overlaps were visualized using Venn diagrams from the `Vennerable` R package (version 3.1.0). Overlaps were further quantified using the Jaccard index defined as

$$J_{(A,B)} = \dfrac{A \cap B}{A \cup B}$$

where A and B represent genomic regions created by flanked called and predicted boundaries. That is, between cell lines, the number of overlapping flanked boundaries divided by the total number of flanked boundaries. Wilcoxon Rank-Sum tests were used to compare chromosome-specific Jaccard indices across cell lines, between *preciseTAD* boundaries and both Arrowhead and Peakachu boundaries, respectively. All statistical analyses were performed in R version 4.0.1. The significance level was set to 0.05 for all statistical tests.

<!--using the `EnrichedHeatmap` R package (version 1.14.0) [@gu2018enrichedheatmap]. For each boundary (both called and predicted), a 5000 bp flank was created on either side. The flanking was broken up into 50 bp segments, for a total of 100 windows on both sides of a given boundary. A coverage value was generated for each window with respect to a given ChIP-seq annotation, as the mean signal averaged by the width of the window, denoted by,

$$v_{c} = \dfrac{\Sigma_{j}^{m} \Sigma_{i}^{n}x_{ij}w_{ij}}{L}$$

where $x_{ij}$ is the total number of bases of annotation $i$ in window $j$ and $w_{ij}$ is the number of bases that overlap between annotation $i$ and window $j$. The denominator, $L$, is the width of the windows (here, $L=50$). The rows of the heatmaps are ordered by enriched scores. The enriched scores are calculated as the sum of coverage values weighted by the distance to the flanked boundary, denoted by

$$s_{e} = \Sigma_{d=1}^{n_{1}}\left(\dfrac{x_{1d}* d}{n_{1}} \right) + \Sigma_{u=1}^{n_{2}}\left(\dfrac{x_{2u}* (n_{2} - u + 1)}{n_{2}} \right)$$
where $n_{1}$ and $n_{2}$ represent the number of downstream and upstream coverage values to sum over. Here, $n_{1}=n_{2}=100$.-->

# Results

## Developing an ML framework for optimal TAD boundary prediction

We developed a machine learning (ML) framework for determining the optimal set of data level characteristics to predict boundaries of Topologically Associating Domains (TADs) and chromatin loops, collectively referred to as domain boundaries. We chose the random forest (RF) algorithm as our binary classification tool. The reason for it is two-fold: (1) to devise a tunable prediction rule in a supervised learning framework that is both robust to overfitting and able to handle multiple correlated predictors, and (2) to allow for an interpretable ranking of predictors [@boulesteix2012overview]. We used Arrowhead-called TAD boundaries [@durand2016juicer] and published Peakachu-predicted loop boundaries [@salameh2019supervised] as ground truth. Data from GM12878 and K565 cell lines at 5-100kb resolution (Arrowhead) and 10kb resolution (Peakachu) were used (Supplementary Table 1). We next developed *preciseTAD*, an R package that leverages an optimally built RF to precisely predict genomic coordinates of TAD and loop boundaries at base resolution from cell line-specific genome annotation data.

![**Resolution-specific data construction and feature engineering for random forest modeling.** (A) The linear genome was binned into non-overlapping resolution-specific intervals using *shifted binning* (see Methods). The response vector **Y** was defined as 1/0 if a genomic bin overlapped/did not overlap with a TAD (or loop) boundary. (B) Four types of associations between bins (blue dashed lines) and genomic annotations (green shapes) were considered to build the predictor space including Average Peak Signal (Signal), Overlap Counts (OC), Overlap Percent (OP), and $log_{2}$ distance (Distance).](Figures/figure1.png)

Boundary regions were defined as genomic bins containing a called boundary ($Y = 1$), while non-boundary regions were defined as bins that did not contain a called boundary ($Y = 0$) (Figure 1A, see Methods). The total number of called TADs, their unique boundaries, and the number of genomic bins expectedly decreased with the decreased resolution of Hi-C data (Table 1, Supplementary Table 3). The number of non-boundary regions highly outnumbered boundary regions. Such a disproportional presence of examples in one class is known as a "class imbalance" problem that negatively affects predictive modeling [@prati2011survey; @wei2013role]. To address the class imbalance, we evaluated the effect of three resampling techniques. *Random over-sampling* (ROS) was defined as sampling with replacement from the minority class (boundary regions). *Random under-sampling* (RUS) was defined as sampling with replacement from the majority class (non-boundary regions). Lastly, we tested *Synthetic minority over-sampling technique* (SMOTE), which is a combination of both random over- and under-sampling to create balanced classes [@chawla2002smote] (see Methods). 


|          Tool         |     Resolution    |     Total number of called TADs/chromatin loops    |     Total number of unique TAD/chromatin loop boundaries    |     Total number of genomic bins    |     Class imbalance    |
|:---------------------:|:-----------------:|:--------------------------------------------------:|:-----------------------------------------------------------:|:-----------------------------------:|:----------------------:|
|        Arrowhead      |       5   kb      |                         8,052                       |                             15,468                           |                535,363               |           0.03         |
|     Arrowhead         |       10   kb     |                         7,676                       |                             14,253                           |                267,682               |           0.05         |
|     Arrowhead         |       25   kb     |                         4,670                       |                             8,363                            |                107,073               |           0.08         |
|     Arrowhead         |       50   kb     |                         2,349                       |                             4,224                            |                 53,537               |           0.08         |
|     Arrowhead         |      100   kb     |                         1,031                       |                             1,883                            |                 26,768               |           0.07         |
|        Peakachu       |       10   kb     |                        16,185                       |                          21,421                              |                267,682               |        0.14            |
Table: Domain boundary data and class imbalance summaries across resolutions for Arrowhead and Peakachu on GM12878.

![**A machine learning framework for building domain boundary region prediction models.** Step 1 employs a range of feature engineering techniques to define the predictor matrix $A_{N \times (p)}$, where $N$ is the number of genomic bins, $p$ is the number of genomic annotations, $i$ is a holdout chromosome. The response vector $Y_N$ is defined as a boundary region ($Y=1$) if it overlaps with a genomic bin (else $Y=0$). Step 2 reserves the predictor-response matrix for the holdout chromosome $i$ as the test data. Step 3 applies a resampling technique to the training data to address the class imbalance. Step 4 trains the random forest model and performs 3-fold cross-validation to tune the mtry parameter. Finally, step 5 validates the model on the separate test data composed of the binned data from the holdout chromosome $i$ and to evaluate model performance using balanced accuracy (BA).](Figures/figure2.png) 

A total of 77 cell line-specific genomic annotations were used to build the predictor space. These included BroadHMM chromatin state data, histone modifications (HM), and transcription factor binding sites (TFBS) (Supplementary Table 2). Four feature engineering procedures were developed to quantify the association between genomic annotations and bins (Figure 1B). These included signal strength association (Signal), direct (OC), proportional (OP), and spatial ($log_2 + 1$ Distance) relationships (see Methods, Supplementary Figure 1). 

In total, we considered combinations of data from two cell lines *L = {GM12878, K562}*, five resolution *R = {5 kb, 10 kb, 25 kb, 50 kb, 100 kb}*, four types of predictor space *P = {Signal, OC, OP, Distance}*, and three re-sampling technique *S = {None, RUS, ROS, SMOTE}* (Figure 2). Once the model inputs were established, a random forest classifier was trained on $n-1$ autosomal chromosomes, while reserving the $i^{th}$ chromosome for testing. Three-fold cross-validation was used to tune the *mtry* hyperparameter, while *ntree* and *nodesize* were fixed at 500 and at 0.1% of the rows in the training data, respectively. Model performance was evaluated by aggregating the mean balanced accuracy (BA) across each holdout chromosome (see Methods). These strategies allowed us to select the best performing model in an unbiased manner.


## Random under-sampling, distance-based predictors, and high-resolution Hi-C data provide optimal performance for boundary prediction

Expectedly, when using data with class imbalance present, that is, no resampling, the models exhibited low balanced accuracies, with minimal variability among different resolutions (Figure 3). Similarly, poor performances were found when using ROS. However, RUS and SMOTE re-sampling led to a drastic improvement in performance, especially at higher resolutions. We found that RUS marginally outperformed SMOTE as the optimal class balancing technique for all resolutions and predictor types when predicting TAD boundary regions. 

Additionally, we found that using a distance-type predictor space yielded substantially higher balanced accuracy compared to the peak signal strength, overlap count, and overlap percent predictor types. As with class balancing techniques, this improvement was less evident at lower resolutions, with results consistent for K562 (Supplementary Figure 3A). Random forest models built on Peakachu-defined loop boundaries with SMOTE resampling only marginally outperformed RUS for both GM12878 and K562 cell lines (Supplementary Figure 3B,3C). Furthermore, 5 kb resolution genomic bins led to the optimal prediction for both TAD and loop boundaries, on both cell lines. Our results indicate that random under-sampling, distance-type predictors, and high-resolution Hi-C data provide the optimal set of data level characteristics for both TAD and chromatin loop boundary prediction.


![**Determining optimal data level characteristics for building TAD boundary region prediction models on GM12878.** Averaged balanced accuracies are compared across resolution, within each predictor-type: Signal, OC, OP, and Distance, and across resampling techniques: no resampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Error bars indicate standard deviation from the mean performance across each holdout chromosome used for testing.](Figures/figure3.png)


## Transcription factor binding sites outperform histone- and chromatin state-specific models

We hypothesized that the type of genomic annotations may also affect predictive performance. Using the established optimal settings (RUS, Distance, 5 kb bins), we built separate random forest models using BroadHMM chromatin state data, histone modifications (HM), and transcription factor binding sites (TFBS). We found that models built on TFBSs outperformed other annotation-specific models, with results consistent for loop boundaries, on both cell lines (Figure 4A; Supplementary Figures 4A). These results suggest that TFBS are the primary drivers of TAD and loop boundary formation in both GM12878 and K562. 

![**SMC3, RAD21, CTCF, and ZNF143 transcription factors accurately predict TAD and loop boundaries in GM12878.** (A) Barplots comparing performances of TAD (Arrowhead) and loop (Peakachu) boundary prediction models using histone modifications (HM), chromatin states (BroadHMM), transcription factor binding sites (TFBS), in addition to a model containing all three classes (ALL). (B) Recursive feature elimination (RFE) analysis used to select the optimal number of predictors. Error bars represent 1 standard deviation from the mean cross-validated accuracy across each holdout chromosome. (C) Clustered heatmap of the predictive importance for the union of the top 8 most predictive chromosome-specific TFBSs. The columns represent the holdout chromosome excluded from the training data. (D) Receiver operating characteristic (ROC) curves and the corresponding average area under the curves (AUCs) when training and testing on GM12878 data (blue, Arrowhead ground truth; red, Peakachu ground truth) versus training on K562 and testing on GM12878 data (black, dashed). The curves represent the average sensitivities and specificities across each holdout chromosome. The shaded areas around each curve represent 1 standard deviation from the average.](Figures/figure4.png)

## Predictive importances confirmed the biological role of CTCF, RAD21, SMC3, and ZNF143 for boundary formation

We sought to further optimize our boundary region prediction models by recursive feature elimination to avoid overfitting and selecting only the most influential TFBS across all autosomal chromosomes. We were able to obtain near-optimal performance using approximately eight TFBS (Figure 4B; Supplementary Figures 4B). However, given that we trained our model on chromosome-specific data, the most significant annotations varied for each chromosome set. To determine transcription factors most important for boundary prediction across all chromosomes, we clustered the predictive importance (mean decrease in accuracy) of the top eight significant TFs across chromosomes. We found four transcription factors, CTCF, RAD21, SMC3, and ZNF143, being consistently predictive of TAD and loop boundaries (Figure 4C; Supplementary Figures 4C). We optimized our model by only considering these top four TFBS when building the random forest model, thereby decreasing computational burden while maintaining high predictive performance. In summary, our model was able to yield the known molecular drivers of the loop extrusion model [@hansen2018recent;@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self; @Mirny:2019aa; @Davidson:2019aa]. 

## Training in one cell line accurately predicts boundary regions in other cell lines 

Having demonstrated the optimal performance in our framework's ability to predict boundary regions, we next attempted to predict boundary regions in one cell line using the model pre-trained on data from another another cell line (See Methods). We found that training and testing using Arrowhead ground truth TAD boundaries and genomic annotation data from GM12878 cell line resulted in an average AUC=0.792 (Figure 4D). Surprisingly, when training on the K562 cell line and testing on GM12878, the average AUC increased slightly to 0.795. Likewise, the average performance of models trained using Peakachu boundaries and genomic annotation data from GM12878 cell line was comparable to models trained on K562-specific Peakachu boundaries and genomic annotations (Avg. AUC=0.881 and 0.874, respectively). These results were consistent when comparing training/testing strategies on K562 with training on GM12878 and testing on K562 data (Supplementary Figure 4D). The average ROC curves were found to be well within 1 standard deviation of each other, suggesting that a model trained on data from one cell line performs well when using the data from another cell line, allowing for the opportunity to predict boundaries for cell lines that do not currently have Hi-C data available.

## *preciseTAD* identifies precise and biologically relevant domain boundaries

Using our optimally built random forest model trained on Arrowhead/Peakachu boundaries, we attempted to predict boundaries at base-level resolution. Intuitively, instead of bin-level annotations, the predictor-response space was built on a base-level, that is, each base was associated with a genomic annotation. The model trained on a bin-level space was then applied on a base-level space to predict each base's probability of being a boundary. Our method, referred to as *preciseTAD*, uses density-based spatial clustering (DBSCAN) and partitioning around medoids (PAM) to predict the location of boundary regions and summit points (highest boundary probability) at base-level resolution (see Methods; Supplementary Figure 2).

When trained using Arrowhead and Peakachu ground truth boundaries at 5 kb and 10 kb, respectively, the *preciseTAD* model predicted a total of 12,258 TAD and 15,707 chromatin loop boundaries in GM12878, as well as 9,603 TAD and 11,154 chromatin loop boundaries in K562 cell line (Supplementary Table 4). We reported less predicted TAD boundaries at 5 kb than Arrowhead on both cell lines (Table 1, Supplementary Table 3). This can be attributed to Arrowhead's inflation of called TADs, that, when visualized, often do not correspond to domains enriched in internal interactions (Figure 5A) and signal of known drivers of domain boundaries (Figure 5B). *preciseTAD* also predicted fewer chromatin loop boundaries than Peakachu (Table 1, Supplementary Table 3). This can be attributed to Peakachu's use of only CTCF sites to call boundaries, while *preciseTAD* leverages four known drivers of 3D chromatin, including CTCF, RAD21, SMC3, and ZNF143. These results suggest that genome annotation-driven predicted boundaries may be more biologically relevant than those detected from Hi-C data alone.

![**preciseTAD boundaries are enriched for known molecular drivers of 3D chromatin.** (A) The location of Arrowhead-called TAD boundaries (blue) vs. *preciseTAD*-predicted TAD boundaries (green) on GM12878 data (chr14:50085000-50800000). The black line represents the predicted probability of each base being a TAD boundary. (B) A zoomed-in portion of the genome shows the preciseTAD boundary region (PTBR, highlighted yellow), a cluster of bases with high probability of being a boundary, and the corresponding signal profiles of CTCF, RAD21, SMC3, and ZNF143. (C, D) Signal profile plots comparing the strength of CTCF, RAD21, SMC3, and ZNF143 binding around Arrowhead-called boundaries (blue, C), Peakachu loop boundaries (red, D) vs. *preciseTAD*-predicted boundaries (green).](Figures/figure5.png) 

Next, we evaluated the biological significance of *preciseTAD* boundary points. Signal of four known molecular drivers of 3D chromatin (CTCF, RAD21, SMC3, and ZNF143) colocalized more frequently around *preciseTAD*-predicted boundary points, as compared to Arrowhead-called TAD and Peakachu loop boundaries, respectively (Figure 5C, 5D; Supplementary Figure 5A, 5B). Surprised by the poor signal distribution around Arrowhead boundaries, we compared signals centered on boundaries called by Arrowhead, Peakachu, and our SpectralTAD [@Cresswell:2020ab] TAD caller. We confirmed the poor signal distribution around Arrowhead boundaries, in contrast to the relatively well-performing Peakachu- and SpectralTAD-called boundaries (Supplementary Figure 5C, D). Signal enrichment heatmaps confirmed that *preciseTAD*-predicted boundaries were more enriched for the same genomic annotations than either Arrowhead or Peakachu boundaries alone (Supplementary Figures 6-7). *preciseTAD* boundaries were statistically significantly closer to the top-ranked TFBSs (Wilcoxon p-value < 0.001 versus Arrowhead and Peakachu boundaries, respectively, Figure 6A, Supplementary Figure 8). These results indicate that *preciseTAD*-predicted boundaries better reflected the known biology of boundary formation.

## *preciseTAD* boundaries are more conserved across cell lines

Previous studies suggest that TAD boundaries are conserved across cell lines [@dixon2012topological; @krefting2018evolutionary; @nora2012spatial; @sexton2012three]. To assess the level of cross-cell-line conservation, we evaluated the overlap between cell line-specific boundaries detected by *preciseTAD*, Arrowhead, and Peakachu. Only 26% and 49% of boundaries were conserved between cell lines for Arrowhead and Peakachu boundaries (J=0.186 and J=0.388), respectively (Figure 6D, 6D. However, 45%/56% of *preciseTAD*-predicted domain boundaries were conserved between GM12878 and K562 cell lines when using models trained on Arrowhead/Peakachu data (J=0.383 and J=0.444, respectively, Figure 6C, 6E). The better conservation of *preciseTAD*-predicted boundaries further supports the notion of their higher biological relevance.

![**preciseTAD-predicted boundaries are closer to CTCF sites and more conserved across cell lines**. (A) Log2 genomic distance distribution from called and predicted boundaries to nearest CTCF sites. The p-values are from the Wilcoxon Rank Sum test. (B-E) Venn diagrams illustrating the levels of conservation (overlap) between domain boundaries for GM12878 (red) and K562 (blue) cell lines identified by Arrowhead (B), Peakachu (D), *preciseTAD*-predicted boundaries using Arrowhead- and Peakachu-trained models () Arrowhead-trained *preciseTAD* models (C), Peakachu boundaries themselves (C, E, respectively). Boundaries involving Arrowhead/Peakachu were flanked by 5 kb/10 kb, respectively.](Figures/figure6.png) 

## Boundaries predicted by *preciseTAD* models trained on TAD and loop boundaries are highly overlapping

The majority of boundaries predicted by the Arrowhead-trained *preciseTAD* model represented a subset of the larger group of boundaries predicted by the Peakachu-trained model (Supplementary Figure 9). <!--We found that 88.8% and 95.8% [??? These numbers are different from what is shown on Supp Fig 9; !!!the %'s shown are (left region)/(left + middle); should I take out the percentages in the figure? ??? Still unclear, calculations don't make sense. TBD] of our predicted TAD boundaries were overlapped by predicted loop boundaries for GM12878 and K562, respectively (Supplementary Figure 9). [??? After clarification, MD to rewrite the first two sentences, unclear] -->This is expected as loop boundaries detected by Peakachu are more abundant, while comparatively wide TAD boundaries detected by Arrowhead likely represent the higher level of the 3D chromatin organization. Together with the results in Figure 4C, the high overlap between boundaries predicted by Arrowhead- and Peakachu-trained models suggests that TAD and loop boundaries may be driven by similar molecular mechanisms. 

Guided by the high predictive performance when training and testing on different cell lines (Figure 4D), we opted to evaluate whether models trained using Arrowhead/Peakachu ground truth data in one cell line could be leveraged to predict boundaries using annotation data from another cell line. We evaluated two scenarios: 1) training on GM12878 and predicting boundaries on GM12878 (GM on GM) vs. training on K562 and predicting on GM12878 (K on GM), and 2) training on K562 and predicting boundaries on K562 (K on K) vs. training on GM12878 and predicting boundaries on K562 (GM on K). Using Arrowhead-trained models, 76% (J=0.701) and 81% (J=0.751) of predicted boundaries overlapped in both cross-cell-line prediction scenarios (Supplementary Figure 10A, 10B). Likewise, when using Peakachu-trained models, we observed 85% (J=0.705) and 88% (J=0.759) overlaps (Supplementary 10C, 10D). Furthermore, boundaries predicted on unseen annotation data exhibited similar level of enrichment for CTCF, RAD21, SMC3, and ZNF143 as did those trained and predicted on the same cell line (Supplementary Figure 10). These results indicate that *preciseTAD* pre-trained models can be successfully used to predict domain boundaries for cell lines lacking Hi-C data but for which genome annotation data is available.

# Discussion

We present *preciseTAD*, a machine learning approach for precise prediction of TAD or loop boundaries from functional genomic annotations. *preciseTAD* leverages a random forest (RF) classification model built on low-resolution domain boundaries obtained from domain calling tools, and high-resolution genomic annotations as the predictor space. *preciseTAD* predicts the probability of each base being a boundary, and identifies the precise location of boundary regions and the most likely boundary points. *preciseTAD* was benchmarked against two boundary calling tools, Arrowhead [@durand2016juicebox], an established TAD-caller, and Peakachu [@salameh2019supervised] a recently published algorithm for predicting chromatin loops. We first optimized our RF model by systematically comparing different combinations of genome binning (resolution), feature engineering procedures, and resampling techniques. Our results demonstrate that genomic distance between boundary regions and genomic annotations coupled with random under-sampling results in best model performance. We show that binding of four transcription factors (SMC3, RAD21, CTCF, ZNF143) is sufficient for accurate boundary predictions. *preciseTAD* is implemented as an R package, provides the pre-trained models, and allows for precise prediction of boundaries for any cell line for which the genomic annotations are available.

Our machine learning framework yielded several interesting observations. We first demonstrated that RF models built using *distance*-type predictors outperformed models built on previously published feature engineering techniques, including signal strength, overlap counts, and overlap percents [@mourad2016computational; @Moore:2015aa; @Al-Bkhetan:2018aa; @kai2018predicting]. We further demonstrated that class imbalance hinders boundary prediction, but can be effectively addressed by a simple random under-sampling technique, an aspect of boundary prediction unaddressed in previous studies [@mourad2016computational; @Moore:2015aa; @Al-Bkhetan:2018aa]. We showed that information about only four transcription factors (CTCF, SMC3, RAD21, ZNF143) is necessary and sufficient for accurate boundary prediction, outperforming histone modification- and BroadHMM-built models [@Moore:2015aa; @Al-Bkhetan:2018aa]. These are known components of the loop extrusion model, an established theory of how loops are made by a ring-shaped adenosine triphosphatase-driven complex called cohesin [@hansen2018recent; @sanborn2015chromatin; @fudenberg2016formation; @alipour2012self; @Mirny:2019aa; @Davidson:2019aa]. Interestingly, the same transcription factors accurately predicted both TAD and loop boundaries, suggesting a similarity of the mechanisms of TAD and loop formation. This suggested that the random forest model, when tuned and feature engineered correctly, is highly effective in predicting biologically relevant domain boundaries. 

Besides balanced accuracy (BA), we investigated five other performance metrics, including accuracy, area under the receiver operating characteristic curve (AUROC), precision, F1-score, and area under the precision-recall curve (AUPRC) (Supplementary Table 5). Our aim was to have a balanced metric sensitive to class imbalance such that it would not favor one component of the confusion matrix. The accuracy metric can be artificially inflated by true negatives (TN), the set of genomic bins correctly predicted as not containing a ground truth boundary. AUROC captures how a model generally performs across different thresholds. However, it doesn't place more emphasis on one class over the other, so it does not reflect the minority class well. Precision indicates the rate at which positive predictions are correct and can be artificially deflated by low proportions of true positives (TP), the set of genomic bins correctly predicted as containing a ground truth boundary. While F1-score is a composite metric, it can be susceptible to different values for precision and recall. Lastly, AUPRC is insensitive toward class imbalance, preventing us from investigating its effect, and also omits from its calculation TN values. All of these are important considerations to make when choosing a performance metric. For these reasons, we opted to report balanced accuracy (BA). The BA benefits from incorporating all components of the confusion matrix, while also being sensitive to class imbalance, a necessary characteristic when comparing performances to models built using no data resampling. 

Guided by both low-resolution Hi-C data and high-resolution genome annotation data, *preciseTAD* predicts base-level resolution boundaries, alleviating resolution limitations of Hi-C data. However, a natural question is how resolution (width) of boundary regions identified by *preciseTAD* (PTBRs) compares with that of Hi-C data. Our preliminary observations indicate that, under most optimal settings, the width of PTBRs parallels the resolution of Hi-C data. Furthermore, each PTBR is formed by several sub-regions with the probability of being a boundary defined by the threshold $t$ ($t=1$ in the current study). Yet, the preciseTAD boundary points (PTBPs, medoids identified within each PTBR) had the highest density of CTCF and other transcription factor binding signal (Figure 5). Our results are in line with the emergent view that domain boundaries are dynamic [@chang2020tads] and their well-defined location arise as a consequence of the population average in bulk Hi-C data [@fudenberg2016formation; @Flyamer:2017aa].

We show that, unlike Arrowhead, *preciseTAD* does not inflate the number of predicted boundaries, providing only the most biologically meaningful boundaries that demarcate regions of high inter-chromosomal interactions. *preciseTAD* boundaries predicted using models trained on either TAD or loop boundaries (Arrowhead and Peakachu data) were enriched for known architectural transcription factors including CTCF, RAD21, SMC3, and ZNF143, supporting recent observations that TADs and loops may be generated by similar mechanisms [@hansen2018recent]. Likewise, *preciseTAD* boundaries were more conserved between GM12878 and K562 cell lines, a known feature among the 3D architecture of the human genome, further highlighting their biological significance.

*preciseTAD* offers flexibility in controlling both the number of predicted boundaries and the distance between them. The two primary parameters are the probability threshold $t$ and $\epsilon$ (referred to as eps; parameter of DBSCAN). The combination of these two quantities changes the resulting number of predicted boundaries from *preciseTAD*. Lower values of $t$ and $\epsilon$ will result in more clusters of bases, and therefore, more boundaries. As a heuristic, we evaluated the pairwise combination of 3 different thresholds ($t=(0.975,0.99,1.0)$) and 6 different $\epsilon$ values ($\epsilon=(1000,5000,10000,15000,20000,25000)$)). We found that the normalized overlaps - calculated as the total number of ChIP-seq peaks that overlapped within a given flanked boundary, divided by the number of boundaries - between top TFBS sites and flanked *preciseTAD* boundaries converged for combinations of $t=1.0$ and $\epsilon=10000$ (Supplementary Figure 11).

A limitation of *preciseTAD* is that it is dependent on the "ground truth" boundaries provided by a domain caller. Given the wide variety of domain callers and their variable performance [@Zufferey:2018aa; @forcato2017comparison], defining "ground truth" boundaries is challenging. <!--Ideally, we would benchmark *preciseTAD* against simulated boundaries. While methods for simulating Hi-C data sets with boundary annotations exist [@forcato2017comparison; @zheng2020freehi], methods for simulating the associated genomic annotations (the main component guiding *preciseTAD* predictions) are lacking.-->Ultra-deep Hi-C sequencing [@bonev2017multiscale] and newer technologies for precise mapping of chromatin interactions (e.g., Micro-C [@Krietenstein:2020aa]), coupled with more precise technologies for genomic annotation profiling (e.g., CUT&RUN for precise mapping of transcription factor binding sites) will help to refine the location and the genomic signatures of the "ground truth" boundaries. In the current work, we feel that the total number of domain boundaries is sufficient to guide learning of the association between genomic annotations and boundaries for precise boundary predictions. Indeed, models trained on the larger number of Peakachu-predicted boundaries performed better than those trained on Arrowhead boundaries. Although we provide models trained on both boundary types, we recommend Peakachu-trained models for the base-level prediction of domain boundaries.

Another limitation of *preciseTAD* is that it does not distinguish boundary types. The hierarchical nature of TAD boundaries [@weinreb2016identification; @fraser2015hierarchical; @Cresswell:2020ab] is not considered by *preciseTAD* due to the lack of gold standard of TAD hierarchy. *preciseTAD* also does not consider the directionality of CTCF binding [@Harrold:2020aa] as it predicts individual boundaries in contrast to pairs of convergent CTCF motifs marking individual domains. Recent research distinguishes CTCF-associated boundaries, CTCF-negative YY1-enriched boundaries, CTCF- and YY1- depleted promoter boundaries, and the fourth class of weak boundaries largely depleted of all three features [@Krietenstein:2020aa]. Furthermore, actively transcribed regions can serve as TAD boundaries themselves, independently of CTCF binding [@Harrold:2020aa]. This may lead to some TAD boundaries being undetected by *preciseTAD* despite being detected by domain callers. Our future work will involve incorporating the directionality of CTCF binding in predictive modeling, including additional predictor types, and defining separate models trained on different boundary types.

Using the provided pre-trained models, *preciseTAD* is able to identify precise domain boundaries using cell line-specific genome annotation information of only four transcription factors, CTCF, SMC3, RAD21, ZNF143. Although the ENCODE, NIH Roadmap Epigenomics, FANTOM5, BLUEPRINT, and other members of the International Human Epigenome Consortium (IHEC) [@Stunnenberg:2016aa] have been actively cataloging cell line-specific genome annotation datasets, these data remain very sparse. To date (September 2020), the ENCODE project has only five cell lines (GM12878, K562, H1-hESC, HeLa-S3, HepG2) with all four transcription factor genomic annotations. Efforts to <!--systematically catalog chromatin conformation capture data [@Wang:2018ab] and -->impute missing cell line-specific genome annotation information [@Schreiber:2020aa] started to appear. We envision the broader availability of cell line-specific genomic annotations will enable a more systematic analysis of domain boundaries using *preciseTAD*. 
<!--*preciseTAD* requires both cell line-specific chromatin conformation capture data (Hi-C in the current study) and genome annotation data. Hi-C data provides information about the approximate location of domain boundaries, while genome annotation data is used to train a model that can predict boundaries at the base-level resolution. The diversity of chromatin conformation capture technologies [@Kempfer:2019aa] and analysis methods [@Hansen:2019aa] limits the availability of harmonized cell line-specific domain boundary information.-->

In summary, we demonstrate that domain boundary prediction is a multi-faceted problem requiring consideration of multiple statistical and biological properties of genomic data. Simply considering the properties of Hi-C contact matrices ignores the fundamental roles of known molecular drivers of the 3D chromatin structures. Instead, we propose *preciseTAD*, a supervised machine learning framework that leverages both Hi-C contact matrix information and genomic annotations. Our method introduces three concepts - *shifted binning*, *distance*-type predictors, and *random undersampling* - which we use to build random forest classification models for predicting boundary regions. Our method can bridge the resolution gap between 1D ChIP-seq annotations and 3D Hi-C sequencing data for more precise and biologically meaningful boundary identification. We introduce *preciseTAD*, an open source R package for leveraging random forests to predict domain boundaries at base-level resolution. We hope that *preciseTAD* will serve as an efficient and easy-to-use tool to further explore the genome's 3D organization.


<!--
The two proteins differ in that YY1 preferentially
occupies interacting enhancers and promoters, whereas CTCF
preferentially occupies sites distal from these regulatory elements that tend to form larger loops and participate in insulation.
Deletion of YY1 binding sites or depletion of YY1 can disrupt
enhancer-promoter contacts and normal gene expression.
Thus, YY1-mediated structuring of enhancer-promoter loops is
analogous to CTCF-mediated structuring of TADs, CTCF contact domains, and insulated neighborhoods. This model of
YY1-mediated structuring of enhancer-promoter loops accounts
for diverse functions reported previously for YY1, including contributions to both gene activation and repression and to gene
dysregulation in cancer. ["YY1 Is a Structural Regulator
of Enhancer-Promoter Loops"]
-->



# Supplementary Legends

**Additional File 1**. An example script for applying Arrowhead to in situ Hi-C data (HIC001-HIC018) to obtain chromosome-specific TAD boundaries on the GM12878 cell line at 5kb, 10kb, 25kb, 50kb, 100kb resolutions. An example of deepTools commands to generate plotProfiels and plotHeatmaps is also provided. 

**Supplementary Table 1.** Data sources for Hi-C matrices used to call topologically associating domains with Arrowhead, as well as loop boundaries obtained by Peakachu.

**Supplementary Table 2.** A complete list of genomic annotations used to build the predictor space for all downstream models. The GRCh37/hg19 human genome assembly was used. "Genomic Class" - broad category of genomic features, "Element" - names of genomic features, "Cell line-Specific Source" - download URL specific to the cell line (not all annotations were provided by the same institutions).

**Supplementary Figure 1. The $log_{2}$ transformation of genomic distances normalizes their distributions.** Distances are measured as the number of bases from the center of a genomic bin to the nearest genomic annotation center. Density curves of distances before (red) and after (blue) performing a $log_{2}$ transformation across 10 kb, 25 kb, 50 kb, and 100 kb data resolutions for both the (A) GM12878 and (B) K562 cell lines. Each density curve represents an individual genomic annotation (77 total).

**Supplementary Figure 2.** Pseudocode of the *preciseTAD* algorithm.

**Supplementary Table 3.** Domain boundary data and class imbalance summaries across resolutions for Arrowhead and Peakachu on K562. 

**Supplementary Figure 3. Determining optimal data level characteristics for building TAD boundary region prediction models on K562.** (A) Averaged balanced accuracies are compared across resolution, within each predictor-type: Signal, OC, OP, and Distance, and across resampling techniques: no resampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple) when using Arrowhead ground truth boundaries for K562. (B) Averaged balanced accuracies are compared for GM12878 and (C) K562  within each predictor-type: Signal, OC, OP, and Distance, and across resampling technique: no resampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Error bars indicate standard deviation away from the mean performance across each holdout chromosome used for testing.

**Supplementary Figure 4. SMC3, RAD21, CTCF, and ZNF143 transcription factors accurately predict TAD and loop boundaries in K562.** (A) Barplots comparing performances of TAD (Arrowhead) and loop (Peakachu) boundary prediction models using histone modifications (HM), chromatin states (BroadHMM), transcription factor binding sites (TFBS), in addition to a model containing all three classes (ALL). (B) Recursive feature elimination (RFE) analysis used to select the optimal number of predictors. Error bars represent 1 standard deviation from the mean cross-validated accuracy across each holdout chromosome. (C) Clustered heatmap of the predictive importance for the union of the top 8 most predictive chromosome-specific TFBSs. The columns represent the holdout chromosome excluded from the training data. (D) Receiver operating characteristic (ROC) curves and the corresponding average area under the curves (AUCs) when training and testing on K562 data (blue, Arrowhead ground truth; red, Peakachu ground truth) versus training on GM12878 and testing on K5628 data (black, dashed). The curves represent the average sensitivities and specificities across each holdout chromosome. The shaded areas around each curve represent 1 standard deviation from the average.

**Supplementary Table 4.** hg19/GRCh37 genomic coordinates of *preciseTAD*-predicted boundary regions (PTBR) and points (PTBP) for GM12878 and K562 cell lines, using models trained on Arrowhead TAD and Peakachu chromatin loop boundaries as ground truth. For PTBRs, the start and end coordinates define the clusters of spatially proximal bases with the probability of being a boundary equal to 1. For PTBPs, the start and end (start+1) coordinates define the most likely boundary point within each PTBR.

**Supplementary Figure 5.** Signal profile plots comparing the binding strength of top TFBSs around flanked (A) Arrowhead called TAD boundaries (blue) and *preciseTAD*-predicted TAD boundaries (green) on K562, (B) Peakachu chromatin loop boundaries (red) and preciseTAD predicted loop boundaries (green) on K562, (C) Arrowhead called TAD boundaries (blue), Peakachu chromatin loop boundaries (red), and SpectralTAD called TAD boundaries (green) on GM12878 and (D) on K562. 

**Supplementary Figure 6. *preciseTAD* boundaries are more enriched for known molecular drivers of 3D chromatin, as compared with Arrowhead boundaries.** Enrichment heatmaps comparing the signal distribution ofCTCF, RAD21, SMC3, and ZNF143 around Arrowhead-called TAD boundaries vs. *preciseTAD*-predicted TAD boundaries for (A) GM12878 and (B) K562 cell lines.

**Supplementary Figure 7. *preciseTAD* boundaries are more enriched for known molecular drivers of 3D chromatin, as compared with Peakachu boundaries.** Enrichment heatmaps comparing the signal distribution ofCTCF, RAD21, SMC3, and ZNF143 around Peakachu chromatin loop boundaries vs. *preciseTAD*-predicted TAD boundaries for (A) GM12878 and (B) K562 cell lines.

**Supplementary Figure 8. *preciseTAD* boundaries are spatially closer to known molecular drivers of 3D chromatin.** Boxplots comparing the log2 genomic distance distributions from predicted and called boundaries to the nearest (A) GM12878-specific and (B) K562-specific CTCF, RAD21, SMC3, and ZNF143 transcription factor binding sites. p-values are derived from the Wilcoxon Rank Sum test.

**Supplementary Figure 9. The agreement between *preciseTAD*-predicted boundaries using Arrowhead- and Peakachu-trained models**. Venn diagrams of boundary overlaps using (A) GM12878 and (B) K562 data. Boundaries involving Arrowhead/Peakachu were flanked by 5 kb/10 kb, respectively.

**Supplementary Figure 10. Pre-trained *preciseTAD* models accurately predict boundaries on cell lines using annotation data only.** Venn diagrams and signal profile plots comparing flanked predicted boundaries using Arrowhead (A, B) and Peakachu (C, D) trained models. (A, C) train on GM12878 and predict on GM12878 (red, GM on GM) vs. train on K562 and predict on GM12878 (blue, K on GM), (B, D) train on K562 predict on K562 (red, K on K) vs. train on GM12878 predict on K562 (blue, GM on K). Boundaries involving Arrowhead/Peakachu were flanked by 5 kb/10 kb, respectively.

**Supplementary Figure 11. Normalized Enrichment levels suggest t=1.0 and $\epsilon=10000$ as the most optimal parameters for biologically relevant *preciseTAD*-predicted boundaries.** Linecharts illustrating the normalized enrichment (NE) between CTCF, RAD21, SMC3, ZNF143 and resolution-flanked *preciseTAD*-predicted boundaries for different combinations of thresholds (t) and epsilon parameter values (eps). NE was calculated as the total number of ChIP-seq peaks that overlapped within a given flanked boundary, divided by the number of boundaries that were predicted, and averaged over the number of annotations included in the model. Data from GM12878 (A) and K562 (B) cell lines, chromosome 22, at 5kb resolution was used. Error bars indicate standard deviation from the mean.

**Supplementary Table 5.** Additional performance metrics when implementing a random forest using Arrowhead/Peakachu ground truth TAD/loop boundaries. Performances are averaged across each holdout chromosome that was reserved for testing.


# Abbreviations

AUC - Area Under the Curve; BA - Balanced Accuracy; bp - base pair; BroadHMM - BroadHMM chromatin state data; ChIP-seq - Chromatin Immunoprecipitation followed by high-throughput DNA sequencing; CI - Class Imbalance; DBSCAN - Density-Based Spatial Clustering of Applications with Noise; FN - False Negatives; FP - False Positives; Hi-C - High-throughput chromosome conformation capture technology; HM - Histone Modifications; kb - kilobase; ML - Machine Learning; OC - Overlap Counts; OP - Overlap Percents; PAM - Partitioning Around Medoids; PTBP - preciseTAD Boundary Point; PTBR - preciseTAD Boundary Region; RF - Random Forest; RFE - Recursive Feature Elimination; ROC - Receiver Operating Characteristic; ROS - Random Over-Sampling; RUS	Random Under-Sampling; SMOTE - Synthetic Minority Over-sampling Technique; TAD - Topologically Associating Domains; TFBS - Transcription Factor Binding Sites; TN - True Negatives; TP - True Positives

# Acknowledgements

_Conflict of Interest._ None.

<!-- # Funding -->

# References

