---
title: "Computational Prediction of Boundaries of 3D Genomic Domains in Class Imbalance Settings"
csl: styles.ref/genomebiology.csl
output:
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT

**Background.** Chromosome conformation capture sequencing technologies have shown that the three-dimensional (3D) structure of the genome is folded into distinct compartments, known as topologically associated domains (TADs) - units of coordinated gene expression. The location of TAD boundaries is highly conserved, suggesting the presence of epigenomic marks aiding in TAD formation. The ability to predict which epigenomic features are most associated with the TAD boundaries will allow to better understand the regulatory role of the 3D structure of the genome.

Existing methods for predicting associations between genomic elements tend to ignore key characteristics of genomic data. Specifically, the number of TAD boundaries is much less than the number of other genomic regions, leading to heavily imbalanced classes. Furthermore, most methods utilize direct overlap as a means to quantify the association, while distance, the measure of spatial relationships, remains unaccounted for. Consequently, distances on a genomic scale vary widely, leaving uncertainty how the heavily right-tailed distribution of distance measures will affect the modelâ€™s performance.

**Methods.** We propose a novel data pre-processing pipeline that addresses those shortcomings. Lasso and Elastic Net were used for feature selection, coupled with five classifier methods. A number of classifier performance metrics were assessed, including the F1 measure and Matthew Correlation Coefficient (MCC), and area under the ROC curve (AUROC).

**Results.** Data preprocessing (log2-transformation and standardization) improves the performance of classification algorithms and allows for the ability to more accurately predict which genomic features are most associated with TAD boundaries. In contrast to overlap-based association measures, the distance between genomic elements was the most predictive measure. The random undersampling strategy addresses the class imbalance problem the most effectively in nearly all settings.

**Conclusions.** Current methods used to model the epigenomic features associated with TAD boundaries are insufficiently robust to handle properties of genomic data. Models applied to unprocessed data can have poor predictive performances. Focusing solely on standard performance assessment metrics, such as AUROC, can mask poor performance of the models; thus, the use of more balanced metrics, such as F1 and MCC, is warranted. Our model results in better performances and more accurate identification of important features associated with the formation of TAD boundaries.

# INTRODUCTION

The advent of various genome-wide sequencing technologies, such as high-throughput conformation capture (notably Hi-C), have revealed how the spatial organization of the human genome may affect genomic regulation [@lieberman2009comprehensive]. Analyses have shown that the genome is folded into distinct compartments termed topologically associating domains (TADs) [@dixon2012topological]. Evidence suggests that regulatory elements and genes tend to interact more frequently within the same TAD [@symmons2014functional]. Approximately 60-80% of TADs remain stable across cell types [@Schmitt:2016aa], suggestingthat the boundaries of TADs may play a role in restricting the function of certain genomic elements such as enhancers, thereby impacting the transcription of genes. Furthermore, studies have shown that changing the 3D structure of the DNA, causing disruptions in these domains, can lead to adverse outcomes and diseases like cancer [@hnisz2016activation; @flavahan2016insulator]. Therefore, it has become increasingly important to be able to identify the key molecular drivers of the formation of TAD boundaries in order to further our understanding of the human genome.

Identifying the mechanisms underlying the formation of TADs are a complex and active area of research. The distinct patterns of epigenomic marks, binding peaks of different factors, and actively transcribed genes [@xiao2012comparative; @milosavljevic2011emerging] [??? REFs; !!!added refs] prompt the use of computational approaches to identify which epigenomic features are most predictive of the TAD boundaries. Current methods rely on the use of classification algorithms to draw associations between the spatial relationship of functional genomic elements and regions that include TAD boundaries [@mourad2016computational; @hong2017computational; @huang2015predicting]. Although several useful insights have been made, such as the convergent binding of CTCF [@ong2014ctcf; @ghirlando2016ctcf] [??? Elaborate with refs.; !!!added refs], many methods ignore key characteristics of domain data, such as heavily imbalanced number of boundaries vs. other regions. In a classification setting [??? are we talking about classification or prediction?; !!!prediction is a byproduct of classification correct?; wait to discuss in meeting], when one is attempting to discriminate between regions with and without TAD boundaries, this will contribute to small imbalance ratios (defined as the ratio of minority to majority classes) [@Orriols-Puig2008]. As has been reported in the machine learning community, imbalanced datasets are known to adversely impact the performance of the classifiers as the learned model is biased towards the majority class to minimize the overall error rate [@lusa2010class;@chen2013influence;@jeni2013facing]. Similarly, careful consideration must also be taken when deciding which performance metric(s) to use for evaluation of predictive models, especially in the case of imbalanced data. 

Commonly used measures of model performance for binary classifiers include the area under the receiver operating curve (AUROC). The popularity surrounding the use of ROC curves is that they do not depend on a threshold. However, when the prevalence of an event is low--that is, a defined genomic region containing a TAD boundary--mean probabilities are biased towards the majority class. This results in the area of interest under an ROC curve to be compressed toward a small corner in the ROC space [@prati2011survey]. While not directly affecting the value of the AUROC, this can have severe consequences on the interpretability of the model's performance. A better threshold-free measure to use when dealing with imbalanced data in the area under the precision-recall curve (AUPRC). It has been shown that the use of AUPRC with imbalanced datasets yields more accurate and intuitive interpretation of model performance when compared to AUROC [@saito2015precision]. Likewise, metrics like accuracy and specificity are also problematic when evaluating model performance on imbalance data due to being highly influenced by the correct classification of the majority class. Better measures to consider are the F1-Score and Matthew's Correlation Coefficient (MCC) [@jeni2013facing]. The F1-Score can be interpreted as a weighted average of the precision and recall values, while MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications. Both metrics use all 4 quadrants of the confusion matrix, thus providing more balanced measures in the case of imbalanced data [@chicco2017ten].

In addition to employing suitable metrics for evaluating models validated on imbalanced data, there also exist several algorithmic or data level solutions. Algorithmic level solutions involve modifyng existing classifiers in order to curtail the bias induced by imbalanced classes [@estabrooks2004multiple; @chawla2003smoteboost; @chen2004using; @khoshgoftaar2007empirical]. Data level solutions mostly constitute as resampling techniques, in which random sampling is used to either replicate or removed data points in order to create a more balanced dataset [@chawla2004special; @chawla2009data]. It has been shown that quite simple resampling techniques can drastically improve model performance in the face of imbalanced data [@japkowicz2002class; @dubey2014analysis]. For the purposes of this study we focus our attention toward data level solutions so that all results may be classifier independent, and thus, more easily comparable. 

Furthermore, there are other aspects of TAD boundary modelling that current methods have failed to consider. Most notably is the concept of genomic distance. As a feature engineering step, the use of distances (in base pairs) from TAD boundaries to regions of various genomic features has largely been ignored. However, these distance characteristics have the potential to offer a more accurate spatial representation of which genomic features are associated with TAD boundaries. 

Additionally, model performance is heavily contingent upon the resolution of the contact matrix used to identify the location of TADs throughout the genome. It is often unclear from previous studies what resolution contact matrices were used when retrieving TAD boundaries, and how other resolutions would affect their results. A larger resolution contact matrix will result in the identification of larger-sized TADs, and subsequently, less TAD boundary points. This will contribute to smaller sample sizes and could effect predictive performances.

In this study we have developed an ensemble framework for predicting TAD boundaries that addresses class imbalance problem. We evaluated Hi-C data for both the GM12878 and K562 cell line obtained at 10 kb, 25 kb, 50 kb, and 100 kb resolution. For each resolution, we examined the inclusion of three different predictor types, including distance as the measure of association between genomic elements. We then tested four re-sampling techniques aimed at creating balanced classes. Random forest classification was performed to both compare each combination of the above ensemble, and investigate which functional genomic features were most associated with the TAD boundaries.  Model performance was evaluated using area under the precision-recall curve (AUPRC), F1-Score, and MCC. 

# METHODS

## DOMAIN DATA

Replicate Hi-C data for the GM12878 and K562 cell lines were downloaded from GEO GSE63525 [@rao20143d] (Supplementary Table 4) [??? Need supplementary ref.; !!!added supp table] The autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries were obtained from data at 10 kb, 25 kb, 50 kb, and 100 kb resolution using the Arrowhead algorithm [@durand2016juicebox]. The start and end coordinates of all TADs were concatenated, sorted, and unique TAD boundary points were obtained. 

## GENOMIC ANNOTATION DATA

Genomic annotation data (GRCh37/hg19 human genome assembly) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 1). The annotations consisted of a variety of different chromatin states, histone marks, and transcription factor binding sites using ChIP-seq peak experiments for the GM12878, K562, ... cell lines. Each genomic annotation was represented by a set of regions (their genomic coordinates) annotated as having a functional property or biological significance.


## MODEL CONSTRUCTION

To establish the data for modeling, the genome was binned into equally sized intervals defined by the data resolution for the corresponding contact matrix (Figure 1). For example, for domain data obtained from a 10 kb resolution contact matrix, the genome was binned into 10 kb intervals. To accommodate for the uncertainty in exact boundary location, the boundary points were flanked on both sides by the size of the resolution of the corresponding contact matrix. Returning to the 10 kb resolution example, each uniquely identified TAD boundary was flanked on either side by 10 kilobases for a total width around each boundary point of 20 kb. A genomic bin was labeled as having a TAD boundary within it (Y=1) if it overlapped with a particular flanked boundary. Otherwise, the bin was labeled as not containing a boundary (Y=0).

## ENSEMBLE FRAMEWORK

An ensemble of models was established for the purposes of systematically comparing the performances of different parameters in the framework (Figure 2). Each system in the framework is assumed to be composed of a model constructed from a particular data resolution, with a particular predictor type for the feature space.

The data sets used for modeling were composed of the full set of genomic bins that either overlapped with a flanked TAD boundary or did not. These two classes made up the minority and majority sets respectively. The data was then split into a 7:3 ratio of training to testing sets. Each training and testing set was composed of a similar majority to minority class ratio as the full data set. For the task of performance evaluation and prediction, the test set remained constant for all ensemble systems to account for a fair and unbiased comparison.

### FEATURE ENGINEERING AND REDUCTION USING GENOMIC ANNOTATIONS

Using the sets of genomic annotations, we created three types of predictors that were used as features for subsequent models to analyze their respective spatial relationship with TAD boundaries (Figure 3):

1. **Overlap Counts (OC)**: For each genomic bin, the total number of instances where a bin overlapped with genomic annotation regions (features) was calculated. In the case of no overlaps the value is set as 0.
   
2. **Overlap Percent (OP)**: For each genomic bin, the percent overlap between the feature width and the total width of the bin was calculated. We defined the feature width as the number of bases that overlapped with the genomic bin. The percent was then calculated by dividing the feature width by the bin width. If multiple overlaps existed, the feature width was defined as the sum of the total number of bases that overlapped with a particular bin. Note that the bin width remained constant for a given resolution (either 10 kb, 25 kb, 50 kb, or 100 kb). Values for bins with no overlaps were set to be 0.

3. **Log2 Distance**: For each genomic bin, the $log_{2}$ of the distance (refered to hereafter simply as $\textit{distance}$), in basepairs, from the center of the bin to the center of the nearest region defined by a specific genomic annotation [??? Decide whether to use "annotation" or "feature"; !!!go with annotation] was calculated. The log transformation was used as a normalization technique in order to account for the skewness of the data (Supplementary Figures 1-8).

### RE-SAMPLING TECHNIQUES

In order to evaluate the problem of imbalanced classes, we then evaluated four re-sampling techniques (in addition to no re-sampling).

1. **No Sampling**: All of the data points from the majority and minority class of the training set were used.

2. **Random Under-Sampling (RUS)**: All of the minority classes from the training set were used. Sampling without replacement was used to obtain the same number of the majority classes. We performed 50 iterations of random under-sampling. At each iteration a random forest classification algorithm was performed. Performance metrics were aggregated by taking the average across all of the iterations.

3. **Random Over-Sampling (ROS)**: All of the majority classes from the training set were used. Sampling with replacement was used to obtain the same number of the minority classes. Similar to the under-sampling technique, 50 iterations were performed and the results of a random forest classification were aggregated by taking the average across iterations.

4. **SMOTE** (Synthetic Minority Over-Sampling Technique): This method incorporates both random under- and over-sampling. Under-sampling is performed without replacement from the majority class, while over-sampling is performed by creating new synthetic observations using the minority class [@chawla2002smote]. To account for the bias introduced by the under-sampling, we performed 50-iterations of SMOTE. We chose the parameters that would create perfectly balanced classes. The `DMwR` R package was used to perform SMOTE (version 0.4.1). 

5. **ROSE** (Random Over-Sampling Examples): Similar to SMOTE, ROSE combines techniques of over-sampling and under-sampling. However, contrary to SMOTE, all of the data generated by ROSE is synthetic. The data is created by sampling from one of the two classes, then generating a new example in its neighborhood using a smoothed bootstrap approach. The shape of the neighborhood is determined by the kernel density estimate of the two classes [@lunardon2013r]. Another important distinction between ROSE and SMOTE is that, for ROSE, the creation of perfectly balanced classes is unlikely due to the randomness in sampling between the two classes. The `ROSE` R package was used to perform ROSE (version 0.0.3).

### CLASSIFICATION & CROSS VALIDATION 

A classification algorithm in the form of a random forest was performed on the various resampled and reduced training sets and validated using the same test set for all models in the ensemble. To reduce bias due to random dataset generation, 10-fold cross validation was used. The default number of features to consider at each node of the random forest algorithm was set as the square root of the number of features in the model. Likewise, the number of ensemble trees to aggregate was set at 500. Once the model was implemented, it was assessed using the test set, and performance metrics were then recorded. 

### PERFORMANCE METRICS 

The efficacy of different ensemble systems was compared using the F1-score and MCC. These metrics are defined as follows:

$$F1 - Score = \frac{2TP}{2TP + FP + FN}$$
$$MCC = \frac{TP \times TN - FP \times FN}{\sqrt{ (TP + FP)(TP+FN)(TN+FP)(TN+FN) }}$$


Here, TP refers to the number of bins correctly identified as containing a TAD boundary (true positives), FP refers to the number of bins incorrectly identified as containing a TAD boundary (false positives), TN refers to the number of bins correctly identified as not containing a TAD boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a TAD boundary. Each of these quantities are obtained from the confusion matrix created by validating the model on the testing set. The evaluation metrics listed above are all considered single threshold measures. That is, they are defined for individual thresholds that separated class prediction. Thererfore they are unable to give an overview of the potential range of performance as thresholds vary. In order to evaluate the trade-off between precision and recall, we also considered a threshold-free measure in form of the area under the precision recall curve (AUPRC).

###MODEL ASSESSMENT AND PREDICTION

We first assessed predictor type performance using random forest classification algorithms by utilizing each set of predictor types as the feature space, in addition to a model with all combined types. Models were evaluated across each combination of resampling technique applied to the training set, for each cell line. Performance was based on the AUPRC. Recursive feature elimination (RFE) was then used to determine the optimal number of features for models with the best performing predictor type, across resampling techniques and cell lines. In doing so, random forest models were again used to recursively assign weights to a set of features of a given size. For computational efficiency we evaluated feature subset sizes in descending order of powers of 2. The optimal number of features was determined by the AUPRC of the corresponding model. Optimal performance was assessed and the feature spaced was reduced accordingly.

For the task of identifying the functional genomic elements that were most predictive of the formation of TAD boundaries, we evaluated the variable importances associated with each predictor of the random forest models associated with each resampling technique. For each model and within each tree, the AUPRC on the out-of-bag portion of the data was recorded. Then the same was done after permuting each predictor variable. The difference between the two AUPRC values was then averaged over all trees, and normalized by the standard error. All random forest algorithms were implemented in R using the `caret` package version 6.0 [@kuhn2012caret]. 

# RESULTS

## ENSEMBLE COMPARISONS

We evaluated and compared multiple random forest classification algorithms built on domain data extracted from Hi-C datasets for 2 cell lines C = {GM12878, K562}, at 4 different resolutions R = {10 kb, 25 kb, 50 kb, 100 kb}. For each classifier, we compared 3 predictor types P = {OC, OP, Distance} and 5 class balancing techniques B = {None, RUS, ROS, SMOTE, ROSE}. A system produced from the ensemble framework was defined as: E = {c, r, p, b}, where $c \in C, \quad R \in R, \quad p \in P,$ and $b \in$ B. Letting |X| denote the cardinality of each set, we evaluated $|C| \times |R| \times |P| \times |B|$ different ensemble systems, totaling X (=$x \times 4 \times 3 \times 5$) separate models. For each model, we aimed to assess both performance and identify the genomic annotations that were most influential in correctly labeling the genomic bins that overlapped with flanked TAD boundaries. For simplicity, we present the most significant results, with the complete results available as Supplementary Tables and Figures.

We first examined the class imbalance of the different data resolutions (Supplementary Table ). While the total number of genomic bins decreased as resolution increased, the imbalance ratios remained relatively stable. We found that there was an average imbalance ratio of 0.154 (SD=0.022) and 0.137 (SD=0.0181) for cell lines GM12878 and K562 respectively. The low imbalance ratios indicate a wide disparity between majority and minority classes, and therefore suitable resampling techniques are needed.  

## DISTANCE TYPE PREDICTORS OUTPERFORM OTHER PREDICTOR-TYPES

For each cell line and resolution, we balanced the data using each resampling technique and performed a series of random forest classification algorithms on the full features space (all combined predictor types), as well as on type-specific feature spaces (overlap counts, overlap predictors, and distances). PR-curves were plotted and compared for 10 kb resolutions (Figures 4-8). For all curves, the baseline threshold used to assign a performance to a random model is defined as the ratio of genomic bins that overlapped with a flanked TAD boundary divided by the total number of genomic bins for the test set, respective to cell line and resolution. It was found that for the 10 kb resolution data, models with distance type predictors had consistently higher AUPRC values compared to other predictor types, across both cell line and resampling technique. This result was observed for the other resolutions as well (Supplementary Figuers 9-23l; Supplementary Table 2). 

Recursive feature elimination was used to establish the optimal number of distance-type features to include in downstream analyses. It was determined that, for 10 kb resolution data, near optimal performance was obtained for models with only 32 predictors, across each resampling technique and for each cell line (Figures 9-13). Results remained consistent across the other data resolutions (Supplementary Figures 24-38). Therefore, the feature space was reduced for subsequent models to the top 32 predictors, respective to cell line, resolution, and resampling technique. 

## DATA RESAMPLING IMPROVES MODEL PREDICTION

 



# Discussion


# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
