---
title: "preciseTAD: A machine learning framework for precise TAD boundary prediction at base pair resolution"
csl: styles.ref/genomebiology.csl
output:
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
  pdf_document: default
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT (Spiro version)

Chromosome conformation capture combined with high-throughput sequencing experiments (Hi-C) have revealed that chromatin undergoes layers of compaction through DNA looping and folding, forming dynamic 3D structures. Among these are Topologically Associating Domains (TADs), which are known to play critical roles in cell dynamics like gene regulation and cell differentiation. Precise TAD mapping remains difficult, as it is strongly reliant on Hi-C data resolution. Obtaining genome-wide chromatin interactions at high-resolution is costly resulting in variability in true TAD boundary location by TAD calling algorithms. To aid in the precise identification of TAD boundaries we developed preciseTAD, a machine learning framework that leverages the spatial relationship of many high resolution ChIP-seq defined genomic elements, coupled with density-based clustering  and scalable partitioning techniques. Our framework precisely predicts chromosome-specific TAD boundaries on multiple cell types. We show that known molecular drivers of 3D chromatin including CTCF, RAD21, and SMC3 are more enriched at our predicted TAD boundaries compared to the boundaries identified by the popular ARROWHEAD TAD caller. Our results provide useful insights into the 3D organization of the human genome.

# ABSTRACT (MD version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUROC) curve, Precision-Recall curve (AUPRC), F1, Matthews Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were the best type of genomic annotations for predicting TAD boundaries when directly compared with histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# Introduction

The advent of chromosome conformation capture (3C) sequencing technologies, and its successor Hi-C, have revealed a hierarchy of important 3-dimensional (3D) spatial constructs within the human genome such as chromatin loops [@rao20143d], Topologically Associated Domains (TADs) [@nora2012spatial;@dixon2012topological], and A/B compartments [@lieberman2009comprehensive]. At kilobase scale, gene promotors are often connected via distal regulatory elements, such as enhancers, through chromatin loops, thereby regulating gene expression [@Lupianez:2015aa; @Franke:2016aa]. Further out, at the megabase scale, TADs represent regions on the linear genome, that are highly self-interacting. Emerging evidence has linked TADs to critical roles in cell dynamics and cell differentiation. Studies have shown that TADs themselves are highly conserved across species and cell lines [@dixon2012topological;@krefting2018evolutionary]. Likewise, disruption of the boundaries that demarcate TADs and loops have been shown to lead to cancer [@hnisz2016activation; @taberlay2016three] and other disorders [@lupianez2016breaking;@Sun:2018aa;@Meaburn:2007aa]. Therefore, precisely identifying the location of TAD and chromatin loop boundaries remains a top priority in our goal to fully understand the functionality of the human genome.

Several methods have been proposed to identify the genomic coordinates of TAD boundaries throughout the linear genome [@dixon2012topological;@rao20143d;@sexton2012three;@chen2016novo;@filippova2014identification;@shin2016topdom;@weinreb2016identification], less so for chromatin loops [@salameh2019supervised;@kai2018predicting]. [??? Check refs, the current ones are suboptimal. Check SpectralTAD paper for refs about TAD callers, use https://github.com/mdozmorov/HiC_tools#loop-callers for loop callers] However, a key limitation among most of these TAD- and loop callers, referred hereafter as TAD callers for brevity, is that they are heavily reliant on Hi-C data *resolution*. Resolution refers to the sizes of genomic bins used to segment the linear genome as a way to deal with low signal-to-noise incurred by the high dimensional nature of Hi-C data [@li2019comparison]. Lower resolutions contribute to larger bin sizes, which lead to uncertainty in the location of TAD boundaries. Due to high sequencing costs, most available Hi-C datasets have relatively low resolution, ranging from 25 kb to 100 kb. Because increasing the resolution of Hi-C data requires a quadratic increase in the total sequencing depth, obtaining high resolution remains difficult [@schmitt2016genome]. While these low resolution Hi-C datasets can be used to probe higher order genomic patterns such as A/B compartments, more refined resolutions are necessary to accurately define TADs and chromatin loops [@fudenberg2016formation].

Another limitation among TAD callers is that they disregard prior knowledge about functional genomic elements associated with TAD and loop boundaries. The insulator binding protein, CTCF, and additional cofactors such as SMC3 and RAD21 have been hypothesized to comprise components of the loop extrusion model, whereby DNA is extruded through cohesin rings forming chromatin loops [@hansen2018recent;@fudenberg2016formation;@alipour2012self;@sanborn2015chromatin]. [??? Rewrite, simplify, describe enrichment first, loop extrusion second] Additionally, distinct patterns of histone modifications have also been shown to be featured at boundaries, including H3K36 timethylation (H3K36me3) [@lieberman2009comprehensive;@dixon2012topological;@libbrecht2015joint]. These elements are annotated on the genome using chromatin immunoprecipitation followed by high-throughput DNA sequencing (ChIP-seq). Even at its highest (750 bp [@bonev2017multiscale]), Hi-C sequencing resolution still remains well below the resolution of ChIP-seq experiments, which ranges in the tens of base pairs [@mendoza2009genome]. Therefore, leveraging such elements in a supervised machine learning framework lends itself to the possibility of more precisely predicting the locations of TAD (or loop) boundaries at much higher resolutions. 

Here we present *preciseTAD*, an R package for precise identification of TAD(or loop) boundaries under a supervised machine learning framework. Our method builds upon an optimized random forest (RF) classification model that leverages high resolution ChIP-seq data in the form of regulatory elements (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS). We introduce a novel technique for building RF classifiers in a boundary prediction setting composed of a resolution-specific binning process and feature engineering procedure. *preciseTAD* then employs density-based clustering (DBSCAN) and scalable partitioning around medoids (CLARA) on prediction probabilities to select the most biologically meaningful boundary coordinates on the linear genome, thereby circumventing resolution restrictions of conventional TAD- and chromatin loop-callers. We demonstrate that predictions made by *preciseTAD* are more enriched for known molecular drivers of 3D chromatin including CTCF, RAD21, SMC3, and ZNF143. Further, we show that *preciseTAD* boundaries are more conserved across cell lines, highlighting their biological significance. This improved precision in TAD and loop boundary location can provide insight into the role that epigenomics has on the spatial architecture of the human genome. [??? Rewrite. Say the main methodology first - model on 4 TFBSs, uses distance, RUS, technique for bp-level prediction, one sentence per point. Detail next - BroadHMM and histone were tested, OC, OP were tested, various sampling techniques. Conclude with findings and package availability]

# Methods

## Data sources

Within this study we used both TAD boundary and loop boundary data derived from the publically available tools ARROWHEAD [@durand2016juicebox] and PEAKACHU [@salameh2019supervised] respectively. ARROWHEAD is part of the Juicer suite of tools for Hi-C data analysis and visualization. To obtain TAD boundaries we applied ARROWHEAD using downloaded primary Hi-C data from Geo GSE63525 [@rao20143d] for the GM12878 and K562 cell lines (libraries HIC001-HIC018 and HIC069-HIC074 for GM12878 and K562 respectively) produced with the MboI restriction enzyme (Supplementary Table 1). The contact matrices were constructed from all read pairs that uniquely aligned to the genome (MAPQ>0). The default tools for ARROWHEAD were used when extracting the autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries at 5 kb, 10 kb, 25 kb, 50 kb, and 100 kb resolutions (Additional File 1). For chromatin loop boundaries, we used loops identified by PEAKACHU at 10 kb for the GM12878 and K562 cell lines [@salameh2019supervised]. Unique boundaries were considered as the midpoints within the coordinate of each chromatin loop. Chromosome 9 was excluded from all downstream analyses due to the sparsity of the contact matrix at the 5 kb and 10 kb resolutions for the K562 cell line. Cell line-specific genomic annotations were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] in the form of BroadHMM chromatin segmentations (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS) (GRCh37/hg19 human genome assembly, Supplementary Table 2). 

## Shifted-binning for binary classification

In Hi-C, each chromosome is binned into non-overlapping regions of the same length *r*, typically 5 kb and above. The *r* parameter defines the resolution of Hi-C data. Consequently, the location of TAD boundaries is limited by the resolution *r*. Here, we designed a strategy called _shifted binning_ that partitions the genome into regions of the same length *r*, but with middle points corresponding to boundaries defined by the original binning. 

To create shifted binning, the first chromosome-specific bin was set at half of the resolution *r*, and continued in congruent intervals of length *r* until the end of the chromosome (*mod r* + *r/2*). The shifted genomic bins, referred hereafted as bins for simplicity, were then defined as boundary regions (*Y = 1*) if they contained a TAD (or loop) boundary, and non-boundary regions (*Y = 0*) otherwise, thus establishing the binary response vector (**Y**) used for classification (Figure 1).

## Feature engineering

Cell line-specific genomic annotations were used to build the feature space for our boundary prediction framework. Bins were annotated by one of either the number of overlaps with an annotation (*Overlap Count (OC)*), the percent of overlap between the bin and the total width of genomic annotation regions overlapping it (*Overlap Percent (OP)*), the average signal strength of the corresponding annotation (*Peak Signal Strength*), or the distance in bases from the center of the bin to the center of the nearest genomic annotation region (*Distance*). A $log_2$-transformation of distance was used to account for the skewness present in the distributions of genomic distances (Supplementary Figure 1). It should be noted that models built using a *Peak Signal Strength* feature space were only composed of histone modifications and transcription factor binding sites. That is, they did not include any BroadHMM chromatin segmentations.

## Addressing class imbalance

To assess the impact of class imbalance (CI), defined as the proportion of boundary regions to non-boundary regions, we evaluated three re-sampling techniques including *Random Over-Sampling (ROS)*, *Random Under-Sampling (RUS)*, and *Synthetic Minority Over-Sampling Technique (SMOTE)*. For ROS, the minority class was sampled with replacement to obtain the same number of data points in the majority class. For RUS, the majority class was sampled without replacement to obtain the same number of data points in the minority class. For SMOTE, under-sampling was performed without replacement from the majority class, while over-sampling was performed by creating new synthetic observations using the $k=5$ minority class nearest neighbors [@chawla2002smote] (implemented in the `DMwR` v.0.4.1 R package). We restricted the algorithm to 100% over-sampling and 200% under-sampling, which creates perfectly balanced classes. 

## Establishing computational framework for assessing model performance

Random forest (RF) classification models (the `caret` v.6.0 R package [@kuhn2012caret]) were built to compare model performances between combinations of data resolutions, feature engineering procedures, and re-sampling techniques. First, the data matrix $A_{N \times (p+1)}$ was constructed by combining the binned genome from each of the $n-1$ autosomal chromosomes ($1,2,\cdots, i-1,i+1,\cdots,22$) as well as a particular feature engineering procedure. Re-sampling was then performed on $A$. An RF classifier was then trained on $A$ using 3-fold cross-validation to tune for the number of annotations, *p*, to consider at each node (*mtry*). The number of trees (*ntree*) that were aggregated for each RF model was set to 500. The minimum number of observations per root node (*nodesize*) was set to 1. To evaluate model performance, the $i^{th}$ chromosome was used for testing. The performance of each combination of model inputs was compared using Balanced Accuracy (BA), defined as the average of sensitivity and specificity:

$$BA = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$

where TP refers to the number of bins correctly identified as containing a boundary (true positives), FP refers to the number of bins incorrectly identified as containing a boundary (false positives), TN refers to the number of bins correctly identified as not containing a boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a boundary (false negatives). Each of these quantities are obtained from the confusion matrix created by validating the model on the test data. Performances were aggregated across each $i^{th}$ holdout chromosome using means and standard deviations.

## Feature selection and predictive importance

Many genomic annotations, notably architectural proteins, tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To suitably reduce the feature space and improve computational efficiency, while maintaining optimal performance, we implemented recursive feature elimination (RFE). In doing so, we estimated the near optimal number of necessary features to include in the best fitting RF model. The best performing subset was identified, ranging from 2 to the maximum number of features by powers of 2. We then aggregated the predictive importance of the union of the optimal set of features across chromosomes using the mean decrease in node impurity among permuted features in out-of-bag samples to determine the most common and top ranked annotations for predicting boundary regions. 

## Boundary prediction at bp resolution using *preciseTAD*

To investigate whether we could alleviate resolution limitations of conventional TAD and chromatin loop calling tools, we developed *preciseTAD*, an algorithm that leverages a random forest model in conjunction with density-based and partitioning techniques to predict boundaries at base pair (bp) resolution. To demonstrate, we implemented an optimally built RF model using the framework described above. That is, a random forest model built on the optimal combination of data resolution (binning), predictor type (feature space), re-sampling technique (class balancing correction), and top ranked annotations (most informative genomic elements).

First the vector, $p_{n_{i}}$, denoting each base pair's probability of being a boundary point from the RF model, was extracted, where $n_{i}$ represents the length of chromosome $i$. Next, *DBSCAN* (version 1.1-5) was applied to the matrix of pairwise genomic distances between base pairs with $p_{n_{i}}=1$. This was to avoid the inflation of predicted boundaries that can occur naturally when dealing with such fine resolution, namely, due to spatially close base pairs with similar predictor profiles (i.e. similar overlap counts, distances, etc.). The DBSCAN algorithm takes two parameters, *eps* ($\epsilon$) and *MinPts*. The eps parameter was assigned as the respective data resolution. The heuristic being that density-reachable base pairs with genomic distances less than resolution should occupy the same designated cluster. The corresponding MinPts parameter was set to recommended value of 3, representing $1+dim(data)$ [@hahsler2019dbscan]. The resulting clusters of highly predictive base pairs identified by DBSCAN were termed *preciseTAD boundary regions* (PTBR). To precisely identify a single base pair among each PTBR, *preciseTAD* then implements an extension of partitioning around medoids (PAM), known as CLARA (Clustering Large Applications) for better scaling and more computational efficiency [@reynolds2006clustering; @wei2000empirical]. Thusly, the corresponding cluster medoid was defined as a *preciseTAD boundary point* (PTBP), making it the most representative and biologically meaningful base pair coordinate within each clustered PTBR.

## Comparing called vs predicted boundaries

Next, we assessed the biological significance of our predicted boundaries with respect to the enrichment profiles of several known molecular drivers of 3D chromatin architecture including CTCF, RAD21, SMC3, and ZNF143 ChIP-Seq peaks. To do so, we used the computeMatrix and plotHeatmap tools offered by DeepTools [@ramirez2016deeptools2] via integration with the Galaxy project [@afgan2018galaxy]. Additionally, we compared the median $log_2$ genomic distances between TAD boundaries and the same top predictive ChIP-seq annotations using Wilcoxon Rank-Sum test. Furthermore, we assessed the hypothesis that TADs and chromatin loops are conserved  [@dixon2012topological; @nora2012spatial; @dixon2015chromatin; @rao20143d; @Schmitt2016aa] by comparing predicted and called boundaries between the GM12878 and K562 cell lines. Boundaries were first flanked by resolution, *r*, and overlaps were visualized using Venn diagrams from the `Vennerable` R package (version 3.1.0) and evaluated using the Jaccard index defined as

$$J_{(A,B)} = \dfrac{A \cap B}{A \cup B}$$

where A and B represent genomic regions created by flanking called and predicted boundaries. That is, between cell lines, the number of overlapping flanked boundaries divided by the total number of flanked boundaries. Wilcoxon rank sum tests were used to compare chromosome-specific jaccard indices across cell lines, between *preciseTAD* boundaries and both ARROWHEAD and PEAKACHU boundaries respectively. All statistical analyses were performed in R version 4.0.1. The alpha level of significance was set to 0.05 for all statistical tests.

<!--using the `EnrichedHeatmap` R package (version 1.14.0) [@gu2018enrichedheatmap]. For each boundary (both called and predicted), a 5000 bp flank was created on either side. The flanking was broken up into 50 bp segments, for a total of 100 windows on both sides of a given boundary. A coverage value was generated for each window with respect to a given ChIP-seq annotation, as the mean signal averaged by the width of the window, denoted by,

$$v_{c} = \dfrac{\Sigma_{j}^{m} \Sigma_{i}^{n}x_{ij}w_{ij}}{L}$$

where $x_{ij}$ is the total number of base pairs of annotation $i$ in window $j$ and $w_{ij}$ is the number of base pairs that overlap between annotation $i$ and window $j$. The denominator, $L$, is the width of the windows (here, $L=50$). The rows of the heatmaps are ordered by enriched scores. The enriched scores are calculated as the sum of coverage values weighted by the distance to the flanked boundary, denoted by

$$s_{e} = \Sigma_{d=1}^{n_{1}}\left(\dfrac{x_{1d}* d}{n_{1}} \right) + \Sigma_{u=1}^{n_{2}}\left(\dfrac{x_{2u}* (n_{2} - u + 1)}{n_{2}} \right)$$
where $n_{1}$ and $n_{2}$ represent the number of downstream and upstream coverage values to sum over. Here, $n_{1}=n_{2}=100$.-->

# Results

## Developing a machine learning framework for optimal TAD boundary prediction

We developed a machine learning framework<!--, refered to as *TADRF*,--> aimed at determining the optimal set of data level characteristics necessary for topologically associated domain (TAD) boundary prediction. We chose the random forest (RF) machine learning algorithm as our binary classification tool. The reason for choosing RF was two-fold: (1) In order to devise a tunable prediction rule in a supervised learning framework that is both robust to overfitting and able to handle multiple correlated predictors, and (2) to allow for an interpretable ranking of predictors with respect to their ability to predict boundary regions [@boulesteix2012overview]. We then developed *preciseTAD*, an algorithm which leverages an optimally built RF to precisely predict the genomic coordinates of both TAD and chromatin loop boundaries at base pair resolution.

Our framework is resolution-specific and uses called TAD or loop boundaries as ground truth to establish TAD(loop)-boundary and non-TAD(loop) boundary regions. To demonstrate, we applied our framework to publically available replicate Hi-C data sets from the GM12878 and K562 cell lines obtained from GEO GSE63525 (Supplementary Table 1). TADs were subsequently called using the ARROWHEAD pipeline from the juicer toolbox [@durand2016juicer] at five data resolutions: 5 kb 10 kb, 25 kb, 50 kb, and 100 kb. For loop boundaries, we opted to use boundaries obtained by the PEAKACHU prediction tool for the GM12878 and K562 cell lines at 10 kb resolution [@salameh2019supervised]. [??? TAD and loop boundaries should be distinguished, to describe when we have all results]

Boundary regions were defined as regions containing a called boundary ($Y = 1$), while non-boundary regions were defined as genomic bins that did not contain a called boundary ($Y = 0$) (Figure 1A, see Methods). The total number of TADs, TAD boundaries, and the number of genomic bins expectedly decreased with the decreased resolution of Hi-C data (Table 1, Supplementary Table 3). The number of non-boundary regions highly outnumbered boundary regions, with results consistent for loop boundaries as well (Supplementary Table 4). Such disproportional presence of examples in one class is known as a "class imbalance" problem that negatively affects model performance [@prati2011survey; @wei2013role]. To address class imbalance, we evaluated the effect of three re-sampling techniques on prediction performance. *Random over-sampling* (ROS) was defined as sampling with replacement from boundary regions. *Random under-sampling* (RUS) was defined as sampling with replacement from non-boundary regions. Lastly, we considered *Synthetic minority over-sampling technique* (SMOTE), which is a combination of both random over- and under-sampling [@chawla2002smote] (see Methods). These strategies allowed us to select the best performing model.

A total of 78 cell line specific functional genomic annotations were used to build the feature space of downstream random forest models. These annotations included BroadHMM chromatin segmentations, histone modifications (HM), and transcription factor binding sites (TFBS, Supplementary Table 2). Four feature engineering procedures were developed for each genomic annotation (Figure 1B, see Methods) with the goal to improve boundary predictions by accounting for signal strength association (Signal), direct (OC), proportional (OP), and spatial (Distance) relationships between TAD boundaries and genomic annotations.

![**Illustrating the resolution-specific data construction and feature engineering used for downstream random forest modeling.** (A) The linear genome was subsequently binned into resolution-specific intervals. The response vector **Y** was defined as a genomic bin that either overlapped with a TAD boundary ($Y = 1$) or did not ($Y = 0$). (B) Four types of associations between bins and genomic annotations (green shapes) were then used to build the feature space including _Peak Signal_ (Signal), _Overlap Counts_ (OC), _Overlap Percent_ (OP), and ($log_{2}$) _Distance_.](Figures/figure1_model_construction_and_predictor_types.png)

In summary, we considered combinations of data from a particular cell line *L = {GM12878, K562}*, at a given resolution *R = {5 kb, 10 kb, 25 kb, 50 kb, 100 kb}*, with a specific engineered feature space *P = {Signal, OC, OP, Distance}*, and re-sampling technique *S = {None, RUS, ROS, SMOTE}* (Figure 2). Once the model inputs were established, a random forest classifier was trained on $n-1$ autosomal chromosomes, while reserving the $i^{th}$ chromosome for testing. Three-fold cross-validation was used to tune the *mtry* hyperparameter, while *ntree* and *nodesize* were fixed to default values of 500 and 1 respectively. Model inputs were evaluated by aggregating the mean balanced accuracy (BA) across each holdout chromosome (See Methods). 

![**A diagram of the machine learning framework detailing the steps for combining different inputs used in the random forest model building pipeline.** Step 1 is to establish the chromosome-specific rowise (genomic bins) and columnwise (genomic annotations) dimensions of the data matrix, **A**, by combining a resolution (5 kb, 10 kb, 25 kb, 50 kb, or 100 kb) and feature engineering procedure (overlap count (OC), overlap percent (OP), $log_{2}$ distance (Distance)) for the $n-1$ autosomal chromosomes. Step 2 is to reserve the binned data matrix for the holdout chromosome $i$ as the test data. Step 3 is to apply a re-sampling technique to the training set (no re-sampling, random over-sampling, random under-sampling, or SMOTE re-sampling). Step 4 is to train the random forest model and perform 3-fold cross-validation to tune for the *mtry* parameter. Finally, step 5 is to validate the model on the separate test data and evaluate model performance using balanced accuracy (BA).](Figures/figure2_ensemble_framework.png)


<!-- Using ARROWHEAD, we obtained a total of 8052, 7676, 4670, 2349, and 1031 called TADs  across each resolution respectively (Table 1). From the called TADs, a total of 15468, 14253, 8363, 4224, and 1883 unique TAD boundaries were obtained (Table 1). The resolution-specific approach we developed to create boundary vs non-boundary regions resulted in a total of 535363, 267682, 107073, 53537, and 26768 genomic bins. The class imbalance among the genomic bins was both severe and invariant toward resolution, resulting in 0.03, 0.05, 0.08, 0.08, and 0.07 for each resolution respectively (Table 1). Boundary data summaries were similar for K562 cell line, as well as for PEAKACHU defined boundaries (Supplementary Tables 3-4). -->

| Resolution | Total number of called TADs | Total number of unique TAD boundaries | Total number of genomic bins | Class imbalance |
|:----------:|:---------------------------:|:-------------------------------------:|:----------------------------:|:---------------:|
|5 kb        |8052                         | 15468                                 | 535363                       | 0.03            |
|10 kb	     |7676 	                       | 14253                                 | 267682                       | 0.05            |
|25 kb	     |4670                         | 8363                                  | 107073                       | 0.08            |
|50 kb	     |2349	                       | 4224                                  | 53537                        | 0.08            |
|100 kb	     |1031                         | 1883                                  | 26768                        | 0.07            |
Table: Domain data and class imbalance summaries across resolutions for ARROWHEAD on the GM12878 cell line.

## 5 kb resolution, random under-sampling, and distance-based predictors provided optimal performance for TAD boundary prediction

Expectedly, when using data with class imbalance present, that is, no re-sampling, the models exhibited low balanced accuracies, with very little variability among different resolutions (Figure 3). Similarly, poor performances were found when using ROS as well. However, RUS and SMOTE re-sampling led to drastic improvement in performance, especially at higher resolutions. We found that RUS marginally outperformed SMOTE as the most optimal class balancing technique for all resolutions and predictor types. These results suggest that simple random under-sampling alone can drastically improve TAD boundary region prediction. 

Additionally, we found that at higher resolutions, random forest classifiers built using a distance-type feature space yielded substantially higher balanced accuracies when compared to the peak signal stength, overlap count, and overlap percent predictor types. As with comparing class balancing techniques, this improvement was less evident at lower resolutions. These results were found to be consistent for the K562 cell line (Supplementary Figure 2). Random forest models built on PEAKACHU loop boundaries with SMOTE resampling only marginally outperformed RUS for both GM12878 and K562 cell lines (Supplementary Figure 3). Five kilobase resolution genomic bins was found to result in optimal prediction for both TAD and loop boundaries, on both cell lines. Our results indicate that random under-sampling, distance-type predictors, and Hi-C data at 5kb resolution are the optimal set of data level characteristics for both TAD and chromatin loop boundary region prediction.

<!--
Interestingly, for the optimal re-sampling and predictor type, we found that performances peaked at 10 kb resolution and then degraded as data resolution decreased. The reason for the improved performance between 5 kb and 10 kb data resolutions can be attributed to the sparsity in the interaction frequency matrices at finer resolutions. This sparsity may lead to inaccurate TAD calling and, subsequently, to worse model performance. These results suggest that finer resolution of Hi-C data is not always optimal for accurate TAD calling and prediction, and sparsity of the data needs to be considered. Furthermore, performances tended to be much more variable between chromosomes at lower resolutions, evidenced by the wider error bars for the 50 kb and 100 kb resolutions. These results were found to be consistent for the K562 cell line (Supplementary Figure 2). In summary, our computational framework showed that using random under-sampling, distance-type predictors, and Hi-C data at 10kb resolution yields an optimally performing cell line-specific random forest classification model for TAD boundary prediction.
-->

![**???[needs updating with new framework, takes longest, do last]Models built using distance-type predictors combined with random under-sampling at 5 kb yielded optimal TAD boundary region predictive performance for GM12878.** Line plots comparing the performances of random forest classifiers built using different combinations of data level characteristics. Performances are compared across resolution, within each predictor-type: (A) Signal, (B) OC, (C) OP, and (D) Distance, and between each re-sampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Performances were evaluated using balanced accuracy.](Figures/figure3_performances.png)

## Transcription factor binding sites outperform HM- and BroadHMM-specific models

We hypothesized that the type of genomic annotations may affect the predictive performance. Using the aforementioned optimal settings, we built separate random forest models using BroadHMM chromatin segmentations, histone modifications (HM), and transcription factor binding sites (TFBS). We found that models built on TFBSs outperformed other annotation-specific models, and these results were consistent for loop boundaries, for both cell lines (Figure 4A; Supplementary Figures 4A). These results suggest that TFBS are the primary drivers of TAD and loop boundary formation in both the GM12878 and K562 cell lines. 

## Predictive importances confirmed biological functionality of CTCF, RAD21, SMC3, and ZNF143 at TAD boundaries

We sought to further optimize our TAD boundary region prediction model by performing recursive feature elimination to avoid overfitting and select only the most influential TFBS across all autosomal chromosomes. We were able to obtain near-optimal performance using approximately eight TFBS (Figure 4B; Supplementary Figures 4B). However, given that we trained our model on chromosome-specific data, the most significant annotations were found to vary for each chromosome. To determine transcription factors most important for boundary prediction, we clustered the predictive importances of the top eight significant TFs across chromosomes. We found that, similar to other studies [@sanborn2015chromatin; @weintraub2017yy1; @gonzalez2017znf143], CTCF, RAD21, SMC3, and ZNF143 transcription factors were most predictive of TAD and loop boundaries for all chromosomes (Figure 4C; Supplementary Figures 4C). We further optimized our model by only considering these top informative TFBS when building the random forest, thereby decreasing computational burden while maintaining high predictive performance. In summary, our framework was able to yield the top molecular drivers of the 3D architecture of the human genome, components of the popular loop extrusion model [@hansen2018recent;@fudenberg2016formation;@alipour2012self;@sanborn2015chromatin]. 

<!--
To summarize, the optimally built cell line-specific random forest model under our computational framework was found to be one built using 10 kb resolution data, leveraging distance-type predictors for the CTCF, RAD21, SMC3, and ZNF143 transcription factor binding sites, while implementing random under-sampling to balance TAD and non-TAD boundary defined genomic bins.
-->

![**Assessing top ranked transcription factors in TAD and loop boundary region prediction for GM12878.** (A) Barplots comparing performances of TAD (ARROWHEAD) and loop (PEAKACHU) boundary region prediction models using distance-type specific classes of functional genomic elements as the feature space including histone modifications (HM), cis-regulated elements (BroadHMM), transcription factor binding sites (TFBS), in addition to a model containing all three classes (ALL). (B) Recursive feature elimination (RFE) analysis used to reduce the number of TFBS to include in the optimally built random forest model. Error bars represent standard deviation from the cross-validated balance accuracy across each autosomal chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes as determined from performing RFE.](Figures/figure4_annots_specifc_rfe_importances_arrowhead_peakachu.png)

## *preciseTAD* identifies precise and biologically relevant locations of TAD boundaries

After obtaining our optimally built random forest model, we investigated whether we could leverage our model to alleviate resolution limitations of Hi-C data by predicting TAD boundaries at base pair resolution. Our method, referred to as *preciseTAD*, leverages a user specified binary classifier with a combination of density-based hierarchical clustering and a scalable version of partitioning around mediods (PAM), known as clustering large applications (CLARA) to precisely identify the most representative and biologically meaningful predicted chromosome-specific domain boundaries (see Methods). 

<!--
At this point, we opted to implement preciseTAD using an optimally built whole genome random forest model. That is, a random forest model built on 5 kb defined genomic bins, with random under-sampling, and distance-type predictors for the top informative TFBS (CTCF, RAD21, SMC3, and ZNF143). The reason for this was because of the improved performance that was exhibited for a whole genome random forest compared to aggregated chromosome-specific models (Supplementary Figure 5). When tuning the full genome RF we found that performances converged when *mtry*=2, *ntree*=250, and *nodesize*=??? (Supplementary Figures 6-7). 
-->

When built using ARROWHEAD ground truth boundaries at 5 kb, *preciseTAD* was able to predict a total of ??? and ??? boundaries for GM12878 and K562 cell lines respectively. We reported less predicted TAD boundaries than ARROWHEAD (Table 1, Supplementary Table 3). This is attributed to ARROWHEAD's inherent inflation of called TAD boundaries, especially near fuzzy areas along the diagonal of Hi-C contact matrices (Figure 5A). However, under the framework of *preciseTAD*, any two TADs are inherently conjoined by only one predicted TAD boundary, leading to comparatively less TAD boundaries. Visually speaking, we also found that predicted TAD boundaries identified by *preciseTAD* were much more likely to demarcate high concentrations of inter-chromosomal contacts as seen on a Hi-C contact matrix (Figure 5A). [???add peakachu totals when ready]

Next, we evaluated the biological significance of *preciseTAD* boundaries. Enriched heatmaps indicated that our predicted boundaries were more enriched for CTCF sites than either ARROWHEAD or PEAKACHU boundaries alone (Figure 5 B-C). Likewise, *preciseTAD* boundaries were also more enriched for other molecular drivers of 3D chromatin including RAD21, SMC3, and ZNF143 for both GM12878 and K562 cell lines (Supplementary Figures 8-9). When compared to called TAD or loop boundaries, *preciseTAD* boundaries were found to be statistically significantly closer to the same cell line-specific top ranked TFBS (p=??? and p=??? vs ARROWHEAD and PEAKACHU respectively) (Figure 5D, Supplementary Figure 10-11). We also inspected the affect of resolution on locations of our predicted TAD boundaries. We found that *preciseTAD* boundaries exhibited higher pairwise Jaccard indices between adjacent resolutions than ARROWHEAD boundaries, thus implying stronger invariance toward resolution (Figure 5E).

![**preciseTAD boundaries are enriched in CTCF sites.** (A) Visualizing locations of predicted vs called ARROWHEAD TAD boundaries on GM12878 (chr 11: 33000000-34600000). The upper triangular contact map shows the normalized number of contacts within the specified window using the JUICER toolbox. The blue vertical lines represent boundaries identified by ARROWHEAD, with the corresponding called TADs outlined in blue on the upper diagonal contact matrix above. The black line represents the predicted probability of each base pair being a TAD boundary when validated by our random forest. The green vertical lines represent the *preciseTAD boundary points* (PTBP) predicted by *preciseTAD*, with the corresponding demarcated TADs outlined above in green. Below is a zoomed in portion of the genome showing a cluster of base pairs with high probability representing a single *preciseTAD boundary region (PTBR) colored in yellow. (B) Enriched heatmaps comparing the enrichment of CTCF sites around 5 kb flanked preciseTAD and ARROWHEAD TAD boundaries respectively. (C) Enriched heatmaps comparing the enrichment of CTCF sites around 10 kb flanked preciseTAD and PEAKACHU loop boundaries respectively. (D) $Log_{2}$ genomic distances from boundaries to nearest CTCF site between preciseTAD boundaries and both ARROWHEAD and PEAKACHU boundaries, respectively. (E) A atrix of pairwise Jaccard indices across resolutions shows more invariance toward resolution for *preciseTAD* (upper triangular) compared to ARROWHEAD (lower triangular).???update when ready](Figures/figure5_enrichment_and_log2dist_ctcf.png) 


## *preciseTAD* boundaries are conserved across cell lines

Subsequent comparisons of *preciseTAD* with both ARROWHEAD and PEAKACHU revealed that our predicted boundaries were more conserved between the GM12878 and K562 cell lines. To assess conservation, we evaluated the overlap between cell line specific boundaries after flanking by the resolution used to make predictions. Boundaries involving ARROWHEAD were flanked by 5 kb, while boundaries involving PEAKACHU were flanked by 10 kb. We found that a significant proportion of flanked boundaries predicted by *preciseTAD* overlapped between GM12878 and K562. Specifically, 45%??? and 61%??? of *preciseTAD* boundaries were conserved across cell lines when built on ARROWHEAD and PEAKACHU boundaries respectively (Figure 6A). Additionally, the median jaccard index (MJI) was found to be 0.472 (IQR=0.441-0.503; p=???) and 0.647 (IQR=0.646-0.647; p=???) for *preciseTAD* predicted TAD and loop boundaries (Supplementary Table 5). Comparatively, only 25% (MJI=0.284; IQR=0.268-0.301) and 49% (MJI=0.509; IQR=0.503-0.515) of called/predicted ARROWHEAD and PEAKACHU boundaries were conserved (Figure 6A, Supplementary Table 5).

Expectedly, the majority of flanked TAD boundaries predicted by *preciseTAD* represented a subset of the larger group of predicted loop boundaries. This makes sense as the boundaries that demarcate chromatin loops make up a broader class of boundaries, with TAD boundaries representing the next level in 3D chromatin hierarchy, followed by A/B compartments. For GM12878, 83% of predicted TAD and loop boundaries overlapped with eachother, while only 3% of predicted TAD boundaries were not contained in the set of flank-predicted loop boundaries (Figure 6B). Similarly, for K562, 75% overlap occured between predicted TAD and loop boundaries, with only 4% of predicted TAD boundaries remaining outside of the set of flank-predicted loop boundaries (Figure 6B).

![**preciseTAD boundaries are conserved between the GM12878 and K562 cell lines.** (A) Venn diagrams illustrating the percentage of overlap between GM12878- and K562-specific predicted and called boundaries. (B) Venn diagrams illustrating the nested structure of predicted TAD and loop boundaries identified by *preciseTAD* for the GM12878 and K562 cell lines. Note, predicted and called TAD boundaries were flanked by 5 kb, while predicted and called loop boundaries were flanked by 10 kb prior to calculating overlaps to reflect the resolution that was used to predict each respectively.](Figures/figure6_venn_diagrams.png) 

# Discussion

Here we present *preciseTAD*, a random forest-based approach coupled with spatial clustering (DBSCAN) and scalable partition around medoids (CLARA) that allows for more precise prediction of TAD or loop boundaries. *preciseTAD* leverages a random forest (RF) classification algorithm built on ground truth TAD(loop) boundaries obtained from a TAD(loop) calling tool and ChIP-seq defined genomic annotations as the feature space. We first optimized our RF model by creating a machine learning framework that systematically compared the performances of different combinations of genome binning (resolution), feature engineering procedures, and re-sampling technique. In doing so, we first introduced a new strategy for establishing boundary region prediction data referred to as *shifted binning*, in which the rows of the data matrix were controlled by the data resolution (either 5kb, 10kb, 25kb, 50kb, and 100kb) used to obtain ground truth boundaries, while assuring ground truth bounderies were located central to any given genomic bin. Furthermore, we introduced a new procedure for drawing spatial associations between boundary regions and genomic annotations, referred to as *distance*-type predictors, defined as as the $log_{2}$ genomic distance in base pairs from each genomic bin the nearest ChIP-seq peak region. We benchmarked *preciseTAD* against two boundary calling tools, ARROWHEAD [@durand2016juicebox], an established TAD-caller, and PEAKACHU [@salameh2019supervised] a recently published algorithm for predicting chromatin loops. For a more in depth study, we also applied *preciseTAD* on cell line-specific boundaries and ChIP-seq data from both the GM12878 and K562 cell lines.

Our machine learning framework alone yielded many interesting results. First, we showed that RF models built using *distance*-type predictors out performed models built on previously published feature engineering techniques including overlap counts and overlap percents. Moreover, we were further able to optimize our RF model by introducing random under-sampling, a technique to correct for the severe class imbalance between TAD(loop) and non-TAD(loop) boundary regions. We showed that in our optimized RF, transcription factor binding sites provided an overwhelming amount of signal in the model, indeed out-performing histone modification- and BroadHMM-specific models, contradicting a previous study that found histone marks alone were effective in predicting TAD boundaries [@huang2015predicting]. Likewise, our RF model built using *distance*-type predictors was consistently able to identify the group of the top known molecular drivers of 3D chromatin including CTCF, RAD21, SMC3, and ZNF143 as being highly predictive of TAD(loop) boundary regions across both cell lines. These are known components of the loop extrusion model, an established theory for how loops are made by a ring-shaped adenosine triphosphatase complex called cohesin [@hansen2018recent;@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. This suggested that random forests, when tuned and feature engineered correctly, are highly effective in accurately predicting boundary regions, contradicting a previous study that found simple multiple linear regression out performed random forests in a TAD boundary prediction setting using both human and drosophila Hi-C data [@mourad2016computational]. Plausible reasons for their findings include lack of class balancing prior to modeling and usage of rudimentary feature engineering procedures such as overlap counts and overlap percents. 

Besides balanced accuracy (BA), we investigated five other performance metrics to evaluate the optimal combination of data level characteristics for RF prediction including accuracy, area under the receiver operating characteristic curve (AUC), precision, F1-score, and area under the precision-recall curve (AUPRC) (Supplementary Tables 6-7). The accuracy metric can be artificially inflated by true negatives (TN), the set of genomic bins correctly predicted as not containing a ground truth boundary. AUC captures how a model generally performs accross different thresholds. However, it doesn't place more emphasis on one class over the other, so it does not reflect the minority class well. Precision indicates the rate at which positive predictions are correct and can be artificially deflated by low proportions of true positives (TP), the set of genomic bins correctly predicted as containing a ground truth boundary. While F1-score is a composite metric, it can be very sensitive towards disparate values for precision and recall. Lastly, AUPRC is insensitive toward class imbalance and also omits from its calculation TN values. All of these are important considerations to make when choosing a performance metric. For these reasons we opted to report the balanced accuracy (BA). The BA benifits from incorporating all components of the confusion matrix, while also being sensitive to class imbalance, a necessary characteristic when comparing performances to models built using no data re-sampling.

The performance of *preciseTAD* was found to be empirically better than both ARROWHEAD and PEAKACHU ground truth boundaries. *preciseTAD* provides bp resolution predicted boundaries, thereby alleviating resolution limitations of conventional TAD-callers. We show that, unlike ARROWHEAD, *preciseTAD* does not inflate the number of predicted boundaries, instead, providing only the most biologically meaningful boundaries that demarcate regions of high inter-chromosomal interactions. *preciseTAD* boundaries were found to be more enriched for known architectural transcription factors including CTCF, RAD21, SMC3, and ZNF143. Likewise *preciseTAD* boundaries were found to be more conserved between GM12878 and K562 cell lines, further highlighting their biological significance.

A limitation of *preciseTAD* is that it is dependent on "ground truth" boundaries provided by a user-defined TAD caller. Ideally we would perform simulation studies to benchmark *preciseTAD* against simulated boundaries. While there are existing methods for simulating Hi-C data sets [@forcato2017comparison; @zheng2020freehi], of which boundaries can be extracted from, to date, there are no such methods for simulating ChIP-seq data. However, given *preciseTAD* is applied on chromosome-specific sections of the linear genome, and on multiple cell lines, we feel that aggregating results across chromosomes provides strong empirical evidence of the superiority of *preciseTAD*. Another limitation is that *preciseTAD* does not provide hierarchical boundaries, but instead simply provides the bp coordinate that demarcates TADs(loops). Studies have shown evidence of the hierarchical nature to the 3D architecture of the human genome [@hansen2018recent; @weinreb2016identification; @fraser2015hierarchical]. In such a hierarchy, the boundaries provided by *preciseTAD* can be interpreted as the lowest level of boundaries.

In summary, we have shown that TAD boundary prediction is a multi-faceted problem composed of many data irregularities and characteristics. Simply quering a contact matrix ignores the fundamental roles of known molecular drivers of 3D chromatin. Instead, we propose a method for turning TAD-calling into a supervised machine learning framework built using ChIP-seq annotations. Specifically, we introduce two novel concepts known as *shifted binning* and *distance*-type predictors which we use to build random forest classification models for predicting boundary regions. We then introduce *preciseTAD*, a method for leveraging random forests to predict TAD(loop) boundaries at base pair resolution. Our method is able to bridge the resolution gap between 1D ChIP-seq annotations and 3D Hi-C sequencing data for more precise and biologically meaningful boundary identification. By releasing our method as an open source R package, we hope to provide an efficient and easy-to-use tool to further explore the 3D organization of the human genome.


<!--
The two proteins differ in that YY1 preferentially
occupies interacting enhancers and promoters, whereas CTCF
preferentially occupies sites distal from these regulatory elements that tend to form larger loops and participate in insulation.
Deletion of YY1 binding sites or depletion of YY1 can disrupt
enhancer-promoter contacts and normal gene expression.
Thus, YY1-mediated structuring of enhancer-promoter loops is
analogous to CTCF-mediated structuring of TADs, CTCF contact domains, and insulated neighborhoods. This model of
YY1-mediated structuring of enhancer-promoter loops accounts
for diverse functions reported previously for YY1, including contributions to both gene activation and repression and to gene
dysregulation in cancer. ["YY1 Is a Structural Regulator
of Enhancer-Promoter Loops"]
-->





# Supplementary Legends

**Additional File 1**. An example script for applying the ARROWHEAD tool to in situ Hi-C data (HIC001-HIC018) for resolutions 5kb, 10kb, 25kb, 50kb, 100kb to obtain chromosome-specific TAD boundaries on the GM12878 cell line.

**Supplementary Table 1.** A complete list of the Hi-C data libraries obtained from GEO GSE63525 [@rao20143d] used to obtain the locations of topologically associating domains for ARROWHEAD, as well as the locations for the cell line specific .txt files of loop boundaries obtained by PEAKACHU. [@salameh2019supervised].

**Supplementary Table 2.** A complete list of the genomic annotations used to build the feature space for all downstream models. The GRCh37/hg19 human genome assembly was used. "Genomic Class" - broad category of genomic features, "Element" - names of genomic features, "Cell Type-Specific Source" - download URL specific to the cell line (not all annotations were provided by the same institutions).

**Supplementary Figure 1.** Density curves illustrate the normalization of distances before (red) and after (blue) performing a $log_{2}$ transformation across differing bin sizes including 10 kb, 25 kb, 50 kb, and 100 kb for both the (A) GM12878 and (B) K562 cell lines respectively. Distances are measured as the number of base pairs from the center of a genomic bin to the center of the nearest genomic annotation. Each density curve represents a particular genomic annotation.

**Supplementary Table 3.** Domain boundary and class imbalance summaries from SpectralTAD across resolution for both the
GM12878 and K562 cell lines respectively.

**Supplementary Table 4.** Loop boundary and class imbalance summaries from PEAKACHU 10 kb resolution for both the GM12878 and K562 cell lines respectively.

**Supplementary Figure 2. Models built using distance-type predictors combined with random under-sampling at 5 kb yielded optimal predictive performance for K562.** Line plots comparing the performances of random forest classifiers built using ARROWHEAD boundaries at different combinations of data level characteristics. Performances are compared across resolution, within each predictor-type: (A) Signal (B) OC, (C) OP, and (D) Distance, and between each re-sampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Performances were evaluated using balanced accuracy.

**Supplementary Figure 3. Models built using distance-type predictors combined with SMOTE re-sampling yielded optimal predictive performance for GM12878 K562.** Line plots comparing the performances of random forest classifiers built using using PEAKACHU boundaries at different combinations of data level characteristics for (A) GM12878 and (B) K562. Performances are compared across resampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple), and within each predictor-type: Signal, OC, OP, and Distance. 

**Supplementary Figure 4. Assessing top ranked transcription factors in TAD and loop boundary prediction for K562.** (A) Barplots comparing performances of TAD (ARROWHEAD) and loop (PEAKACHU) boundary prediction models using specific classes of functional genomic elements as the feature space including histone modifications (HM), cis-regulated elements (BroadHMM), transcription factor binding sites (TFBS), in addition to a model containing all three classes (ALL). (B) Recursive feature elimination (RFE) analysis used to reduce the number of TFBS to include in the optimally built random forest model. Error bars represent standard deviation from the cross-validated accuracy across each chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes as determined from performing RFE.

**Supplementary Figure 5. Whole genome random forest model outperform chromosome-specific random forest models across resolutions.** Random forest models built on ARROWHEAD boundaries using 5 kb defined genomic bins, with random under-sampling, and distance-type predictors for the top informative TFBS (CTCF, RAD21, SMC3, and ZNF143) applied to the entire genome (A) exhibit higher balanced accuracies than their chromosome specific counterparts (B) across each resolution respectively.

**Supplementary Figure 6. Whole genome RF performances converge when ntree=250.** Scatter plot of cross-validated balanced accuracies (BA) for random forests built using both ARROWHEAD (circles) and PEAKACHU (triangles) boundaries, with random under-sampling, distance-type predictors (CTCF, RAD21, SMC3, and ZNF143), over the entire binned genome, for GM12878 (red) and K562 (blue) respectively. For the ARROWHEAD model, 5kb bins were used, while 10kb bins were used for the PEAKACHU model. Performances converged after trees of size 250, marked by the dashed line.

**Supplementary Figure 7. Whole genome RF performances converge when nodesize=???.** Scatter plot of cross-validated balanced accuracies (BA) for random forests built using both ARROWHEAD (circles) and PEAKACHU (triangles) boundaries, with random under-sampling, distance-type predictors (CTCF, RAD21, SMC3, and ZNF143), over the entire binned genome, for GM12878 (red) and K562 (blue) respectively. For the ARROWHEAD model, 5kb bins were used, while 10kb bins were used for the PEAKACHU model. Performances converged after nodesize values of ???, marked by the dashed line.

**Supplementary Figure 8. preciseTAD boundaries are more enriched for GM12878-specific SMC3, RAD21, and ZNF143 sites.** Enriched heatmaps comparing preciseTAD boundaries with both (A) ARROWHEAD and (B) PEAKACHU boundaries for the GM12878 cell line.

**Supplementary Figure 9.preciseTAD boundaries are more enriched for K562-specific CTCF, SMC3, RAD21, and ZNF143 sites.** Enriched heatmaps comparing preciseTAD boundaries with both (A) ARROWHEAD and (B) PEAKACHU boundaries for the K562 cell line.

**Supplementary Figure 10. preciseTAD boundaries are spatially closer to GM12878-specific SMC3, RAD21, and ZNF143 sites.** Boxplots comparing the $log_{2}$ genomic distances from each preciseTAD boundary (green) and called boundary (either ARROWHEAD or PEAKACHU; blue) to the nearest SMC3, RAD21, and ZNF143 sites on the linear genome for GM12878.

**Supplementary Figure 11. preciseTAD boundaries are spatially closer to K572-specific CTCF, SMC3, RAD21, and ZNF143 sites.** Boxplots comparing the $log_{2}$ genomic distances from each preciseTAD boundary (green) and called boundary (either ARROWHEAD or PEAKACHU; blue) to the nearest CTCF, SMC3, RAD21, and ZNF143 sites on the linear genome for K562.

**Supplementary Table 5.** Median chromosome-specific Jaccard indices (IQR) for called and *preciseTAD* predicted TAD (ARROWHEAD) and loop (PEAKACHU) boundaries between the GM12878 and K562 cell lines. The p-value is from a nonparametric Wilcoxon-rank sum test. Note, predicted and called *TAD* boundaries were flanked by 5 kb, while predicted and called *loop* boundaries were flanked by 10 kb prior to calculating Jaccard indices to reflect the resolution that was used to predict each respectively.

**Supplementary Table 6.**

**Supplementary Table 7.**

<!--
**Supplementary Figure 3. The proportional importance of optimally built cell line specific models was outweighed by transcription factor binding sites across resolutions.** Stacked barplots illustrate the proportion of total predictive importance in cell line specific random forest classifiers for (A) GM12878 and (B) K562 respectively. Model inputs included the optimal data level characteristics determined by the ensemble framework (SMOTE re-sampling; with distance-type predictors). Predictive importance was averaged for models built on chromosome specific regions of the linear genome. Proportional importance was then calculated according to the annotation group of the feature space including histone modifications (HM), chromatin state segmentations (ChromHMM), and transcription factor binding sites (TFBS).

**Supplementary Figure 5. Venn diagrams illustrating the amount of overlap between predicted and called TAD boundary regions.** For each of the GM12878 (A) and K562 (B), both predicted and called TAD boundaries were flanked other either side by 10 kb and the amount of overlap was enumerated.

**Supplementary Figure 6. Predicted TAD boundaries were centered closer to peak signal regions defined by known architectural transcription factors when compared to called TAD boundaries for K562.** Enrichment
heatmaps illustrate the average signal strength of K562-specific ChIP-seq peak regions defined by (A) SMC3, (B) ZNF143, (C) CTCF, and (D) RAD21 around the centers of 10 kb flanked predicted and called TAD boundaries.

**Supplementary Figure 7. Known architectural transcription factors were statistically significantly closer to predicted TAD boundaries then called TAD boundaries.** Barplots illustrate the log2 distance in basepairs from classes of TAD boundaries to transcription factors including CTCF, RAD21, SMC3, and ZNF143 for the GM12878 (A) and K562 (B) cell lines. TAD boundaries were classified into the 3 classes. Predicted but not called (PBNC) were the set of 10 kb flanked predicted TAD boundaries the did not overlap with 10 kb flanked called TAD boundaries. Called and predicted (CAP) were the set of 10 kb flanked called and predicted TAD boundaries that overlapped with each other. Called but not predicted (CBNP) were the set of 10 kb flanked called TAD boundaries that did not overlap with 10 kb flanked predicted TAD boundaries.
-->

# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
