---
title: "Predictive modeling using genomic annotations"
subtitle: "Pitfalls and Recommendations"
author: "Spiro Stilianoudakis"
date: "August 9, 2018"
header-includes:
    - \usepackage{setspace}\doublespacing
    - \usepackage{amsmath}
output: 
  pdf_document:
    toc: yes
  word_document:
    fig_caption: yes
    fig_height: 3
    fig_width: 2
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# ABSTRACT

###Background 
Chromosome conformation capture sequencing technologies have shown that the three-dimensional (3D) structure of the genome is folded into distinct compartments, known as topologically associated domains (TADs) - units of coordinated gene expression. The location of TAD boundaries is highly conserved, suggesting the presence of epigenomic marks aiding in TAD formation. The ability to predict which epigenomic features are most associated with the TAD boundaries will allow to better understand the regulatory role of the 3D structure of the genome.

Existing methods for predicting associations between genomic elements ignore key characteristics of the data. Specifically, the number of TAD boundaries is much less than the number of other genomic regions, leading to heavily imbalanced classes. Furthermore, most methods utilize direct overlap as a means to quantify the association, while distance, the measure of spatial relationships, remains unaccounted for. Consequently, distances on a genomic scale vary widely, leaving uncertainty how the heavily right-tailed distribution of distance measures will affect the modelâ€™s performance.


###Methods 
We propose a novel data pre-processing pipeline that addresses those shortcomings. A number of classifier performance metrics were assessed, including the F1 measure and Matthew Correlation Coefficient (MCC), and area under the ROC curve (AUROC).

###Results 
Data preprocessing (log2-transformation and standardization) improves the performance of classification algorithms and allows for the ability to more accurately predict which genomic features are most associated with TAD boundaries. 

###Conclusions 
Current methods used to model the epigenomic features associated with TAD boundaries are not robust to handle properties of genomic data. Models applied to unprocessed data can have poor predictive performances. Focusing solely on typical performance assessment metrics, such as AUROC, can mask poor performance of the models; thus, the use of more balanced metrics, such as F1 and MCC, is warranted. Our model results in better performances and more accurate identification of important features associated with the formation of TAD boundaries.

# INTRODUCTION

The advent of various genome-wide sequencing technologies, such as high-throughput conformation capture (notably Hi-C), have revealed how the spatial organization of the human genome may affect several epigenetic functions (Aiden et al.). Analyses have shown that the genome is tightly compacted into distinct compartments.  There exist regions within these compartments that have been shown to be highly conserved and self-interacting, and termed topologically associating domains (TADs) (Dixon et al.). Evidence suggests that regulatory elements and genes tend to interact more frequently within the same TAD (Symmons et al.). This suggests that the boundaries of TADs may play a role in restricting the function of elements such as enhancers, thereby impacting the transcription of genes. Furthermore, changing the 3D structure of the DNA and distrupting these domains can lead to adverse outcomes and diseases like cancer. Therefore, it has become increasingly important to be able to identify the key molecular drivers of the formation of TAD boundaries in order to further our understanding of the human genome.

The mechanisms underlying the formation of TADs are a complex and active area of research. Recently, it has been discovered that insulator sequences have a primary role in orchestrating the topological arrangement of higher-order chromatin architecture (Phillips-Cremins et al). Insulators are multi-faceted regulatory sequences that moderate a variety of genomic processes including activation, repression, and enhancer blocking. Specifically, the insulator binding protein CTCF has been found to be enriched at the boundary sequences of topologically associating domains in human cells and may therefore act as a mediator of long range chromatin contacts (Zuin et al.). Other regulatory elements like DNase I hypersensitive sites, particular histone modifications, and transcription factors such as ZNF274 and YY1 have been more frequently observed at TAD boundaries than in other regions (Hong et al). These have been found to be associated with more open chromatin which alter the accessibility of genes for transcription. The distinct patterns of some of these different proteins and functional elements point toward the opportunity of computational approaches in predicting which epigenomic features are most predictive of the development of TAD boundaries. This will allow us to better understand what leads to their formation. 

Most accepted methods for predicting TAD boundaries consider performing a classification algorithm by binning the genome and labeling the bins as either containing a TAD boundary or not. However, a pervasive issue becomes the introduction of highly imbalanced classes. Due to the sparcity of TADs throughout the genome, most bins will not contain a TAD boundary. Therefore, the minority class of genomic bins with TAD boundaries in them becomes proportionally smaller than that of the majority class of genomic bins with no TAD boundaries. It is commonly agreed that imbalanced datasets adversely impact the performance of the classifiers as the learned model is biased towards the majority class to minimize the overall error rate (Estabrooks et al). Furthermore, it is unclear what sized bins best improves predictive performance or even if the results are stable across differently sized genomic bins. 

Another problem lies in identifying the feature space of subsequent classification algorithms.  A variety of positional characteristics of the genomic features can be used to describes the relationship between TAD boundaries and genomic elements, but it is unclear which is best. The use of distances (in base pairs) from genomic bins to regions of functional genomic elements has largely been ignored. However, these distance characteristics offer a more accurate spatial representation of which genomic features are associated with TAD boundaries. 

In this study we have developed an ensemble framework that aims to address and compare a variety of data irregularities and characteristics that are associated with the prediction of TAD boundaries. We examined the inclusion of three different predictor types as the feature space for predictive modelling. We also examined four resampling techniques aimed at improving class imbalance. Lastly, we examined the effect of three separate variable reduction techniques on model performance. Random forest classification algorithms were performed to investigate which functional genomic features were most associated with the formaton of TAD boundaries. Model performance was evaluated based on a total of four metrics which included test accuracy, area under the ROC curve (AUC), F1 score, and Matthew's Correlation Coefficient (MCC). 

[Insert last paragram for introduction]
 

# METHODS



## DOMAIN DATA

Publicly available topologically associating domain data was obtained from GEO with accession GSE53525 for the GM12878 cell line (Rao et. al).  The domain data was constructed from in situ Hi-C contact matrices at 5kb, 25kb, 50kb, and 100kb resolutions. Identified TAD boundaries were represented by their location in the genome (hg19 human genome assembly), including chromosome, start and end coordinates for chromosomes 1 through 22. The start and end coordinates were concatenated, sorted, and unique coordinates representing the borders demarcating TADs were obtained. 

Figure 1 presents a diagram of how the data was established for modeling purposes. First, the genome was binned into equally sized intervals. This was performed at the same resolution of the contact matrix that the domain data was obtained from. For example, for the domain data obtained from a 5kb resolution contact matrix, the genome was subsequently binned into 5kb intervals. The same processes was repeated for each of the other three resolutions. Next, the boundary coordinates were flanked on either side by a unit corresponding to the resolution of the particular domain data. Again siting the domain data obtained from a 5kb resolution contact matrix, each uniquely identified TAD boundary was flanked on either side by 5kb. Again, the same process was repeated for the other three resolutions. A genomic bin was labeled as having a TAD boundary within it (Y=1) if it overlapped with a particular flanked boundary. Otherwise, the bin was labeled as not containing a boundary (Y=0).

## GENOMIC ANNOTATIONS

Annotation data, in the form of functional genomic element, from ChIP-seq experiments was obtained from the Encyclopedia of DNA Elements (ENCODE) Consortium. The annotations consisted of a variety of different chromatin states using hidden markov models, histone marks, and DNase I hypersensitive sites. Each annotation is represented by their location throughout the genome via chromosomal coordinates. Using each genomic element we created three types of predictors that were used as features in downstream modelling. 

## Ensemble Framework

Figure 2 details the ensemble system proposed and evaluated in this study. For each predictor type chosen, the data, composed of majority and minority classes, was split into a 7:3 ratio of training to testing sets. For the trainig set, a variable reduction technique was performed. The feature space was reduced and a resampling technique was performed. A classification algorithm in the form of a random forest was performed and validated using the test set. 

A total of three increasingly refined predictor types where used to describe the relationship between TAD boundaries and each function genomic element of interest. Two different overlap types and a distance type were considered. Figure 3 presents an illustration of each type and is detailed below:

1. Overlap Counts (OC): For each genomic bin, a total count of each overlapped region, defined by a functional genomic element, with a bin was calculated. If no regions overlapped with a bin, then the predictor was assigned as 0.

2. Overlap Percent (OP): For each genomic bin, the percent of overlap between a bin and a region was calculated as the ratio of feature width (in base pairs) and the total width of each bin. Here, the feature width was defined as the number of base pairs that overlapped between a bin and an elemental region. Note, that the denominator of this ratio was given to be that same for each respective domain data used. That is, for the 5kb resolution domain data, the total width of each bin was 5kb.

3. Distance: For each genomic bin, the distance (in base pairs) from the center of the bin to the center of the nearest region of a respective functional genomic elemnt was calculated.

Variable reduction techniques generally fall into one of three main categories: filter methods, wrapper methods, or embedded methods. Filter methods are generally used as a preprocessing step. The selection of features is independent of any machine learning algorithms. Instead, features are selected on the basis of their scores in various statistical tests for their correlation with the outcome variable. In wrapper methods, we try to use a subset of features and train a model using them. Based on the inferences drawn from the previous model, we decide to add or remove features. Embedded methods combine the qualities of filter and wrapper methods by implementing algorithms that have their own built-in feature selection methods. 

In this paper, we considered one technique from each group. For a filter method, we performed nonparametric two-sample Kolmogorovâ€“Smirnov tests to assess how different the underlying distributions of a respective genomic element, for each predictor type, were between bins with and with a TAD boundary located inside it. The predictors were sorted according the p-value provided from the test. To determine the suitable number of features to reduce to, we plotted the performance of random forest models when iteratively including a new feature (up to the first 20). For a wrapper method, we performed forward selection. Forward selection is an iterative procedure in which began with an intercept only logistic regression model. At each step a feature is added, defined by a particular predictor type, and the AIC of the model is recorded. The method continues as long as there is a suitable reduction in AIC. Lastly, for an embedded method, we performed LASSO regularized regression. The LASSO regression is a penalization technique that uses L1-norm regularization in logistic regression. Here, the coefficients of irrelevant features are reduced to 0. The subsequent features are chosen for downstream modelling. 

To address the class imbalance problem we evaluated four different resampling techniques (in addition to no sampling). For each resampling technique, the majority class is defined as the set of genomic bins that did not include a TAD boundary, while the minority class is defined as the set of bins that included a TAD boundary. The techniques included no sampling, random under-sampling, random over-sampling, SMOTE, and K-Medoids and are disussed in detail below:

1. No Sampling: The full dimension of the training set was used.

2. Random Under-Sampling (RUS): All data points (bins) from the minority class (bins with TAD boundaries) of the training set were used. Data from the majority class (bins without TAD boundaries) are randomly sampled to match the number of minority class instances.  We considered 50 interations, each time randomly sampling from the majority class (without replacement) and performed a separate model. Performance metrics were aggregated by taking the average across all iterations.

3. Random Over-Sampling (ROS): All data points from the majority class of the training set were used. Data from the minority class was randomly sampled (with replacement) to match the size of the majority class instances. Similar to the random under-sampling, we considered 50 iterations and aggregated the results by taking the average.

4. SMOTE (Synthetic Minority Over-Sampling Technique): The method combines both random under-sampling and random over-sampling by creating synthetic observations. The SMOTE algorithm was implemented in R through the DmWR package. The user is able to specify the parameters that control the percent under- and over-sampling. Specific details of the SMOTE algorithm can be found in Chawla et al. For the purposes of this study, the parameters of the algorithm were chosen to create perfectly balanced classes.

5. K-Medoids: This is an unsupervised clustering algorithm. The set of majority classes is isolated and the clustering algorthm is performed. The cluster centers are represented by actual data points. The set of cluster centers are chosen to match the set of minority classes. The final resampled training set is composed of the set of minority classes and the set of cluster centers from the majority classes.

The graphical display of the disparity in classes across the different class-balancing techniques is provided is Figure X. Using set notation the formulation of the ensemble framework can be defined as follows:

Set of Predictor Types:

P = {OC, OP, Distance}

Set of Variable Reduction Techniques:

V = {KS-Test, Forward Selection (AIC), LASSO}

Set of Class-Balancing Methods:

C = {None, RUS, ROS, SMOTE, K-Medoids}

Therefore, given domain data at a specific resolution, a system produced from the ensemble framework is defined as follows:

E = {p, v, c}, where $p \in P$, $v \in V$, and $c \in C$

Once a system was defined, a random forest algorithm was performed using 10-fold cross-validation to reduce bias. Letting $|X|$ denote the cardinality of each set, then for each domain data, we evaluated $|P| \times |V| \times |C|$ different ensemble systems in this paper.

# RESULTS

In this section we provide the details of the comprehensive experiments performed in order to assess each ensemble system. For the task of predicting TAD boundaries using functional genomic elements, we used 5 class balancing techniques, and for each balancing technique we used 3 predictor types and 3 variable reduction techniques. We evaluated each ensemble system using 10-fold cross-validated random forest models.
Thus, we generated 45 (=5x3x3) separate models.

After binning the genome at 5kb, [[25kb, 50kb, and 100kb]] there were a total of X [[, X, X, and X]] genomic bins respectively. Table X provides the data distributions of the different classes across each of the balancing techniques. We see that each class balancing technique considered resulted in perfectly balanced classes. 

We implemented 3 types of variable reduction techniques that fell into 1 of 3 categories: filter methods, wrapper methods, or embedded methods. For the filter method, we performed nonparametric KS-tests for each predictor in the feature space. P-values of the tests were sorted into increasing order. Figure X presents AUCs and F1 scores results by iteratively adding 1 predictor at a time according to the sorted p-values across each resampling technique. We see that the performances begin to stabilize after the inclusion of the top 15 most significant features. Thus, the top 15 features were selected when applying this filter method for all downstream analyses. 

Table X presents the complete results comparing each ensemble system in terms of accuracy, AUC, F1 score, and MCC. We see that the inclusion of distance type predictors significantly improved model performance compared to the other two overlap types. This improvement was apparant across all class balancing techniques. Furthermore, the improvement was irrespective of which variable reduction technique was chosen. Performance metrics remained relatively constant within each model with a specific predictor type and resampling technique, regardless of the variable reduction technique chosen. With respect to the class balancing technique, when isolated to models using distance type predictors, it was found that X preformed best.

Figure X presents the variable importance plot including the top X most predictive functional genomic elements associated with the formation of TAD boundaries. At the top we see X, X, and X
.
.
.







# DISCUSSION
