---
title: Computational Prediction of Boundaries of 3D Genomic Domains in Class Imbalance
  Settings
csl: styles.ref/genomebiology.csl
output:
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
  pdf_document: default
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance problem, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUCROC) curve, Precision-Recall curve (AUPRC), F1, Matthew Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were best type of genomic annotations for predicting TAD boundaries than histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# INTRODUCTION

The advent of various genome-wide sequencing technologies, such as high-throughput conformation capture (notably Hi-C), have revealed how the spatial organization of the human genome may affect genomic regulation [@lieberman2009comprehensive]. Analyses have shown that the genome is folded into distinct compartments termed topologically associating domains (TADs) [@dixon2012topological]. Approximately 60-80% of TADs remain stable across cell types [@Schmitt2016aa], suggesting that the boundaries of TADs may play a role in restricting the function of certain genomic elements such as enhancers, thereby impacting the transcription of genes. Furthermore, studies have shown that changing the 3D structure of the DNA, causing disruptions in these domains, can lead to adverse outcomes and diseases like cancer [@hnisz2016activation; @Flavahan2016; @song2017ctcf]. Therefore, it has become increasingly important to be able to identify the key molecular drivers of the formation of TAD boundaries in order to further our understanding of the human genome.

The distinct patterns of epigenomic marks, binding peaks of different factors, and actively transcribed genes prompt the use of computational approaches to identify which epigenomic features are most predictive of TAD boundaries. Evidence suggests that regulatory elements and genes tend to interact more frequently within the same TAD [@symmons2014functional]. Likewise, emerging research has shown that particular binding sites are enriched at TAD boundaries. It has been established that the zinc finger protein, CTCF, is associated with TAD boundaries and is an architectural protein responsible for the organization of the genome within the nucleus [@ghirlando2016ctcf; @ong2014ctcf]. CTCF along with other transcription factors, including SMC1, SMC3, and RAD21 have been shown to colocalize together near TAD boundaries, leading to the recruitment of a cohesin ring [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Together, these components are hypothesized to be what causes loops in the DNA, contributing to higher order chromatin organization. Other binding proteins have been shown to mediate enhancer-promoter interactions, including the transcription factors YY1 and ZNF143, thereby influencing the delineation of topological domains [@weintraub2017yy1; @gonzalez2017znf143; @heidari2014genome]. Even more so, loci within a domain have been shown to interact more frequently with one another, exhibiting similar patterns of chromatin modifications, and long-range contact patterns [@rao20143d]. However, the growing number of annotated genomic features and the methods used to identify the ones assocaited with TAD boundaries still leaves much to be desired.

Current methods used to predict TAD boundaries rely on the use of classification algorithms to draw associations between the spatial relationship of functional genomic elements and regions that include TAD boundaries. However, this is a multi-faceted probelem, and many methods ignore key characteristics of domain data when building models and reporting performances. We aim to address these characteristics in order to create a more unified approach toward identifyig the genomic annotations most predictive of TAD boundaries.

A major ignored aspect of TAD boundary prediction is the introduction of heavily imbalanced classes resulting from the number of boundaries vs. nonboundary regions. In a classification setting, when one is attempting to discriminate between regions with (minority class) and without (majority class) TAD boundaries, the class imbalance issue will contribute to small imbalance ratios. The imbalance ratio is defined as the ratio of minority to majority classes [@Orriols-Puig2008]. Such heavily imbalanced datasets are known to adversely impact the performance of the classifiers as the learned model is biased towards the majority class to minimize the overall error rate [@lusa2010class; @chen2013influence; @jeni2013facing].

In a study published in 2016 [@mourad2016computational], an L1-regularized multiple logistic regression model was proposed using transcription factor binding sites on GM12878 to discriminate between regions with and without TADs, and to identify the most predictive factors associated with boundaries. However, the class imbalance issue was not addressed. Even worse, when performing cross-validation [@mourad2016computational], used the misclassification error rate of the model to establish the optimal L1-regularization lambda. They also reported ROC (receiver operating characteristic) curves and subsequent AUC (area under the curve) as measures of their model's performance. This is problematic in an imbalanced setting. Such heavily imbalanced datasets are known to adversely impact the performance of the classifiers as the learned model is biased towards the majority class to minimize the overall error rate [@lusa2010class; @chen2013influence; @jeni2013facing]. For AUC, when the prevalence of an event is low--that is, a defined genomic region truly containing a TAD boundary--mean probabilities are biased towards the majority class. This results in the area of interest under an ROC curve to be compressed toward a small corner in the ROC space [@prati2011survey; @wei2013role]. While not directly affecting the value of the AUROC, this can have severe consequences on the interpretability of the model's performance. Consequently, careful consideration must be taken when deciding which performance metric(s) to use for evaluation of predictive models, especially in the case of imbalanced data. 

A better threshold-free measure to use when dealing with imbalanced data is the area under the precision-recall curve (AUPRC).  The interpretation of the AUPRC of a model is similar to that of the AUROC. However, where as the baseline model performance according to the AUROC is given by 0.50, the baseline model performance according to the AUPRC is given by the ratio of true positives over the total number of cases. This enables a more realistic benchmark to evaluate a model on as this ratio will be lower than 0.5 for a test data set with imbalanced classes [@saito2015precision]. Another threshold-specific measure to consider is a composite metric involving all quadrants of the confusion matirix, such as Matthew's Correlation Coefficient (MCC) [@jeni2013facing; @chicco2017ten]. The MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications [@brodersen2010balanced]. A value of positive 1 represents a perfect prediction, while a value of negative 1 represents total disagreement between prediction and observation. A value of 0 indicates no better than random prediction.

In addition to employing suitable metrics for evaluating models validated on imbalanced data, there also exists several data level solutions. Data level solutions mostly constitute as re-sampling techniques, in which random sampling is used to either replicate or remove data points in order to create a more balanced dataset [@chawla2004special; @chawla2009data]. It has been shown that quite simple re-sampling techniques can drastically improve model performance in the face of imbalanced data [@japkowicz2002class; @dubey2014analysis]. 

Two studies published in 2015 and 2017 respectively [@huang2015predicting; @hong2017computational] used re-sampling techniques prior to modeling, and both implemented random under-sampling to balance classes. However, given the large disparity between the number of genomic regions labeled as containing a TAD boundary or not, performing random under-sampling alone is likely to be unstable and introduce bias. Also, it is unclear how other data level solution compare to simple random under-sampling, for instance random over-sampling or a weighted combination of both random under- and over-sampling together. Equally concerning is how the test set is treated in scenarios of re-sampling. It was found that [@huang2015predicting] performed re-sampling prior to splitting the data into training and testing. However, this is problematic because re-sampling prior to splitting can further introduce bias into the testing set, as it is no longer an accurate representation of real-world data [@wei2013role]. 

Furthermore, determing the optimal representation of the relationship between features in the model and genomic regions, known as feature engineering, plays a significant role in TAD prediction. Typical approaches include overlap counts [@hong2017computational] or percents [@mourad2016computational; @huang2015predicting]. Counts refer to the number of times a genomic annotation directly overlaps with a TAD boundary region, e.g., a TAD boundary may overlap with several ChIP-seq peaks for a given transcription factor. Percents refers to the percent of overlap between a genomic annotation and a TAD boundary region, e. g., 60% of a TAD boundary region may overlap with a genomic region annotated as heterochromatin. While direct overlaps represent the most straightforward measure of functional associations, they do not account for indirect relationships shown to be one of the major driving force of genome regulation [??? REFS]. The most prominent example being the enhancer-promoter relationships, where enhancers are frequently located hundreds or thousands of bases away from their target promoters [@whalen2016enhancer]. Using percents offers a more granular approach, but it is unclear how it performs when compared to counts. On the other hand, the use of distances between TAD boundaries and genomic annotations has largely been ignored. However, these distance features have the potential to offer the most accurate spatial representation of which genomic features are associated with TAD boundaries.

Additionally, model performance is heavily contingent upon the resolution of the contact matrix used to identify the location of TADs throughout the genome. It is often unclear from previous studies what resolution contact matrices were used when retrieving TAD boundaries, and how other resolutions would affect their results. A lower resolution contact matrix will result in the identification of larger-sized TADs, and subsequently, less TAD boundary points. This will contribute to smaller sample sizes and could effect predictive performances. To date, no study has compared the results of their predictive models built across multiple data resolutions.

Therefore, in this study, we aimed to develop an ensemble framework composed of different combinations of data driven characteristics used to address the issues presented above. We used our framework to systematically compare performances of random forest classifiers using a composite performance metric, MCC. Our goal was to determine a suitable combination of data re-sampling and feature engineering techniques that both corrected the inherent imbalanced classes and provided the best spatial representation linking genomic annotations to TAD boundaries. Additionally, we used our framework to directly compare the influence of Hi-C contact matrix resolution on model performances. We evaluated Hi-C data for both the GM12878 and K562 cell line obtained at 10 kb, 25 kb, 50 kb, and 100 kb resolution. For each resolution, we examined the inclusion of three different predictor types, including distance as the measure of association between genomic elements. We then tested four different re-sampling techniques aimed at creating balanced classes. This process was performed on chromosome-specific regions on the linear genome. Noninformative features were removed using a feature reduction algorithm and classifiers were built to assess model performance, and investigate which genomic annotations were most predictive of TAD boundaries for a given system from the complete ensemble.



# METHODS

## DOMAIN DATA

Biological replicate Hi-C data for the GM12878 and K562 cell lines were downloaded from GEO GSE63525 [@rao20143d] (Supplementary Table 1). The autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries were obtained from data at 10 kb, 25 kb, 50 kb, and 100 kb resolution using the Arrowhead algorithm [@durand2016juicebox]. The start and end coordinates of all TADs were concatenated, sorted, and unique TAD boundary points were obtained throughput the autosomal chromosomes. TAD boundaries called on Chromosome 9 were omitted from all downstream analyses, for both cell lines, due to the sparcity of the contact matrix at the 10 kb resolution for the K562 cell line.

## GENOMIC ANNOTATION DATA

Genomic annotation data in the form of cell type specific chromatin states, histone marks, and transcription factor binding sites (GRCh37/hg19 human genome assembly) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 2).

## MODEL CONSTRUCTION

To establish the data for modeling, the genome was binned into equally sized intervals defined by the data resolution for the corresponding contact matrix <!--(Figure 1)-->. As an example, for domain data obtained from a 10 kb resolution contact matrix, the genome was binned into 10 kb intervals. For the boundary data, in order to accommodate for the uncertainty in exact boundary location, the boundary points were flanked on both sides by the size of the resolution of the corresponding contact matrix. That is, for 10 kb resolution data, each uniquely identified TAD boundary was flanked on either side by 10 kilobases for a total width around each boundary point of 20 kb. A genomic bin was then labeled as having a TAD boundary within it (Y=1) if it overlapped with a particular flanked boundary. Otherwise, the bin was labeled as not containing a boundary (Y=0). The process was repeated for each subsequent resolution.

### FEATURE ENGINEERING 

Using the sets of genomic annotations, we created three types of predictors that were used as features for subsequent models to analyze their respective spatial relationship with TAD boundaries <!--(Figure 3)-->:

1. **Overlap Counts (OC)**: For each genomic bin, the total number of instances where a genomic bin overlapped with regions assigned to genomic annotations (features) was calculated. In the case of no overlaps, the value was set to 0.
   
2. **Overlap Percent (OP)**: For each genomic bin, the percent of overlap between the width of the region assigned to a genomic annotation and the total width of the bin was calculated. We defined the feature width as the number of bases that overlapped with a particular genomic bin. The percent was then calculated by dividing the feature width by the bin width. If multiple overlaps existed, the feature width was defined as the sum of the total number of bases that overlapped with a particular bin. Note that the bin width remained constant for a given resolution (either 10 kb, 25 kb, 50 kb, or 100 kb). Values for bins with no overlaps were set to be 0.

3. **Log2 Distance**: For each genomic bin, the $log_{2}$ of the distance in bases (refered to hereafter simply as $\textit{distance}$), from the center of the bin to the center of the nearest region defined by a specific genomic annotation was calculated. The log transformation was used as a normalization technique in order to account for the skewness present in the distance densities <!--(Supplementary Figures 1)-->.

### DATA RE-SAMPLING

Class imbalance was characterized by each dataset's imalance ratio, defined as the ratio of minority to majority classes given by the genomic bins. Minority classes were the set of genomic bins that overlapped with flanked TAD boundaries, while majority classes were the set of genomic bins that did not overlap with a flanked TAD boundary.

In order to assess the influence of imbalanced classes, we then evaluated four re-sampling techniques (in addition to no re-sampling), composed of the genomic bin class and data from each predictor type. The re-sampling techniques considered are described in detail below.

1. **No Sampling**: All of the data points from the majority and minority class of the training set were used.

2. **Random Under-Sampling (RUS)**: All of the minority classes from the training set were used. Sampling without replacement was used to obtain the same number of the majority classes. 

3. **Random Over-Sampling (ROS)**: All of the majority classes from the training set were used. Sampling with replacement was used to obtain the same number of the minority classes. 

4. **SMOTE** (Synthetic Minority Over-Sampling Technique): This method incorporates both random under- and over-sampling. Under-sampling is performed without replacement from the majority class, while over-sampling is performed by creating new synthetic observations using the minority class [@chawla2002smote]. The *DMwR* R package was used to perform SMOTE (version 0.4.1). 

### FEATURE REDUCTION

Architectural proteins tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To measure the pairwise correlation of each cell line specific annotation, we used the Jaccard Index. The Jaccard Index measures the ratio of the number of intersecting base pairs between two sets to the number of base pairs in the union, subtracted by the intersection [@favorov2012exploring]. The quantity ranges from 0.0 to 1.0, where 0.0 represents no overlap (no correlation) and 1.0 represent complete overlap (complete correlation).

In order to suitably reduce the feature space while also maintaining the inclusion of possibly colocalized annotations, we implemented an elastic-net regularization technique prior to modelling. The technique using a combination of the *l_{1}* and *l_{2}* penalties. For each elastic-net regularization, 10-fold cross-validation was used to tune the parameters that controlled the penalty terms. <!--The parameters that maximized the mean AUPRC across each cross-fold was chosen.--> All elastic-net algorithms were implemented in R using the *caret* package version 6.0 [@kuhn2012caret]. 

### FINAL MODEL ASSESSMENT AND PERFORMANCE

Once the appropriate predictor type and set of features were determined, a final set of random forest classification models were built to compare model performances between re-sampling techniques and predictor types. <!--To reduce bias due to random dataset generation, 10-fold cross validation was used to tune the number of features to consider at each node and the number of ensemble trees to aggregate.--> The default number of features to consider at each node of the random forest algorithm was set as $\sqrt{p}$, where *p* refers to the number of predictors included in the model. Likewise, the number of ensemble trees to aggregate was set at 500. All random forest algorithms were implemented in R using the *caret* package version 6.0 [@kuhn2012caret]. Once the model was implemented, it was validated on the test set. 

The performance of each ensemble system in the framework was compared using the composite metric, Matthew's Correlation Coefficient. The MCC is defined as follows:

<!--
$$F1 - Score = \frac{2TP}{2TP + FP + FN}$$
-->

$$MCC = \frac{TP \times TN - FP \times FN}{\sqrt{ (TP + FP)(TP+FN)(TN+FP)(TN+FN) }}$$
<!--
$$Balanced \quad Accuracy = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$
-->

<!--
$$Informedness = \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} - 1$$
-->

Here, TP refers to the number of bins correctly identified as containing a TAD boundary (true positives), FP refers to the number of bins incorrectly identified as containing a TAD boundary (false positives), TN refers to the number of bins correctly identified as not containing a TAD boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a TAD boundary (false negatives). Each of these quantities are obtained from the confusion matrix created by validating the model on the testing set. Additional performance metrics were reported in the supplementary materials.

<!--Additional threshold-free metrics where also reported including the AUROC and AUPRC.--> 

<!--
Lastly, for the task of identifying the genomic annotations that were most predictive of the formation of TAD boundaries, we ranked the variable importances associated with each feature of the ensemble system-specific random forest model. To calculate variable importances for each model, the F1-Score on the out-of-bag portion of the data within each tree was recorded. Then the same was done after permuting each predictor variable. The difference between the two AUPRC values was then averaged over all trees, and normalized by the standard error of the differences. Annotations with low rankings correspond to functional epigenomic features that are either positively or negatively enriched near TAD boundaries, and thus, better allow the model to discriminate between genomic bins that overlap with flanked TAD boundaries.
-->

## ENSEMBLE FRAMEWORK

An ensemble framework was then established for the purposes of systematically creating a variety of models that each included different sets of parameters described above <!--(Figure 2)-->. Each system in the framework is assumed to be composed of a model constructed from a particular cell line and data resolution, with a particular predictor type for the feature space. The data sets used for modeling were composed of the full set of genomic bins that either overlapped with a flanked TAD boundary or did not. The data was then split into a 7:3 ratio of training to testing sets. For the task of performance evaluation and prediction, the test set remained constant for all ensemble systems to account for an unbiased comparison. Furthermore, this process was repeated, and performances were aggregated across each autosomal chromosome (omitting chromosome 9) using means and standard deviations. 

# RESULTS

## THE PROPOSED MODEL 

We illustrate a unified approach toward preparing TAD boundary data for predictive modelling purposes from different cell lines and data resolutions (Figure 1). To enable us to build a model that suitably discriminates between regions on the linear genome that are near TAD boundaries and regions that are not, we created a cell line- and resolution-specific response variable that indicated whether a genomic bin of unit width overlapped with a flanked TAD boundary (Y=1), or not (Y=0). The unit width of each bin on the linear genome corresponded to the resolution of the contact matrix on which domains were called. Moreover, to account for uncertainty in boundary location, TAD boundaries were flanked on either side by the same unit length as the genomic bins.

To investigate the relationship between functional genomic annotations and the development of TAD boundaries, we devised 3 different predictor-types that composed the feature space of downstream random forest classification models. These predictor-type specific annotations represented the data matrix, **X** = {$X_{1}$, $X_{2}$, ..., $X_{p}$}, where $p$ is the complete set of functional annotations. Overlap counts was the least granular predictor-type considered, simply defined by the number of annotation-specific regions located within each genomic bin (Figure 2A). Overlap percents was a more refined predictor-type, defined as the total percent overlap between each annotation-specific region and genomic bins (Figure 2B). Lastly, distances, the most granular predictor-type, described the spatial relationship between the center of each genomic bin to the center of the nearest annotation-specific region (Figure 2C).  <!--When measured in basepairs, distances were found to be highly skewed across all resolutions, for both the GM12878 and K562 cell lines (Supplementary Figures 1-2 A-D).--> As a normalization procedure, we performed a $log_{2}$ transformation on the distances prior to modelling (See Methods Section).

## ENSEMBLE COMPARISONS

We then developed an ensemble framework to systematically compare and establish the best set of data level characteristics for predicting TAD boundaries using random forest classifiers built on cell line specific functional genomic annotations (Figure 3). TAD boundaries were called using the Arrowhead algorithm [@durand2016juicebox] from replicate Hi-C data for 2 cell lines $L = \{GM12878, K562\}$, at four different resolutions $R = \{10 kb, 25 kb, 50 kb, 100 kb\}$. For each classifier, we considered 3 separate predictor types as the feature space including $P = \{OC, OP, Distance\}$. To assess the severity of class imbalance, four re-sampling techniques were implemented prior to modelling, including $B = \{None, RUS, ROS, SMOTE\}$. Each combination of the above parameters was implemented on chromosome specific sections of the linear genome, and thus, was repeated repeated 21 times (number of autosomal chromosomes, excluding chromosome 9 [??? In Methods, not here, you should explain why]), $C = \{1, 2, ..., 22\}$. A system produced from the ensemble framework was defined as: $S = \{c, r, p, b\}$, where $l \in L$, $r \in R$, $p \in P$, $b \in B$, and $c \in C$. Letting $|X|$ denote the cardinality of each set, we evaluated $|L| \times |R| \times |P| \times |B|$ different ensemble systems, totaling 2016 ($2 \times 4 \times 3 \times 4 \times 21$) separate models. Performances were then aggregated across each of the 21 chromosomes (omitting chromosome 9; see Methods). We report our main results for the GM12878 cell line and have included results from the K562 cell line in the supplementary materials. [??? Currently, Methods and the results are redundant. Look how others present their models in the results section, and how methods section is structured in such cases. Specifically, some key definitions, like OC, OP, Distance should be solidly introduced in the results section, while log2 status of distance and supplementary figure 1 description are for methods]

## IMBALANCE RATIOS REMAINED SEVERE AND INVARIANT TOWARD RESOLUTION

[???probably better to report medians here]
Expectedly, the average number of genomic bins, TADs called, and TAD boundaries increased with the increased resolution of cell-line specific Hi-C data (Table 1; Supplementary Tables 3). The correlation of each cell-line specific quantity with resolution was found to be significant, as evidenced by large intraclass correllation coefficients (ICC) (Table 2; Supplementary Tables 4). However, the class imbalance problem was found to be much more consistant across resolutions in both cell lines. Likewise, this relationship was not significantly correlated with resoltuion for either cell line, indicating that there exists a severe disparity between classes at all resolutions of Hi-C data.

[??? Very bad and boring. You should tell the STORY, not the report. E.g., "Expectedly, the average number of genomic bins, TADs, and TAD boundaries increased with the increased resolution of Hi-C data (Table 1). However, the imbalance ratio, defined as ..., remained severe (~XXX) at all resolutions (Table 1). In summary, the problem of class imbalance exists at all resolutions of Hi-C data."; !!! editted]

## HIGH JACCARD INDICES IMPLY COLOCALIZATION OF MANY GROUP-SPECIFC ANNOTATIONS 

Using the Jaccard index, we were able to determine which group-specific annotations were enriched together on the linear genome. As could be expected, for GM12878, the class of 14 annotations segmenting the chromation into different states exhibited very low pairwise correlations, with a Jaccard index range of (0-0.001) (Table 3). However, specific sets of pairwise histone modifications and transcription factor binding sites (TFBS) were found to be highly enriched. The range of Jaccard indices for the 10 histone modifications was found to be (0.022-0.539). Specifically, H3k27ac, H3k4me1, H3k4me2, and H3k4me3 were found to be highly correlated, high levels of which have been detected surrounding transcription start sites [@barski2007high]. For the 79 TFBS, the range of Jaccard indices was (<0.001-0.425). Here, known mediators of loop extrusion components were found to have high pairwise correlations including CTCF, RAD21, and SMC3 [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Similar profiles of Jaccard indices were exhibited across the sets of annotations for the K562 cell line (Supplementary Table 4). <!--The high Jaccard indices among select sets of genomic annotations gave precedent for the use of an elastic-net regularization algorithm in order to suitably reduce the feature space while retaining possibly correlated annotations that could be predictive of TAD boundary formation.-->

## CLASS IMBALANCE SEVERELY IMPACTS MODEL PERFORMANCE

We then went on the evaluate the effect of class imbalance on model performance. In order to set a baseline in performance, we first evaluated the least granular model, that is one implemented using OC predictor types, without data re-sampling. We found that models built on data with fully imbalanced classes exhibited poor MCCs across all resolutions (Figure 5A). We gradually increased the granularity by assessing models built on OP predictor types, while still not performing any data resampling. Here, the results were similar to that of OC types (Figure 5B). However, when including distance-type predictors, models were found to exhibit a marked improvement in MCC compared to the other two types (Figure 5C). Results were found to be consistently poor across resolutions and predictor types, when implemented on the K562 cell line (Supplementary Figure 3). Furthermore, we found that conventional metrics of performance (i.e. accuracy, sensitivity and specificity, and AUC) were less sensitive for models built on imbalanced data, regardless of cell line, resolution, or predictor-type (Supplementary Tables 6-7). [??? Need to know why PRCs give different results.]

## DATA RE-SAMPLING COUPLED WITH THE INCLUSION OF A DISTANCE-TYPE FEATURE SPACE SUBSTANTIALLY IMPROVED MODEL PERFORMANCE

When models were built on randomly oversampled (ROS) data for GM12878, we found an increase in MCC compared to no resampling, at higher resolutions. The biggest improvement in performance was seen at the 10 kb resolution in each of the 3 predictor types, with a downward trend as resolutions decreased. Again, distance-type predictors were found to offer the highest MCCs for ROS data, across all resolutions. Similar results were obtained when applying ROS to the K562 cell line (Supplementary Figure 3). 

Further improvement in performance was found when models were implemented on both randomly undersampled (RUS) and SMOTE resampled data, especially between models with OC and OP predictor types. When using RUS or SMOTE, implemented on distance type predictors, we found a clear improvement compared to the other predictor types. However, there was less of a separation between ROS and RUS techniques within the ensemble of models using distance type predictors. 

Overall, SMOTE resampled data was found to yield the best MCCs among all classes of models developed and compared in the ensemble framework. Here, distance type predictors were found to again yield the highest MCC values across resolutions. It should be noted that, SMOTE re-sampled data, with the inclusion of distance-type predictors was also found to be the best set of characteristics when applied to the K562 cell line (Supplementary Figure 3). Moreover, for these sets of data characteristics, we again found that there was a degredation in model performance as data resolution decreased. Results indicated that performances exhibited much more variability when built on lower resolutions, specifically at the 100 kb level. These results imply that a best performing model, one that is able to best discriminate between genomic bins that are near versus far away from TAD boundaries, is one that is built on 10 kb resolution data, with distance-type predictors, and incorporates both random under- and over-sampling. 

## MODELS IMPLEMENTED ON TRANSCRIPTION FACTOR BINDING SITES, USING OPTIMAL PARAMETERS, OUTPERFORM OTHER ANNOTATION-SPECIFIC MODELS

After determining the optimal set of data level characteristics, we then aimed to compare the performances of annotation-specifc models. In doing so, we inspected the proportion of predictive importance that was explained by each of the 3 groups of annotations (chromatin states, histone modifications, and transcription factor binding sites), across resolution. We found that for random forest models implemented on distance-type predictors, using SMOTE re-sampling, the majority of the proportional predictive importance was obtained by the cell-line specific transcription factor binding sites, and was consistent across resolutions (Figure 5; Supplementary Figure 4). 

We further assessed the predictive capabilities of the specific sets of annotations by comparing the performances of annotation-specific random forest models implemented on the optimal set of characteristics. We found that models built simply on cell line specific TFBS out performed other cell line and annotation-specific models at all resolutions (Figure 6; Supplementary Figure 5). Additionally, when compared to the optimal model identified from the ensemble framework, one that included all 3 sets of annotations, the TFBS-specific models performed similarly at the 10 kb and 25 kb resolutions. Interestingly though, at the 50 kb and 100 kb resolutions, the TFBS-specific models out performed models built on the combined sets of annotations, indicating that, at lower resolution, the inclusion of other types of annotations contributed to additional noise rather than signal. 

## PREDICTIVE IMPORTANCES CONFIRMED BIOLOGICAL FUNCTIONALITY OF KNOWN TRANSCRIPTION FACTOR BINDING SITES

We then sought to determine which of the transcription factor binding sites were most influential in predicting TAD boundaries. We found that the cell line specific binding sites associated with SMC3, CTCF, RAD21, ZNF143, and YY1 were given consistently high predictive importance, across all resolutions (Figure 7; Supplementary Figure 6). The SMC3, CTCF, and RAD21 binding sites are known factors associated with the cohesin protein complex, which is hypothesized to be what creates loops in the DNA  [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Likwise, ZNF143 and YY1 are known structural regulators of enhancer-promoter interactions that also contribute to looping of the linear genome [@weintraub2017yy1; @gonzalez2017znf143]. With the identification of these factors known to be enriched at TAD boundaries, we are able to confirm the validity of our model building process. Additionally, we were also able to identify some lesser known cell line specific transcription factors as predictive of TAD boundaries including the MYC-associated zing-finger protein (MAZ), jun-D proto-oncogene (JUND), and ELF1. These transcription factor binding sites tended to be given higher predictive importance at lower resolution. In fact, at lower resolutions, the predictive importance was found to be much more evenly dispersed across the top transcription factor binding sites, than for higher resolutions. This is again suggestive of the fact that the spatial relationship between functional annotations, specifically transcription factor binding sites, and TAD boundaries changes at different Hi-C data resolutions.

<!--
## DISTANCE PROFILES HIGHLIGHT THE RELATIONSHIP BETWEEN KEY TFBS AND TAD BOUNDARIES

When evaluating the 10 kb data resolution, on GM12878, the distance profiles (on a $log_{2}$ scale) were found to be smaller for top predictive TFBS annotations in relation to genomic bins that overlapped with flanked TAD boundaries, when compared to bins that did not overlap with a flanked TAD boundary (Figure 7). At the same resolution and cell line, direct comparisons of quantiles obtained from the distribution of distances for the transcription factors SMC3a, ZNF143, CTCF, and RAD21, led to the the conclusion that these TFBS were more often located near TAD boundaries on the linear genome, as can be seen from the shifted quantile differences (Figure 7). Similar distance profiles for the same TFBS were obtained when assessed at lower resolutions, across both cell lines (Supplementary Figures 12-17 A-D).
-->

# DISCUSSION

In this study we developed an ensemble framework to systematically compare model performances for different combinations of re-sampling and feature engineering techniques aimed at determining a consensus approach to predicting TAD boundary formation using genomic annotations. We provide direct comparisons of cell line specific random forest models built on a variety of different data dependent characteristics, including 4 contact matrix resolutions ($\{\textit{10 kb}, \textit{25 kb}, \textit{50 kb}, \textit{100 kb\}}$), 3 predictor-types ($\{\textit{OC}, \textit{OP}, \textit{Distance}\}$), and 4 re-sampling techniques ($\{\textit{None}, \textit{RUS}, \textit{ROS}, \textit{SMOTE}\}$), for both the GM12878 and K562 cell lines. 

The experiments derived from our ensemble framework provide a useful guide toward the exploration of chromatin organization, while highlighting the impact of both class imbalance and feature engineering in TAD boundary prediction. Using the framework we demonstrated that random forest models built on imbalanced data exhibited poor performance, regardless of cell line, resolution, or predictor type. In our framework, we opted to leave the test set as a completely untouched holdout set, only performing the re-sampling, and subsequent learning, on the training set. This was done so as to incur the least amount of bias and so that all conclusions could be made soley between model characteristics. As a result, we showed that Matthew's Correlation Coefficient was much more sensitive to the imbalance present in the test set, compared to more popular metrics like AUROC, which attain artificially high values by correctly predicting majority class samples. When comparing re-sampling techniques, we found that SMOTE re-sampling outperformed models built on both simple ROS or RUS, implying that, in fact, a combination of both random under-sampling and random over-sampling is preferred in TAD boundary prediction. Furthermore, the most improvement was seen when combining SMOTE resampling on distance-type features when building random forest classifiers. These two characteristics exhibited consistenly better performance results across both the GM12878 and K562 cell lines. This confirmed our hypothesis that the most granular and spatially informative representation of the relationship between TAD boundaries and annotations resulted in the best performing model.

When applying SMOTE re-sampling, with distance-type predictors, on transcription factor binding sites alone, we found that random forest classifiers performed better when compared to other annotation-specific models, including histone modifications and chromatin state segmentations. Likewise the performance of TFBS specific models was comparable to models with all annotations combined, across both cell line and resolution. This suggests that the majority of the key molecular drivers aiding in the formation of TAD boundaryies are transctiption factors, and that modifications at histone tails such as acetylation or methylation play less of a role.

We also observed that, while the best set of data characteristics were consistent across resolutions, performances of classifiers built on said characteristics tended to decline as resolution decreased. For both cell lines, top performance was achieved at the 10 kb data resolution, while the worst performance was obtained at the 50 kb and 100 kb resolution for the GM12878 and K562 cell line respectively. This suggests that TAD boundary prediction is intimately related to the resolution that boundaries are called on. This result can be attributed to one of two factors in the scheme of our framework. First, at higher resolutions, by model construction genomic bins were smaller, thereby resulting in more accurate profiles for the genomic annotations. Secondly, because we are mainly concerned with TAD boundaries, and not the heirarchical structure of domains as a whole, the sparcity of contact matrices at higher resolutions does not impair boundary prediction. Conversely, lower resolutions imply larger bin windows and lead to less specificity in boundary location on the linear genome, resulting in worse performances. Given the trend in performances across resolution, future research is necessary in determining the optimal resolution to use when predicting TAD boundaries. 

Given the formulation of our framework, we were not able to make direct comparisons between cell lines. This was due to the fact that we opted to use cell line specific annotations, and there existed some annotations that were available on the K562 cell line that were not available to GM12878, and vice versa. However, we were able to confirm that final models built on the optimal set of data characteristics identified consistent and well known proteins related to the 3D structure of chromatin, for both cell lines. Our results indicated that the zinc finger protein CTCF, as well components of the cohesin complex, SMC3 and RAD21, were found to be the key molecular drivers most predictive of TAD boundary formation, consistent across cell line and resolution. In addition, models derived from our framework recognized known mediaters of long range enhancer-promoter interactions, the ZNF143 and YY1 transcription factors, as being highly predictive of TAD boundaries. 

In summary, we have shown that TAD boundary prediction is a multi-faceted problem composed of many data irregularities and characteristics that previous model have failed to address. We have shown that our proposed framework suitably addresses each of these components, while also providing the optimal considerations to make in an effort to build consensus classifiers used to predict TAD boundary formation from functional genomic annotations. Given the stability of our results, this opens up the possibility of widening our framework to include a broader set of classifiers, cell lines, and Hi-C contact matix resolutions that will inevitably assist us in gaining a better understanding of the mechanisms associated with the architecture of the human genome.



# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
