---
title: Computational Prediction of Boundaries of 3D Genomic Domains in Class Imbalance
  Settings
csl: styles.ref/genomebiology.csl
output:
  pdf_document: default
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# ABSTRACT (Spiro version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance problem, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUCROC) curve, Precision-Recall curve (AUPRC), F1, Matthews Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were best type of genomic annotations for predicting TAD boundaries than histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# ABSTRACT (MD version)

**Background.** Topologically Associated Domains (TADs) represent fundamental building blocks of the three-dimensional (3D) genome organization and regulation. Several epigenetic marks, such as CTCF, are highly enriched at TAD boundaries, suggesting a strong epigenetic component of TAD boundary formation. Several methods have been developed to leverage the growing amount of (epi)genomic annotation data for predicting TAD boundaries; however, they overlooked key characteristics of the data, such as spatial associations between a genomic annotation and a boundary. Furthermore, the number of regions containing TAD boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods.** To investigate the problem of spatial genomic relationships and class imbalance, we developed an ensemble framework for optimizing prediction of TAD boundaries. We utilized several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) coupled with various metrics for classifier performance evaluation (Area Under Receiver Operating Characteristics (AUROC) curve, Precision-Recall curve (AUPRC), F1, Matthews Correlation Coefficient (MCC), among others) to select the best predictive model in class imbalance settings.

**Results.** Using different feature engineering and resampling strategies, coupled with elastic net-based feature selection, we systematically compared the performance of models predicting the location of TAD boundaries (Gm12878 and K562 Hi-C data at 10kb to 100kb resolutions). Precision-Recall curves were found to be best suited for evaluating model performance in class imbalance settings. Random undersampling and SMOTE-resampled data performed best for optimal parameter selection. Transcription Factor Binding Sites (TFBSs) were the best type of genomic annotations for predicting TAD boundaries when directly compared with histone modifications and chromatin states; we identified well-known CTCF, SMC3, RAD21, and several novel TFBSs as best predictors of TAD boundaries. Importantly, spatial relationships (distance between TAD boundaries and genomic annotations) were best predictors of TAD boundaries as compared with commonly used overlap counts or percent overlap measures.

**Conclusions.** Our results outline strategies for predictive modeling of 3D genomic domains using 1D genome annotation data. The accurate modeling of TAD boundaries will improve our understanding of how epigenetics shapes the 3D structure of the genome.

# INTRODUCTION

The advent of various genome-wide sequencing technologies, such as high-throughput conformation capture (notably Hi-C), have revealed the spatial organization of the human genome  [@lieberman2009comprehensive]. Analyses have shown that the genome is folded into distinct compartments termed topologically associating domains (TADs) [@dixon2012topological]. Given their high degree of conservation between cell types and species, it has been proposed that TADs represent the fundamental building blocks of organization, giving the genome its 3D structure [@rocha2015breaking]. Studies have shown that changing the 3D structure of the DNA, causing disruptions in these domains, can lead to adverse outcomes and diseases like cancer [@hnisz2016activation; @Flavahan2016; @song2017ctcf]. Therefore, it has become increasingly important to be able to identify the key molecular drivers of the formation of TAD boundaries in order to further our understanding of the human genome. It has been thought that the collection of particular functional genomic elements may contribute to the formation of TADs. Namely, the CCCTC-binding factor (CTCF) has been shown to be enriched at TAD boundaries [@ghirlando2016ctcf]. Additional transcription factors including SMC1, SMC3, and RAD21 have also been shown to colocalize near TAD boundaries [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Likewise, chromatin has been shown to exhibit certain modifications at TAD boundaries, evidenced by the frequent observation of the H3K36me3 histone mark [@dixon2012topological]. However, the growing number of publically available annotated genomic elements and there collinearity pose a significant problem for establishing methods robust enough to determine the most important epigenomic factors that contribute to TAD boudandary formation. Moreover, the plasticity of functional annotations make TAD boundary prediction a cell line specific problem [@wang2012widespread]

## PROBLEMS AND LIMITATIONS

The distinct patterns in epigenomic marks, binding peaks of different factors, and actively transcribed genes prompt the use of computational approaches toward TAD boundary prediction  <!--[@dixon2015chromatin; @di2017novo]-->. Currently, classification models have been used to draw associations between these functional annotations and genomic regions that include TAD boundaries. However, many key characteristics of 3D genomic data remain unexplored, potentially limiting the performance of predictive modeling. 

First, the use of different levels of association between TAD boundary regions and genomic features as predictors remain underexplored. These associations are termed predictor-types and make up the feature space of classifiers. Counts of overlaps (**overlap count**, OC) is the most commonly used type of association as it simply counts the total number of peaks defining a genomic annotation that are inside a given TAD boundary region. Percents of overlaps (**overlap percent**, OP) represents a more precise measure of association as it considers the proportion of overlap betwee a genomic annotation and a boundary region. However, the distance between boundary regions and genomic annotations are currently unexplored. This association allows for quantifying the spatial relationship between genomic elements and boundary regions. We hypothesize that the inclusion of this type of association will correspond with more accurate identification of functional genomic elements association with TAD boundary formation.

Secondly, the minority of TAD boundaries as compared with the rest of the genome creates proportionally small TAD boundary regions versus non-TAD boundary regions. These boundary regions make up the minority class, while the non-boundary regions make up the remaining majority class. The large disparity is sizes between minority and majority classes creates imbalanced data, as can be characterized by the *proportion of imbalance* (PI). Here, PI is defined as twice the ratio of the set minority class observations to the sum of both sets of minority and majority class observations. That is, the PI is bounded by 0 and 1. A dataset is considered balanced if PI $\approx$ 1, and imbalanced if PI$<<$1. Classifiers implemented on heavily imbalanced datasets are known to exhibit poor performances as the learned model is biased towards the majority class [@lusa2010class; @chen2013influence; @jeni2013facing]. We hypothesize that the implementation of novel data resampling techniques will significantly improve model performance. Futhermore, the choice of performance metric can mask the poor performance of models implemented on imbalanced data.  It has been shown that conventional metrics such as AUC and accuracy are insensitive to class imbalance, creating artificially inflated values influenced by the majority class [@prati2011survey; @wei2013role]. Therefore, we hypothesize that a more composite metric such as Matthews Correlation Coefficient (MCC), might provide more sensitive assessment of model performance, given that it penalizes for both false positives and false negatives.

Thirdly, TAD boundary location is known to be heavily influenced by the resolution of the subsequent Hi-C interaction matrix [@sexton2012three]. Higher resolutions require a larger number of mapped reads, thereby creating larger data matrices, increasing computation time. However, the increased resolution available in newer datasets has highlighted the existence of a hierarchical structure of TADs inside other TADs [@pal2019hi]. It is unclear how varying levels of resolution affect the association between functional genomic annotations and TAD boundary regions.
<!--[link these two paragraphs]-->
<!--[Paragraph about disconnect between Hi-C and epigenomic data] 
Despite the success of Hi-C technologies in revealing the high-order levels of genome organization (chromosome territories, A/B compartments, TADs, sub-TADs, chromatin loops), resolution of Hi-C data (~750kb [??? REF, Bonev 2017?]) remains well below resolution of epigenomic assays (hundredths of bases for ChIP-seq). One of the key missing pieces to connect 3D genome architecture to transcription has been the inability of Hi-C to attain the resolution necessary to discern gene-level features of chromatin. This resolution gap between 1D and 3D genomic maps, as well as a paucity of chromatin features below the level of TADs, has significantly limited our understanding of epigenomic regulation of gene expression and its potential link to chromatin architecture. Although computational [??? REFS?](Zhang et al., “HiCPlus.”, Carron et al., “Boost-HiC.”) and technological [@Hsieh:2019aa][@Ma:1998aa][@Belaghzal2017] methods for enhancing resolution of Hi-C data started to appear, they [??? remains to write about their pitfalls]. One aim of this study is to predict the location of TAD boundaries on a scale comparable to epigenomic assays. For this, we developed a model trained on a relatively coarse Hi-C data for predicting 3D features (TAD boundaries) from 1D data (epigenomic datasets).-->


## PREVIOUS METHODS

<!--[still revising]-->
Recently, an L1-regularized multiple logistic regression model, called *HiCfeat*, was developed to predict TAD boundary regions using a collection of 69 transcription factor binding sites on the GM12878 cell line [@mourad2016computational]. 

Another method used an L2-based regularized linear model to discriminate between what was termed as "boundary-" and "TAD-segments" of the hESC cell line [@hong2017computational].

A method devolped in 2015 used a tree based model to predict TAD boundary regions using 9 different cell line specific histone modifications, as well as consensus CTCF peaks, across 8 cell lines [@huang2015predicting]. 

## OUR APPROACH

In this study, we have developed a computational framework aimed at determining the optimal set of data level characteristics necessary for optimal prediction of TAD boundaries and identification of influential genomic annotations. These include the best combination of predictor type, resampling technique, and data resolution. We utilize random forest classification models built on Hi-C data from two cell lines, including GM12878 and K562, obtained at four different resolutions, including 5 kb 10 kb, 25 kb, 50 kb, and 100 kb. For each cell line and resolution, we compared three predictor types, including counts, percents, and distances. Additionally, we compared three re-sampling techniques, including random under-sampling, random over-sampling, and a weighted combination of both referred to as Synthetic Minority Over-sampling Technique (SMOTE). Model performances were evaluated and compared using Matthews Correlation Coefficient (MCC). We performed 10-fold cross-validation for parameter tuning and to guard against over-fitting. We found that the optimally built cell line specific model is one implemented using distance type predictors, with SMOTE resampling, at the 10-kb data resolution. We also determined that simply including transcription factor binding sites was necessary for optimal performance. The inclusion of chromatin states and histone modifications was found to induce noise into the model. Specifically, we found that the transcription factors CTCF, ZNF143, RAD21, and SMC3 exhibited high predictive importances, confirming there involvement in TAD boundary formation. We then aimed to evaluate the cell line specificity of TAD boundaries by validating our optimally built models across cell lines. We found that models built and validated on the same cell line outperformed models that were validated across cell lines, confirming the cell line specific nature and plasticity of TAD boundaries and functional genomic annotations respectively. <!--[add conclusion from results of predicting finer bins]-->

# METHODS

## DOMAIN DATA

Biological replicate Hi-C data for the GM12878 and K562 cell lines were downloaded from GEO GSE63525 [@rao20143d] (Supplementary Table 1). The autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries were obtained from data at 10 kb, 25 kb, 50 kb, and 100 kb resolution using the Arrowhead algorithm [@durand2016juicebox]. TAD boundaries called on chromosome 9 were omitted from all downstream analyses, for all cell lines and resolutions, due to the sparsity of the contact matrix at the 10 kb resolution for the K562 cell line.

## GENOMIC ANNOTATION DATA

Genomic annotation data in the form of cell type specific chromatin states, histone marks, and transcription factor binding sites (GRCh37/hg19 human genome assembly) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 2). Functional annotations were represented by chromosome-specific regions on the linear genome obtained by ChIP-seq peak calling algorithms.

## DEVELOPING FRAMEWORK TO DETERMINE OPTIMAL DATA LEVEL CHARACTERISTICS FOR TAD BOUNDARY PREDICTION

### CLASSIFICATION

To establish the data for classification, the genome was binned into equally sized intervals defined by the data resolution of the corresponding contact matrix. As an example, for domain data obtained from a 100 kb resolution contact matrix, the genome was binned into 100 kb intervals. In order to accommodate for the uncertainty in exact boundary location, the boundary points were flanked on both sides by the size of the resolution of the corresponding contact matrix. That is, for 100 kb resolution data, each uniquely identified TAD boundary was flanked on either side by 100 kilobases for a total width around each boundary point of 200 kb. A genomic bin was then labeled as having a TAD boundary within it ($Y = 1$) if it overlapped with a particular flanked boundary. Otherwise, the bin was labeled as not containing a boundary ($Y = 0$).

### FEATURE ENGINEERING 

Using the sets of cell line specific genomic annotations, we created three types of predictors that were used to describe the relationship between functional annotations and genomic bins. They are described in detail below:

1. **Overlap Counts (OC)**: For each genomic bin, the total number of instances where a bin overlapped with regions assigned to genomic annotations (features) was calculated. 
   
2. **Overlap Percent (OP)**: For each genomic bin, the percent of overlap between the width of the region assigned to a genomic annotation and the total width of the bin was calculated. We defined the feature width as the number of bases that overlapped with a particular genomic bin. If multiple overlaps existed, the feature width was defined as the sum of the total number of bases that overlapped with a particular bin. [??? Unclear.]

3. **Log2 Distance**: For each genomic bin, the $log_{2}$ of the distance in bases (refered to hereafter as $\textit{distance}$) [??? Learn Markdown!], from the center of the bin to the center of the nearest region defined by a specific genomic annotation was calculated. 

### TRAINING AND TESTING

The data set used for modeling was composed of the response vector **Y**, and the predictor-type specific data matrix, **X** = {$X_1$, $X_2$, ..., $X_p$}, where $p$ was the complete set of functional annotations. Each composed data set was specific to cell line and resolution. We then split the data into a 7:3 ratio of training to testing sets. For the task of performance evaluation and prediction, the test set remained constant to account for an unbiased comparison between data level characteristics.

### DATA RE-SAMPLING

Class imbalance was characterized by each dataset's *proportion of imbalance* (PI) defined as the ratio of minority class observations to the sum of minority and majority class observations:

$$PI = 2\dfrac{C_{min}}{C_{min} + C_{maj}}$$

where $C_{min}$ denotes the set of genomic bins that overlap with a flanked TAD boundary and $C_{maj}$ denotes the complementary set of genomic bins that did not overlap with a flanked TAD boundary. Here, PI is bounded by 0 and 1, with heavily imbalanced classes approaching 0.

In order to assess the influence of imbalanced classes, we then evaluated four re-sampling techniques (in addition to no re-sampling). They are described in detail below.

1. **No Sampling**: All of the data points from the majority and minority classes of the training set were used.

2. **Random Under-Sampling (RUS)**: All of the data points from the minority class of the training set were used. Sampling without replacement was used to obtain the same number of data points from the majority class. 

3. **Random Over-Sampling (ROS)**: All of the data points from the majority class of the training set were used. Sampling with replacement was used to obtain the same number of data points from the minority class. 

4. **SMOTE** (Synthetic Minority Over-Sampling Technique): This method incorporates both random under- and over-sampling [@chawla2002smote]. Under-sampling is performed without replacement from the majority class. Over-sampling is performed by creating new synthetic observations using the minority class. The minority class is over-sampled by taking each minority class sample and introducing synthetic examples along the feature vectors joining any/all of the k=5 (default) minority class nearest neighbors. The SMOTE algorithm can have many variations in the weightinh of over- and under-sampling. In this study we restricted the algorithm to 100% over-sampling and 200% under-sampling. That is, 1 new synthetic observation was to be created for each existing minority class data point, while 2 data points were simultaneously randomly selected from the majority class. This combination results in perfectly balanced classes. The *DMwR* R package was used to perform SMOTE (version 0.4.1). 

<!--
### FEATURE REDUCTION

Architectural proteins tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To measure the pairwise correlation of each cell line specific annotation, we used the Jaccard Index. The Jaccard Index measures the ratio of the number of intersecting base pairs between two sets to the number of base pairs in the union, subtracted by the intersection [@favorov2012exploring]. The quantity ranges from 0.0 to 1.0, where 0.0 represents no overlap (no correlation) and 1.0 represent complete overlap (complete correlation). [?? Jaccard is a distance measure, I would avoid using the term correlation here ??]

In order to suitably reduce the feature space while also maintaining the inclusion of possibly colocalized annotations, we implemented an elastic-net regularization technique prior to modelling. The technique using a combination of the *l_{1}* and *l_{2}* penalties. For each elastic-net regularization, 10-fold cross-validation was used to tune the parameters that controlled the penalty terms. All elastic-net algorithms were implemented in R using the *caret* package version 6.0 [@kuhn2012caret]. 
-->

### FINAL MODEL ASSESSMENT AND PERFORMANCE

Random forest classification models were built to compare model performances between combinations of predictor types and re-sampling techniques. To reduce bias due to random dataset generation, 10-fold cross validation was used to tune the number of annotations, *p*, to consider at each node. The number of trees that were aggregated for each random forest model was set to the default, n=500. All random forest algorithms were implemented in R using the *caret* package version 6.0 [@kuhn2012caret]. Once the model was implemented, it was validated on the test set. The performance of each model was compared using the composite metric, Matthews Correlation Coefficient (MCC). The MCC is, in essence, the correlation coefficient between the observed and predicted binary classifications [@brodersen2010balanced]. A value of 1 represents a perfect prediction, while a value of -1 represents total disagreement between prediction and observation. A value of 0 indicates no better than random prediction. The MCC is defined as

<!--
$$F1 - Score = \frac{2TP}{2TP + FP + FN}$$
-->

$$
MCC = \frac{TP \times TN - FP \times FN}{\sqrt{ (TP + FP)(TP+FN)(TN+FP)(TN+FN) }}
$$
<!--
$$Balanced \quad Accuracy = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$
-->

<!--
$$Informedness = \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} - 1$$
-->

where TP refers to the number of bins correctly identified as containing a TAD boundary (true positives), FP refers to the number of bins incorrectly identified as containing a TAD boundary (false positives), TN refers to the number of bins correctly identified as not containing a TAD boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a TAD boundary (false negatives). Each of these quantities are obtained from the confusion matrix created by validating the model on the testing set. <!--Additional performance metrics including AUROC, accuracy, sensitivity, and specificity were [??? are?] reported in the supplementary materials. [??? Where is it?]-->

## USING OPTIMALLY BUILT CLASSIFIER TO PREDICT BOUNDARIES AT FINER RESOLUTION

[needs editting]

# RESULTS

## DEVELOPING A RESOLUTION-SPECIFIC APPROACH FOR DATA CONSTRUCTION

We analyzed publically available replicate Hi-C data sets from two cell lines including GM12878 and K562 (Supplementary Table 1). Topologically associating domains (TADs) were subsequently called using the ARROWHEAD pipeline [@durand2016juicer] at 4 resolutions: 5 kb 10 kb, 25 kb, 50 kb, and 100 kb. To account for the varying levels of data resolution, we developed a resolution-specific approach toward constructing TAD versus non-TAD boundary regions on the linear genome (See Methods), labeled as a response vector **Y** (Figure 1). <!--First, we binned the linear genome according to the data resolution. We then flanked the boundaries on either side by the same number of basepairs as the previously constructed genomic bins.--> TAD boundary regions were defined as regions on the linear genome where a genomic bin overlapped with a flanked TAD boundary ($Y = 1$), while non-TAD boundary regions were defined as the regions that did not overlap with a flanked TAD boundary ($Y = 0$).  

![**An illustration of the proposed approach for resolution-specific data construction used for TAD boundary prediction.** Resolution-specific TADs were called and unique boundaries were identified, denoted by the blue dots. The yellow regions denote the flanked boundary regions according to the Hi-C data resolution (i.e. 100 kb resolution corresponded to a 100 kb flank for a total boundary region of 200 kb). The linear genome was subsequently binned into resolution-specific intervals. The response vector **Y** was defined as a variable that indicated whether a genomic bin overlapped with a flanked TAD boundary ($Y = 1$) or not ($Y = 0$).](Figures/figure1_model_construction_V7.png)

## TAD BOUNDARY REGIONS EXHIBIT CELL-LINE SPECIFICITY

<!--## TOTAL NUMBER OF TAD BOUNDARY REGIONS DECREASE AS Hi-C DATA RESOLUTION DECREASES-->
Expectedly, the total number of TAD boundary regions increased with the increased resolution of Hi-C data irrespectively of cell line (Figure 2A). At higher resolutions, GM12878 was found to contain more TADs, and thus more TAD boundary regions compared to K562. This can largely be attributed to the increased sequencing depth of the GM12878 library. However, this difference was found to dissipate as resolution decreased, resulting in much more similar numbers of TAD boundary regions between the two cell lines at the 50 kb and 100 kb resolutions respectively. 

We sought to compare the cell-line specificity of TAD boundary regions. Between the two cell lines, a significant number of TAD boundary regions were found to be specific to one of the two cell lines (Figure 2B). It was found that X%, 45%, 38%, 37%, and 36% of TAD boundary regions were specific to the GM12878 cell line, while X%, 27%, 27%, 29%, and 33% of TAD boundary regions were specific to the K562 cell line across each of the 5 kb, 10 kb, 25 kb, 50 kb, and 100 kb resolutions respectively. Given the plasticity in boundary location, we opted to build cell line specific models for predicting TAD boundary regions.

## PROPORTIONAL IMBALANCE REMAINED SEVERE IRRESPECTIVELY OF Hi-C DATA RESOLUTION 

We then investigated the proportional class imbalance (PI) featured at each resolution and among each cell line. The PI was defined as twice the ratio of TAD boundary regions ($Y=1$) versus the sum of both TAD and non-TAD boundary regions. The proportional imbalance was found to be invariant to resolution, remaining consistently low for both cell lines, with very little variability between chromosomes (Figure 3C). Such low proportions of imbalance gave precedence for the implementation of re-sampling prior to classification. 

![**Cell-line specific boundary summaries.** (A) Barplots illustrating that the total number of TAD boundary regions decreases as Hi-C data resolution decreases. (B) Venn diagrams showing that TAD boundary regions exhibit cell-line specificity across resolution. (C) Boxplots indicating that the proportion of class imbalance remained severe irrespective of resolution.](Figures/bounds_and_pis_v2.png)

<!--
**Table 1. Sparse TAD boundaries throughout the linear genome create high imbalance ratios irrespective of resolution.** 

| Resolution (Bin Width) | Average Number of Genomic Bins (SD) | Average Number of TAD Boundaries (SD) | Average Imbalance Ratio (SD) |
|------------------------|-------------------------------------|---------------------------------------|------------------------------|
| 10 kb                  | 12214.1 (6068.5)                    | 1254.6 (598.8)                        | 8.7 (1.2)                    |  
| 25 kb                  | 4892.7 (2431.2)                     | 714.0 (358.4)                         | 5.9 (0.5)                    |
| 50 kb	                 | 2450.9 (1217.6)                     | 364.4 (184.5)                         | 5.8 (0.5)                    |
| 100 kb                 | 1227.5 (609.8)                      | 165.3 (90.9)                          | 6.8 (1.1)                    |
-->

<!--**Table 2. Inter-Resolution ICCs.** For GM12878, two-way intraclass correlation coefficients between resolutions for the following quantities: Number of genomic bins, number of TADs called, number of unique TAD boundaries, and imbalance ratio. The ICC was computed across 21 of the autosomal chromosomes. P-values indicate the level of significance of the correlation between data resolution and each quantity.[??? What is ICC? !!!editted]

| Quantity                 | Inter-Resolution | 95% CI          | P-Value  |
|--------------------------|------------------|-----------------|----------|
| Number of Genomic Bins   | 0.464            | [0.244, 0.690]  | < 0.0001 |
| Number of TADs Called    | 0.627            | [0.426, 0.802]  | < 0.0001 |
| Number of TAD Boundaries | 0.611            | [0.407, 0.791]  | < 0.0001 |
| Imbalanced Ratio         | 0.114            | [-0.058, 0.372] | 0.109    |
-->

## DEVELOPING THREE TYPES OF PREDICTORS DESCRIBING THE ASSOCIATIONS BETWEEN TAD BOUNDARY REGIONS AND EPIGENOMIC FEATURES 

A total of 78 cell type specific functional genomic annotations derived from ChIP-Seq experiments (See Methods) for the GM12878 and K562 cell line were used to build the feature space of downstream predictive models. We constructed three different predictor types for each functional genomic annotation. Counts of overlaps (**overlap count**, OC) were defined as the total number of peak regions that overlapped within a given genomic bin (Figure 3A). Percents of overlaps (**overlap percent**, OP) were defined as the proportion of a bin overlapped by a peak region (Figure 3B). **Distance** was defined as log2 transformed number of base pairs between between the centers of bins and regions defining genomic annotations (Figure 3C). A log2-transformation of distance was used to account for the skewness present in the distributions of distances (Supplementary Figures 1-2). Each predictor type provides a granular way to draw associations between functional genomic annotations and the spatial structure of the genome. 

![**Three types of associations between TAD boundaries and genomic annotations.** (A) _Overlap Counts_ (OC) consider the total number of genomic features that overlapped with each genomic bin. (B) _Overlap Percent_ (OP) was calculated by dividing the sum of all feature widths within each specific bin by the total bin width. (C) _Distance_ ($log_{2}$ transformed) was calculated by measuring the distance (in bases) from the center of each genomic bin to the center of the nearest region defining a genomic annotation. ](Figures/figure2_predictor_types.png)

## ENSEMBLE FRAMEWORK & COMPARISONS

In order to identify the optimal combination of data resolution, predictor type, and re-sampling technique necessary for TAD boundary prediction, we created an ensemble framework to systematically compare performances of different model inputs (Figure 4). Each system in the framework was assumed to be composed of a random forest classifier constructed from a particular cell line (*L = {GM12878, K562}*). Within each cell line we evaluated 5 data resolutions (*R = {5 kb, 10 kb, 25 kb, 50 kb, 100 kb}*). We then considered 3 separate predictor types as the feature space (*P = {OC, OP, Distance}*). To assess the severity of class imbalance, 4 re-sampling techniques were implemented prior to modelling (*B = {None, RUS, ROS, SMOTE}*). Each combination of the above cell line specific parameters was implemented on chromosome specific sections of the linear genome (omitting chromosome 9). Letting $|X|$ denote the cardinality [??? TBD] of each set, we evaluated *|L|* $\times$ *|R|* $\times$ *|P|* $\times$ *|B|* different ensemble systems, totaling 2520 (*2* $\times$ *5* $\times$ *3* $\times$ *4* $\times$ *21*) separate models. For each combination of inputs, we performed 10-fold cross-validation to tune for the number of features to consider at each node of the random forest classifier. Each optimally tuned model was then validated on the same separate cell line specific testing set for unbiased comparisons. Performances were evaluated using MCCs and aggregated across each chromosome by averaging. <!--We report our main results for the GM12878 cell line and have included results from the K562 cell line in the supplementary materials.-->

![**Ensemble Framework.** A diagram of the ensemble framework detailing the combination of inputs used in the model building pipeline. Cell line specific TAD boundary data was extracted from a Hi-C produced contact matrix of a given resolution. A variety of functional genomic annotations were used to establish the feature space. Each subsequent random forest classification model was trained on an isolated predictor type in conjunction with a particular data re-sampling technique to account for class imbalance. Ten-fold cross validation was performed to tune for the number of predictor-type specific genomic features to consider at each node. Each model was then validated on a testing set and performances were evaluated using MCC.](Figures/figure3_ensemble_framework_V3.png)

## SMOTE RE-SAMPLING COUPLED WITH A DISTANCE-TYPE FEATURE SPACE PROVIDED OPTIMAL CELL LINE SPECIFIC MODEL PERFORMANCE

<!--## CLASS IMBALANCE SEVERELY IMPACTS PERFORMANCE FOR CELL LINE SPECIFIC MODELS-->

We first evaluated the least granular cell line specific random forest classifiers [??? Unclear; !!!discuss in meeting], those were ones implemented using OC predictor types, without data re-sampling. We found that these models exhibited poor MCCs across all resolutions (Figure 5 A-B (red)). We then assessed models built on OP predictor types, while still not performing any data resampling. Performances were similar to that of OC types, with no visible improvement across resolution, for either cell line (Figure 5 C-D (red)). This indicated that the added complexity of using overlap percents from the functional genomic annotations did not affect predictive performance. However, when cell line specific classifiers were implemented using distance-type predictors and no re-sampling, we found that there was a significant improvement in MCC compared to the other two predictor types (Figure 5 E-F (red)). We then further explored the impact of data re-sampling on model performance, across cell line and resolution. For cell line specific random forest classifiers built on balanced data, we found that SMOTE re-sampling provided the optimal MCC, consistently outperforming models built on ROS or RUS respectively, across both predictor type and resolution (Figure 5). Specifically, the most improvement was seen when combining SMOTE re-sampling with distance-type predictors for both the GM12878 and K562 cell lines. 

## MODEL PERFORMANCE PEAKED AT 10 KB AND DECREASED AS Hi-C DATA RESOLUTION DECREASED

Interestingly, we found that performance peaked at 10 kb resolution and then degraded as data resolution decreased (Figure 5). The reason for the improved performance between 5 kb and 10 kb data resolutions can be attributed to the added sparcity in the interaction frequency matrices at finer resolutions. This sparcity corresponds to inaccurate TAD calling which leads to worse model performance. The difference in performance between 5 kb and 10 kb resolutions for K562 was found to be slightly steaper, indicating that this issue of sparcity was more severe at finer resolutions for K562. Additionally, performances tended to be much more variable between chromosomes at lower resolutions. Moreover, this decreasing trend in performance was much more evident under the optimal combination of data level characteristics (SMOTE resampling with distance type predictors). These results imply that there is an ideal resolution in TAD boundary prediction, and that that an optimally performing cell line specific random forest classifier <!--, one that is able to best discriminate between genomic bins that are near versus far away from TAD boundaries,--> is one that is built on 10 kb resolution data, with distance-type predictors, and incorporates both random under- and over-sampling.

![**SMOTE re-sampling coupled with distance-type predictors yielded optimal MCCs across resolution for cell line specific models.** Performances of random forest classification models built on each of the 3 predictor types including (A-B) OC, (C-D) OP, and (E-F) Distance, and across both the GM12878 and K562 cell lines. Within each cell line, predictor type, and resolution, chromosome-specific training data was re-sampled using different techniques including no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and *Synthetic Minority Over Sampling TEchnique* (SMOTE; purple). Performances were evaluated using MCCs and aggregated by averaging across each chromosome, with error bars representing 1 standard deviation away from the mean.](Figures/model_performances_v3.png)

## MODELS IMPLEMENTED ON TRANSCRIPTION FACTOR BINDING SITES, USING OPTIMAL PARAMETERS, OUTPERFORM OTHER ANNOTATION-SPECIFIC MODELS

After determining the optimal set of data level characteristics, we then aimed to compare the performances of annotation-specifc models [??? Never defined, what is this?; !!!defined above; should i mention again?], within each cell line and across resolution. In doing so, we inspected the proportion of predictive importance that was explained by each of the 3 groups of annotations (chromatin states, histone modifications, and transcription factor binding sites). We found that for cell line specific random forest models implemented on distance-type predictors, using SMOTE re-sampling, the majority of the proportional predictive importance was obtained by the functional genomic annotations classified as transcription factor binding sites (TFBS) (Figure 6 A). This was consistent across resolutions, suggesting that TFBS were the primary drivers of TAD boundary formation in both the GM12878 and K562 cell lines.

We further assessed the predictive capabilities of the specific sets of annotations by comparing the performances of annotation-specific random forest models implemented on the optimal set of characteristics. We found that cell line specific models built simply on TFBS outperformed other cell line and annotation-specific models at all resolutions (Figure 6 B). Additionally, when compared to models built on all 3 sets of functional genomic annotations, the TFBS-specific models performed similarly at the 10 kb and 25 kb resolutions. Interestingly though, at the 50 kb and 100 kb resolutions, the TFBS-specific models out performed models built on the combined sets of annotations for both cell lines. This indicated that, at lower resolutions, the inclusion of other types of annotations contributed to additional noise rather than signal. 

<!--
![**The proportional importance of optimally built cell line specific models was outweighed by transcription factor binding sites across resolutions.** Stacked barplots illustrate the proportion of total predictive importance in cell line specific random forest classifiers for (A) GM12878 and (B) K562 respectively. Model inputs included the optimal data level characteristics determined by the ensemble framework (SMOTE re-sampling; with distance-type predictors). Predictive importance was averaged for models built on chromosome specific regions of the linear genome. Proportional importance was then calculated according to the annotation group of the feature space including chromatin state segmentations (CSS), histone modifications (HM), and transcription factor binding sites (TFBS).](Figures/proportion_of_importances_v2.png)
-->

<!--
![**Optimally built TFBS-specific models outperformed other annotation-specific models across resolution for GM12878** Performances were compared for random forest models built using specific groups of annotations as the feature space including: histone modifications (HM; red), chromatin state segmentations (CSS; green), transcription factor binding sites (TFBS; blue), and a feature space containing all 3 types combined (combined; purple). Each bar represents a model ran through the ensemble framework, with re-sampling technique and predictor-type characteristics fixed to SMOTE and distance-types. The error bars represent $\pm$ 1 standard deviation from the mean MCC across each chromosome.](Figures/comparing_sets_of_annotations_v2.png)
-->

## PREDICTIVE IMPORTANCES CONFIRMED BIOLOGICAL FUNCTIONALITY OF KNOWN TRANSCRIPTION FACTOR BINDING SITES

We then sought to determine which of the transcription factor binding sites were most influential in predicting TAD boundaries. We found that the cell line specific binding sites associated with SMC3, CTCF, RAD21, and ZNF143 were given consistently high predictive importance, across all resolutions (Figure 6 C). The SMC3, CTCF, and RAD21 binding sites are known factors associated with the cohesin protein complex, which is hypothesized to be what creates loops in the DNA  [@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self]. Likewise, ZNF143 is a known structural regulators of enhancer-promoter interactions that also contribute to looping of the linear genome [@weintraub2017yy1; @gonzalez2017znf143]. With the identification of these factors known to be enriched at TAD boundaries, we are able to confirm the validity of our model building process. Additionally, we were also able to identify some lesser known cell line specific transcription factors as predictive of TAD boundaries including the MYC-associated zing-finger protein (MAZ), jun-D proto-oncogene (JUND), and ELF1. These transcription factor binding sites tended to be given higher predictive importance at lower resolutions. [??? Where do you show it?] In fact, at lower resolutions, the predictive importance was found to be much more evenly dispersed across the top transcription factor binding sites, than for higher resolutions. This is again suggestive of the fact that the spatial relationship between functional annotations, specifically transcription factor binding sites, and TAD boundaries changes at different Hi-C data resolutions.


<!--
![**Top 20 most predictive transcription factor binding sites sorted by 10 kb for GM12878.** A heatmap illustrating the importance of the key cell line specific transcription factors for optimally built random forest models across resolution. Importances were sorted according to the 10 kb resolution and rescaled to values between 0 and 100. Transcription factors with higher values of predictive importance are denoted in dark red.](Figures/figure7_variable_importance_heatmaps_tfbs_gm12878.png)
-->

![**Assessing optimally built cell line specific random forest classifiers for predicting TAD boundary formation.** (A) Comparing the proportion of total predictive importance carried by each group of functional genomic annotation across resolution. Predictive importance was averaged for models built on chromosome specific regions of the linear genome. Proportional importance was then calculated according to the annotation group of the feature space including chromatin state segmentations (CSS), histone modifications (HM), and transcription factor binding sites (TFBS). (B) Comparing models using specific groups of functional genomic annotation as the feature space in addition to a model with a feature space containing all 3 types combined (combined; purple). Error bars represent $\pm$ 1 standard deviation from the mean MCC across each chromosomal section of the linear genome. (C) Heatmaps illustrating the importance of the top 20 cell line specific transcription factors across resolution. Importances were sorted according to the row-wise average across resolution and rescaled to values between 0 and 100. Transcription factors with higher values of predictive importance are denoted in dark red and dark blue for the GM12878 and K562 cell line respectively.](Figures/prop_imps_a_comparing_annots_b_var_imps_c_v2.png)


## INFLUENTIAL TFBS TENDED TO BE CLOSER TO BINS THAT OVERLAPPED WITH FLANKED TAD BOUNDARIES THAN THOSE THAT DID NOT

We then investigated the spatial patterns of the most predictive TFBS around TAD boudaries by characterizing the distance profiles between the classes of genomic bins. For GM12878, at 10 kb resolution, we found that SMC3, CTCF, RAD21, ZNF143, and YY1sc281 were all statistically significantly closer to bins that overlapped with flanked TAD boudaries than those that did not (Figure 8). [??? not there] Although still statistically significantly different, the distance profiles between genomic class were found to be much more similar at lower resolutions (Supplementary Figure 9). For the K562 cell line, we reported the distance profiles of the same transcription factor binding sites, excluding the YY1sc281, due to its unavailability in the UCSC Genome Browser. These cell line specific distance profiles were again found to be significantly different between genomic class, at all resolutions (Supplementary Figure 10). 

![**Figure 8. Top predictive transcription factor binding sites were significantly closer to bins that overlapped with 10 kb flanked TAD boundaries on GM12878** Boxplots illustrating the distribution of distances (on $log_{2}$ scale) around 10 kb regions on the linear genome that either overlapped with a flanked TAD boundary (Yes) or did not (No), for the most influential cell line specific TFBS including SMC3, CTCF, RAD21, ZNF143, and YY1sc281. P-values obtained from Wilcoxon Rank Sum tests are provided assessing the significance [$NS$ = p > 0.05; $*$ = 0.01 < p $\le$ 0.05; $**$ = 0.001 < p $\le$ 0.01; $***$ = 0.0001 < p $\le$ 0.001; $****$ = p $\le$ 0.00001].](Figures/figure8_distance_boxplots_key_tfbs_gm12878_10kb.png)

# DISCUSSION

In this study we developed an ensemble framework to systematically compare model performances for different combinations of re-sampling and feature engineering techniques aimed at determining a consensus approach to predicting TAD boundary formation using genomic annotations. We provide direct comparisons of cell line specific random forest models built on a variety of different data dependent characteristics, including 4 contact matrix resolutions ($\{\textit{10 kb}, \textit{25 kb}, \textit{50 kb}, \textit{100 kb\}}$), 3 predictor-types ($\{\textit{OC}, \textit{OP}, \textit{Distance}\}$), and 4 re-sampling techniques ($\{\textit{None}, \textit{RUS}, \textit{ROS}, \textit{SMOTE}\}$), for both the GM12878 and K562 cell lines. [??? Learn Markdown] [??? What's the summary of your work?]

The experiments derived from our ensemble framework provide a useful guide toward the exploration of chromatin organization, while highlighting the impact of both class imbalance and feature engineering in TAD boundary prediction. Using the framework we demonstrated that random forest models built on imbalanced data exhibited poor performance, regardless of cell line, resolution, or predictor type. In our framework, [??? Too much "framework" mentioning] we opted to leave the test set as a completely untouched holdout set, only performing the re-sampling, and subsequent learning, on the training set. This was done so as to incur the least amount of bias and so that all conclusions could be made soley between model characteristics. As a result, we showed that Matthews Correlation Coefficient was much more sensitive to the imbalance present in the test set, compared to more popular metrics like AUROC, which attain artificially high values by correctly predicting majority class samples [??? I don't see results for it]. When comparing re-sampling techniques, we found that SMOTE re-sampling outperformed models built on both simple ROS or RUS, implying that, in fact, a combination of both random under-sampling and random over-sampling is preferred in TAD boundary prediction. Furthermore, the most improvement [??? In what?] was seen when combining SMOTE resampling on distance-type features when building random forest classifiers. These two characteristics exhibited consistently better performance results across both the GM12878 and K562 cell lines. This confirmed our hypothesis that the most granular and spatially informative representation of the relationship between TAD boundaries and annotations resulted in the best performing model.

When applying SMOTE re-sampling, with distance-type predictors, on transcription factor binding sites alone, we found that random forest classifiers performed better when compared to other annotation-specific models, including histone modifications and chromatin state segmentations. Likewise the performance of TFBS specific models was comparable to models with all annotations combined, across both cell line and resolution. This suggests that the majority of the key molecular drivers aiding in the formation of TAD boundaryies are transctiption factors, and that modifications at histone tails such as acetylation or methylation play less of a role.

We also observed that, while the best set of data characteristics were consistent across resolutions, performances of classifiers built on said characteristics tended to decline as resolution decreased. For both cell lines, top performance was achieved at the 10 kb data resolution, while the worst performance was obtained at the 50 kb and 100 kb resolution for the GM12878 and K562 cell line respectively. This suggests that TAD boundary prediction is intimately related to the resolution that boundaries are called on. This result can be attributed to one of two factors in the scheme of our framework. First, at higher resolutions, by model construction genomic bins were smaller, thereby resulting in more accurate profiles for the genomic annotations [?? Don't understand this sentence ??]. Secondly, because we are mainly concerned with TAD boundaries, and not the hierarchical structure of domains as a whole, the sparsity of contact matrices at higher resolutions does not impair boundary prediction [?? Doesn't this contradict your results about low resolution causing worse boundary prediction ??]. Conversely, lower resolutions imply larger bin windows and lead to less specificity in boundary location on the linear genome, resulting in worse performances. Given the trend in performances across resolution, future research is necessary in determining the optimal resolution to use when predicting TAD boundaries [?? Might want to qualify the statement in results about 10kb data being best by saying that it only applies to your data ??]. [??? This paragraph is a mess]

Given the formulation of our framework, we were not able to make direct comparisons between cell lines. This was due to the fact that we opted to use cell line specific annotations, and there existed some annotations that were available on the K562 cell line that were not available to GM12878, and vice versa. [??? Did they provide additional advantage?] However, we were able to confirm that final models built on the optimal set of data characteristics identified consistent and well known proteins related to the 3D structure of chromatin, for both cell lines. Our results indicated that the zinc finger protein CTCF, as well components of the cohesin complex, SMC3 and RAD21, were found to be the key molecular drivers most predictive of TAD boundary formation, consistent across cell line and resolution. In addition, models derived from our framework recognized known mediaters of long range enhancer-promoter interactions, the ZNF143 and YY1 transcription factors, as being highly predictive of TAD boundaries. 

In summary, we have shown that TAD boundary prediction is a multi-faceted problem composed of many data irregularities and characteristics that previous model have failed to address. We have shown that our proposed framework suitably addresses each of these components, while also providing the optimal considerations to make in an effort to build consensus classifiers used to predict TAD boundary formation [?? formation or location. Formation implies some temporal element ??] from functional genomic annotations. Given the stability of our results, this opens up the possibility of widening our framework to include a broader set of classifiers, cell lines, and Hi-C contact matix resolutions that will inevitably assist us in gaining a better understanding of the mechanisms associated with the architecture of the human genome.



# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References
