---
title: "preciseTAD: A machine learning framework for precise TAD boundary prediction at base level resolution"
csl: styles.ref/genomebiology.csl
output:
  word_document:
    reference_docx: styles.doc/NIH_grant_style.docx
  pdf_document: 
bibliography: References/references.bib
--- 

Spiro C. Stilianoudakis^1^ (stilianoudasc@mymail.vcu.edu), Mikhail G. Dozmorov^1\*^ (mikhail.dozmorov@vcuhealth.org)

^1^ Dept. of Biostatistics, Virginia Commonwealth University, Richmond, VA, 23298, USA  
^\*^ To whom correspondence should be addressed: Virginia Commonwealth University, Richmond, VA, 23298, 804-827-2055, mikhail.dozmorov@vcuhealth.org

# Abstract

**Background** High-throughput chromosome conformation capture technology (Hi-C) has revealed extensive DNA looping and folding into discrete 3D domains. Among these are Topologically Associated Domains (TADs) and chromatin loops, known to play critical roles in cellular processes like gene regulation and cell differentiation. The relatively low resolution of Hi-C data (regions of several kilobases in size) prevents precise mapping of domain boundaries. TAD and loop boundaries have been characterized by the presence of CTCF and members of cohesin complex binding sites, and other genomic annotations. They are typically defined at high resolution (~100bp), suggesting they can inform the precise location of domain boundaries. Several methods have been developed to leverage genome annotation  data for predicting TAD/loop boundaries; however, they overlooked key characteristics of the data, such as spatial associations between an annotation and a boundary. Furthermore, the number of regions containing boundaries is much smaller than the rest of the genome, creating a class imbalance problem. 

**Methods** To improve the detection of domain boundaries, we developed *preciseTAD*, a random forest-based framework that leverages high-resolution genome annotation data coupled with density-based clustering and scalable partitioning techniques. We investigated several resampling techniques (random over- and undersampling, Synthetic Minority Over-sampling TEchnique (SMOTE)) to address class imbalance. We benchmarked our method against the Arrowhead domain caller and a novel chromatin loop prediction algorithm, Peakachu, on multiple cell lines.

**Results** Using different feature engineering and resampling strategies, we found spatial relationships (distance) between boundaries and annotations have the most predictive power. Transcription factor binding sites outperformed other genome annotation types. Random under-sampling significantly improved model performance. Boundaries predicted by preciseTAD were more enriched for CTCF, RAD21, SMC3, and ZNF143 sites and more conserved across cell lines, highlighting their higher biological significance. The model trained in one cell line perform well in predicting boundaries in another cell line.

**Conclusion** Our results outline strategies for predictive modeling of 3D genomic domains using genome annotation data. The precise identification of domain boundaries will improve our understanding of how genomic regulators are associated with the 3D structure of the genome.

**Availability.** Our method is implemented as an R package, *preciseTAD*, available on https://github.com/dozmorovlab/preciseTAD and Bioconductor (submitted).

# Introduction

The advent of chromosome conformation capture (3C) sequencing technologies, and its successor Hi-C, have revealed a hierarchy of 3-dimensional (3D) spatial constructs within the human genome such as chromatin loops [@rao20143d], Topologically Associated Domains (TADs) [@nora2012spatial;@dixon2012topological], and A/B compartments [@lieberman2009comprehensive]. At kilobase scale, gene promoters are often connected via distal regulatory elements, such as enhancers, through chromatin loops, thereby regulating gene expression [@Lupianez:2015aa; @Franke:2016aa]. Further, at the megabase scale, TADs represent regions on the linear genome, that are highly self-interacting. Emerging evidence has linked TADs to critical roles in cell dynamics and cell differentiation. Studies have shown that TADs themselves are highly conserved across species and cell lines [@dixon2012topological; @krefting2018evolutionary; @nora2012spatial; @dixon2015chromatin; @rao20143d; @Schmitt2016aa]. Likewise, disruption of boundaries demarcating TADs and loops promotes cancer [@hnisz2016activation; @taberlay2016three] and other disorders [@lupianez2016breaking;@Sun:2018aa;@Meaburn:2007aa]. Therefore, precisely identifying the location of TAD and chromatin loop boundaries remains a top priority in our goal to fully understand the functionality of the human genome.

Several methods have been proposed to identify genomic coordinates of TADs (reviewed in [@Zufferey:2018aa; @forcato2017comparison]), and chromatin loops [@rao20143d; @Ay:2014aa; @Heinz:2010aa; @salameh2019supervised].  However, a key limitation among most of these TAD- and loop callers is that they are heavily reliant on Hi-C data *resolution*. Resolution refers to the sizes of genomic regions (bins) used to segment the linear genome [@li2019comparison]. Lower resolutions contribute to larger bin sizes, which lead to uncertainty in the location of boundaries. Due to high sequencing costs, most available Hi-C datasets have relatively low resolution, ranging from 25 kb to 100 kb. Because increasing the resolution of Hi-C data requires a quadratic increase in the total sequencing depth, obtaining high resolution remains difficult [@schmitt2016genome]. While these low resolution Hi-C datasets can be used to probe higher order genomic patterns such as A/B compartments, more refined resolutions are necessary to accurately define TAD and loop boundaries [@fudenberg2016formation].

Another limitation among domain callers is that they disregard prior knowledge about functional genomic annotations associated with TAD and loop boundaries, referred hereafter as domain boundaries. The insulator binding protein, CTCF, and additional cofactors such as SMC3 and RAD21 have been identified as components of the loop extrusion model, whereby DNA is extruded through cohesin rings forming chromatin loops [@hansen2018recent;@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self; @Mirny:2019aa; @Davidson:2019aa]. Additionally, distinct patterns of histone modifications have also been shown to be present at boundaries, including H3K36me3 [@lieberman2009comprehensive;@dixon2012topological;@libbrecht2015joint]. These genomic annotations are obtained using chromatin immunoprecipitation followed by high-throughput DNA sequencing (ChIP-seq). The resolution of ChIP-seq experiments is typically on the order of tens to hundreds of bases [@mendoza2009genome], well below the resolution of Hi-C data (tens of kilobases; 750 bp is the highest resolution of Hi-C data to date [@bonev2017multiscale]). Therefore, leveraging precisely mapped genomic annotations in a supervised machine learning framework provides the possibility of more precisely predicting the locations of domain boundaries. 

We present *preciseTAD*, an R package implementing an optimally tuned machine learning framework for precise identification of domain boundaries using genome annotation data. Our method utilizes the random forest (RF) algorithm trained on chromatin segmentation (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS) data. We introduce a systematic pipeline for building optimal boundary prediction classifier. We found that spatial associations (distance) between boundaries and annotations perform best, transcription factor binding sites further improve prediction accuracy, and a simple random undersampling technique effectively addresses the negative effect of class imbalance. Translated from Hi-C data resolution level to base level (annotating each base and predicting its boundary probability), *preciseTAD* employs density-based clustering (DBSCAN) and scalable partitioning around medoids (CLARA) on prediction probabilities to select genome annotation-guided boundary coordinates. This approach circumvents resolution restrictions of Hi-C data, allowing for the precise detection of biologically meaningful boundaries. We demonstrate that predictions made by *preciseTAD* are more enriched for known molecular drivers of 3D chromatin including CTCF, RAD21, SMC3, and ZNF143. Further, we show that *preciseTAD* boundaries are more conserved across cell lines, and the model trained in one cell line can predict boundaries in another cell line. This improved precision in domain boundary location can provide insight into the association between genomic regulators and the 3D genome organization. The *preciseTAD* R package is freely available on GitHub (https://github.com/dozmorovlab/preciseTAD) and on Bioconductor (submitted).

# Methods

## Data sources

Publicly available TAD and loop boundaries called by Arrowhead [@durand2016juicebox] and Peakachu [@salameh2019supervised] tools, respectively, were used for training. Arrowhead TAD boundaries were called from primary Hi-C data for the GM12878 and K562 cell lines [@rao20143d]. The contact matrices were constructed from all read pairs that uniquely aligned to the genome (MAPQ>0). The default parameters for Arrowhead were used when extracting the autosomal genomic coordinates (GRCh37/hg19 human genome assembly) of TAD boundaries at 5 kb, 10 kb, 25 kb, 50 kb, and 100 kb resolutions (Additional File 1). Peakachu chromatin loop boundaries called at 10 kb for the GM12878 and K562 cell lines were downloaded from the Yue lab web site (Supplementary Table 1). Unique boundaries were considered as the midpoints within the coordinate of each chromatin loop anchor. Chromosome 9 was excluded from all downstream analyses due to sparsity of contact matrices at 5 kb and 10 kb resolutions for the K562 cell line. Cell-line-specific genomic annotations (BroadHMM chromatin segmentations (BroadHMM), histone modifications (HM), and transcription factor binding sites (TFBS)) were obtained from the UCSC Genome Browser Database [@Tyner:2017aa] (Supplementary Table 2). 

## Shifted-binning for binary classification

In Hi-C, each chromosome is binned into non-overlapping regions of length *r*, typically 5 kb and above. The *r* parameter defines the resolution of Hi-C data. Here, we designed a strategy called _shifted binning_ that partitions the genome into regions of the same length *r*, but with middle points corresponding to boundaries defined by the original binning. 

To create shifted binning, the first shifted bin was set to start at half of the resolution *r*, and continued in intervals of length *r* until the end of the chromosome (*mod r* + *r/2*). The shifted bins, referred hereafted as bins for simplicity, were then defined as boundary-containing regions (*Y = 1*) if they contained a TAD (or loop) boundary, and non-boundary regions (*Y = 0*) otherwise, thus establishing the binary response vector (**Y**) used for classification (Figure 1A).

## Feature engineering

Cell line-specific genomic annotations were used to build the predictor space for our boundary prediction framework. Bins were annotated by one of either the number of overlaps with an annotation (*Overlap Count (OC)*), the percent of overlap between the bin and the total width of genomic annotation regions overlapping it (*Overlap Percent (OP)*), the average signal strength of the corresponding annotation (*Peak Signal Strength*), or the distance in bases from the center of the bin to the center of the nearest genomic annotation region (*Distance*) (Figure 1B). A ($log_2 + 1$)-transformation of distance was used to account for the skewness present in the distributions of genomic distances. Models built using a *Peak Signal Strength* predictor space were only composed of histone modifications and transcription factor binding sites because BroadHMM chromatin segmentations lack signal values.

## Addressing class imbalance

To assess the impact of class imbalance (CI), defined as the proportion of boundary regions to non-boundary regions, we evaluated three re-sampling techniques including *Random Over-Sampling (ROS)*, *Random Under-Sampling (RUS)*, and *Synthetic Minority Over-Sampling Technique (SMOTE)*. For ROS, the minority class was sampled with replacement to obtain the same number of data points in the majority class. For RUS, the majority class was sampled without replacement to obtain the same number of data points in the minority class. For SMOTE, under-sampling was performed without replacement from the majority class, while over-sampling was performed by creating new synthetic observations using the $k=5$ minority class nearest neighbors [@chawla2002smote] (implemented in the `DMwR` v.0.4.1 R package). We restricted the SMOTE algorithm to 100% over-sampling and 200% under-sampling, which creates perfectly balanced classes. 

## Establishing optimal data level characteristics for TAD boundary region prediction

Random forest (RF) classification models (the `caret` v.6.0 R package [@kuhn2012caret]) were built to compare model performances between combinations of data resolutions, feature engineering procedures, and re-sampling techniques. A *holdout chromosome* technique was used for model evaluation on each autosomal chromosome. The $i^{th}$ holdout chromosome was identified and a data matrix, $A_{N \times (p+1)}$, was constructed by combining the binned genome from the remaining chromosomes ($1,2,\cdots, i-1,i+1,\cdots,21,22$), where $N=[n_{1} n_{2} \cdots n_{21} n_{22}]'$ and $n_{k}$ is the length of chromosome *k* after being binned into non-overlapping regions of resolution *r*, such that $k \ne i$. $p$, the number of annotations, together with the response vector **Y**, defined the column-wise dimension of the matrix $A$. Re-sampling was then performed on $A$ and an RF classifier was trained using 3-fold cross-validation to tune for the number of annotations to consider at each node (*mtry*). The number of trees (*ntree*) that were aggregated for each RF model was set to 500. The minimum number of observations per root node (*nodesize*) was set to .01% of the rows in the data. The binned data for the holdout chromosome $i$ was reserved for testing. Models were evaluated using Balanced Accuracy (BA), defined as the average of sensitivity and specificity:

$$BA = \dfrac{1}{2}\left( sensitivity + specificity \right) = \dfrac{1}{2}\left( \dfrac{TP}{TP+FN} + \dfrac{TN}{TN+FP} \right)$$

where TP refers to the number of bins correctly identified as containing a boundary (true positives), FP refers to the number of bins incorrectly identified as containing a boundary (false positives), TN refers to the number of bins correctly identified as not containing a boundary (true negatives), and FN refers to the number of bins incorrectly identified as not containing a boundary (false negatives). Each of these quantities are obtained from the confusion matrix created by validating the model on the test data. The process was repeated for each $i^{th}$ holdout chromosome and performances were aggregated using mean and standard deviation.

## Feature selection and predictive importance

Many genomic annotations, notably architectural proteins, tend to exhibit an extensive pattern of colocalization (correlation) [@van2014insulator]. To suitably reduce the predictor space and improve computational efficiency, while maintaining optimal performance, we implemented recursive feature elimination (RFE). In doing so, we estimated the near optimal number of necessary features to include in the best fitting RF model. The best performing subset was identified, ranging from 2 to the maximum number of features by powers of 2. We then aggregated the predictive importance of the union of the optimal set of features across chromosomes using the mean decrease in node impurity among permuted features in out-of-bag samples to determine the most common and top ranked annotations for predicting boundary regions. 

## Evaluating performance across cell lines

When evaluating performance across cell lines, a random forest model was built using ground truth boundaries and genomic annotations from one cell line such that the data matrix was constructed using genomic bins from $n-1$ chromosomes. The test data used to evaluate performance was constructed from the binned data of the $i^{th}$ holdout chromosome with genomic annotations from the other cell line. Thus, the models were trained on cell line specific data and tested on the other cell line. This process was repeated for each holdout chromosome. To evaluate performance, we constructed receiver operating characteristic (ROC) curves composed of the average sensitivies and specificities, and reported the area under the curve (AUC).

## Boundary prediction at bp resolution using *preciseTAD*

To investigate whether we could alleviate resolution limitations of conventional domain calling tools, we developed *preciseTAD*, an algorithm that leverages a random forest model in conjunction with density-based and partitioning techniques to predict boundaries at base resolution. To demonstrate, we implemented an optimally trained RF model using the framework described above. That is, a random forest model was built on the optimal combination of predictor type (predictor space), re-sampling technique (class balancing correction), and top ranked annotations (most informative genomic elements) to maximally accurately predict boundaries at base _Hi-C resolution level_.

To precisely identify boundary locations, we first construct a _base-level resolution_ predictor space for a particular chromosome *i*. We then apply our optimally trained RF model on the base-level predictor space to extract the probability vector, $p_{n_{i}}$, denoting each base's probability of being a boundary point. Here, $n_{i}$ represents the length of chromosome *i*. A threshold (*t*) is specified on $p_{n_{i}}$, such that base coordinates with $p_{n_{i}} \ge t$ are extracted. Next, density-based spatial clustering of applications with noise (DBSCAN; version 1.1-5) is applied to the matrix of pairwise genomic distances between bases with $p_{n_{i}} \ge t$. The resulting clusters of highly predictive bases identified by DBSCAN were termed *preciseTAD boundary regions* (PTBR). To precisely identify a single base among each PTBR, *preciseTAD* then implements an extension of partitioning around medoids (PAM), known as CLARA (**Clustering Large Applications**) for better scaling and more computational efficiency [@reynolds2006clustering; @wei2000empirical]. The corresponding cluster medoid was defined as a *preciseTAD boundary point* (PTBP), making it the most representative and biologically meaningful base coordinate within each clustered PTBR.

The DBSCAN algorithm takes two parameters, *MinPts* and *eps* ($\epsilon$). The *MinPts* parameter was set to the recommended value of 3, representing $1+dim(data)$ [@hahsler2019dbscan]. To decide on the optimal value of *t* and $\epsilon$ in *preciseTAD*, we considered the normalized enrichment of flanked predicted boundaries, defined as

$$N_{e} = \dfrac{1}{b} \Sigma_{i=1}^{n}e_{i}$$

where $e_{i}$ is the number ChIP-seq peak regions that overlap with the $i^{th}$ predicted boundary, $b$, flanked by resolution $r$. We evaluated $N_{e}$ for combinations of $t=\{0.95, 0.975, 0.99, 1.0 \}$ and $\epsilon = \{5000, 10000, 15000, 20000, 25000 \}$. The heuristic of $\epsilon$ is that density-reachable bases with genomic distances less than $\epsilon$ should occupy the same designated cluster.


## Comparing called vs. predicted boundaries

Next, we assessed the biological significance of our predicted boundaries by their association with several known molecular drivers of 3D chromatin architecture including CTCF, RAD21, SMC3, and ZNF143 ChIP-Seq peaks. To do so, we used the **computeMatrix** and **plotHeatmap** tools offered by deepTools (version 2.0) [@ramirez2016deeptools] via integration with Galaxy (version 3.3.2.0.0) [@afgan2018galaxy]. Additionally, we compared the median $log_2$ genomic distances between TAD boundaries and the same top predictive ChIP-seq annotations using Wilcoxon Rank-Sum tests. Furthermore, we compared the overlap between predicted and called boundaries in GM12878 and K562 cell lines. Boundaries were first flanked by resolution, *r*, and overlaps were visualized using Venn diagrams from the `Vennerable` R package (version 3.1.0). Overlaps were further quantified using the Jaccard index defined as

$$J_{(A,B)} = \dfrac{A \cap B}{A \cup B}$$

where A and B represent genomic regions created by flanking called and predicted boundaries. That is, between cell lines, the number of overlapping flanked boundaries divided by the total number of flanked boundaries. Wilcoxon Rank-Sum tests were used to compare chromosome-specific Jaccard indices across cell lines, between *preciseTAD* boundaries and both Arrowhead and Peakachu boundaries, respectively. All statistical analyses were performed in R version 4.0.1. The alpha level of significance was set to 0.05 for all statistical tests.

<!--using the `EnrichedHeatmap` R package (version 1.14.0) [@gu2018enrichedheatmap]. For each boundary (both called and predicted), a 5000 bp flank was created on either side. The flanking was broken up into 50 bp segments, for a total of 100 windows on both sides of a given boundary. A coverage value was generated for each window with respect to a given ChIP-seq annotation, as the mean signal averaged by the width of the window, denoted by,

$$v_{c} = \dfrac{\Sigma_{j}^{m} \Sigma_{i}^{n}x_{ij}w_{ij}}{L}$$

where $x_{ij}$ is the total number of bases of annotation $i$ in window $j$ and $w_{ij}$ is the number of bases that overlap between annotation $i$ and window $j$. The denominator, $L$, is the width of the windows (here, $L=50$). The rows of the heatmaps are ordered by enriched scores. The enriched scores are calculated as the sum of coverage values weighted by the distance to the flanked boundary, denoted by

$$s_{e} = \Sigma_{d=1}^{n_{1}}\left(\dfrac{x_{1d}* d}{n_{1}} \right) + \Sigma_{u=1}^{n_{2}}\left(\dfrac{x_{2u}* (n_{2} - u + 1)}{n_{2}} \right)$$
where $n_{1}$ and $n_{2}$ represent the number of downstream and upstream coverage values to sum over. Here, $n_{1}=n_{2}=100$.-->

# Results

## Developing a ML framework for optimal TAD boundary prediction

We developed a machine learning (ML) framework aimed at determining the optimal set of data level characteristics necessary for precise prediction of boundaries of Topologically Associated Domains (TADs) and chromatin loops, collectively referred to as domain boundaries. We chose the random forest (RF) algorithm as our binary classification tool. The reason for it is two-fold: (1) to devise a tunable prediction rule in a supervised learning framework that is both robust to overfitting and able to handle multiple correlated predictors, and (2) to allow for an interpretable ranking of predictors [@boulesteix2012overview]. We use published Arrowhead-called TAD boundaries [@durand2016juicer] and Peakachu-predicted loop boundaries [@salameh2019supervised] as ground truth. Data from GM12878 and K565 cell lines at 5-100kb resolution (Arrowhead) and 10kb resolution (Peakachu) were used (Supplementary Table 1). We next developed *preciseTAD*, an R package which leverages an optimally built RF to precisely predict genomic coordinates of TAD and loop boundaries at base resolution from cell type-specific genome annotation data.

In our ML framework boundary regions were defined as genomic bins containing a called boundary ($Y = 1$), while non-boundary regions were defined as bins that did not contain a called boundary ($Y = 0$) (Figure 1A, see Methods). The total number of called TADs, their unique boundaries, and the number of genomic bins expectedly decreased with the decreased resolution of Hi-C data (Table 1, Supplementary Table 3). The number of non-boundary regions highly outnumbered boundary regions. Such disproportional presence of examples in one class is known as a "class imbalance" problem that negatively affects predictive modeling [@prati2011survey; @wei2013role]. To address class imbalance, we evaluated the effect of three re-sampling techniques. *Random over-sampling* (ROS) was defined as sampling with replacement from the minority class (boundary regions). *Random under-sampling* (RUS) was defined as sampling with replacement from the majority cluss (non-boundary regions). Lastly, we tested *Synthetic minority over-sampling technique* (SMOTE), which is a combination of both random over- and under-sampling to create balanced classes [@chawla2002smote] (see Methods). 

A total of 77 cell line-specific functional genomic annotations were used to build the predictor space. These included BroadHMM chromatin segmentations, histone modifications (HM), and transcription factor binding sites (TFBS) (Supplementary Table 2). Four feature engineering procedures were developed to quantify the association between genomic annotations and bins (Figure 1B). These included signal strength association (Signal), direct (OC), proportional (OP), and spatial (Distance) relationships (see Methods). A ($log_2 + 1$)-transformation of distance was used to account for the skewness present in the distributions of genomic distances (Supplementary Figure 1). 

![**Resolution-specific data construction and feature engineering used for downstream random forest modeling.** (A) The linear genome was binned into non-overlapping resolution-specific intervals (see Methods). The response vector **Y** was defined as 1/0 if a genomic bin overlapped/did not overlap with a TAD boundary, respectively. (B) Four types of associations between bins and genomic annotations (green shapes) were considered to build the predictor space including Average Peak Signal (Signal), Overlap Counts (OC), Overlap Percent (OP), and $log_{2}$ distance (Distance).](Figures/figure1.png)


In total, we considered combinations of data from two cell lines *L = {GM12878, K562}*, five resolution *R = {5 kb, 10 kb, 25 kb, 50 kb, 100 kb}*, four types of predictor space *P = {Signal, OC, OP, Distance}*, and three re-sampling technique *S = {None, RUS, ROS, SMOTE}* (Figure 2). Once the model inputs were established, a random forest classifier was trained on $n-1$ autosomal chromosomes, while reserving the $i^{th}$ chromosome for testing. Three-fold cross-validation was used to tune the *mtry* hyperparameter, while *ntree* and *nodesize* were fixed to default values of 500 and 1 respectively. Model performance was evaluated by aggregating the mean balanced accuracy (BA) across each holdout chromosome (see Methods). These strategies allowed us to select the best performing model in an unbiased manner.

![**A diagram of the machine learning framework used to build TAD/loop boundary region prediction models.** Step 1 defines the dimensions of the data matrix $A_{N \times (p+1)}$ used for training and testing by establishing a resolution, $r$, at which to bin the linear genome into $N$ genomic bins for each chromosome, except a holdout chromosome *i*. The response vector **Y** is defined as a boundary region ($Y=1$) if it overlaps with a genomic bin (else $Y=0$) A feature engineering procedure is selected to define the feature space of $A$ composed of $p$ genomic annotations. Step 2 is to reserve the predictor-response matrix for the holdout chromosome *i* as the test data. Step 3 is to apply a re-sampling technique to the training data to address class imbalance. Step 4 is to train the random forest model and perform 3-fold cross-validation to tune for the mtry parameter. Finally, step 5 is to validate the model on the separate test data composed of the binned data from the holdout chromosome *i* and to evaluate model performance using balanced accuracy (BA).](Figures/figure2.png) 


|          Tool         |     Resolution    |     Total number of called TADs/chromatin loops    |     Total number of unique TAD/chromatin loop boundaries    |     Total number of genomic bins    |     Class imbalance    |
|:---------------------:|:-----------------:|:--------------------------------------------------:|:-----------------------------------------------------------:|:-----------------------------------:|:----------------------:|
|        Arrowhead      |       5   kb      |                         8,052                       |                             15,468                           |                535,363               |           0.03         |
|     Arrowhead         |       10   kb     |                         7,676                       |                             14,253                           |                267,682               |           0.05         |
|     Arrowhead         |       25   kb     |                         4,670                       |                             8,363                            |                107,073               |           0.08         |
|     Arrowhead         |       50   kb     |                         2,349                       |                             4,224                            |                 53,537               |           0.08         |
|     Arrowhead         |      100   kb     |                         1,031                       |                             1,883                            |                 26,768               |           0.07         |
|        Peakachu       |       10   kb     |                        16,185                       |                          21,421                              |                267,682               |        0.14            |
Table: TAD/loop boundary data and class imbalance summaries across resolutions for Arrowhead and Peakachu on the GM12878 cell line. Class imbalance is calculated as the ratio of total number of unique boundaries to the total number of genomic bins.


## Random under-sampling, distance-based predictors, and high-resolution Hi-C data provide optimal performance for boundary prediction

Expectedly, when using data with class imbalance present, that is, no re-sampling, the models exhibited low balanced accuracies, with very little variability among different resolutions (Figure 3). Similarly, poor performances were found when using ROS. However, RUS and SMOTE re-sampling led to drastic improvement in performance, especially at higher resolutions. We found that RUS marginally outperformed SMOTE as the most optimal class balancing technique for all resolutions and predictor types when predicting TAD boundary regions. 

Additionally, we found that using a distance-type predictor space yielded substantially higher balanced accuracies when compared to the peak signal stength, overlap count, and overlap percent predictor types. As with class balancing techniques, this improvement was less evident at lower resolutions with results consistent for K562 (Supplementary Figure 2). Random forest models built on Peakachu-defined loop boundaries with SMOTE resampling only marginally outperformed RUS for both GM12878 and K562 cell lines (Supplementary Figure 3). Furthermore, 5 kb resolution genomic bins led to optimal prediction for both TAD and loop boundaries, on both cell lines. Our results indicate that random under-sampling, distance-type predictors, and high-resolution Hi-C data provide the optimal set of data level characteristics for both TAD and chromatin loop boundary prediction.


![**Distance-type predictors combined with random under-sampling at 5 kb resolution genomic bins yielded optimal TAD boundary region predictive performance for GM12878 Arrowhead TAD boundaries.** Averaged balanced accuracies are compared across resolution, within each predictor-type: (A) Signal, (B) OC, (C) OP, and (D) Distance, and across re-sampling techniques: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Error bars indicate 1 standard deviation away from the mean performance across each holdout chromosome used for testing.](Figures/figure3.png)


## Transcription factor binding sites outperform histone- and chromatin segmentation-specific models

We hypothesized that the type of genomic annotations may also affect predictive performance. Using the established optimal settings (RUS, Distance, 5 kb bins), we built separate random forest models using BroadHMM chromatin segmentations, histone modifications (HM), and transcription factor binding sites (TFBS). We found that models built on TFBSs outperformed other annotation-specific models, with results consistent for loop boundaries, on both cell lines (Figure 4A; Supplementary Figures 4A). These results suggest that TFBS are the primary drivers of TAD and loop boundary formation in both GM12878 and K562. 

## Predictive importances confirmed biological functionality of CTCF, RAD21, SMC3, and ZNF143 at TAD boundaries

We sought to further optimize our boundary region prediction models by performing recursive feature elimination to avoid overfitting and select only the most influential TFBS across all autosomal chromosomes. We were able to obtain near-optimal performance using approximately eight TFBS (Figure 4B; Supplementary Figures 4B). However, given that we trained our model on chromosome-specific data, the most significant annotations varied for each set of chromosomes. To determine transcription factors most important for boundary prediction, we clustered the predictive importances (mean decrease in accuracy) of the top eight significant TFs across chromosomes. We found that, similar to other studies [@sanborn2015chromatin; @weintraub2017yy1; @gonzalez2017znf143], CTCF, RAD21, SMC3, and ZNF143 transcription factors were consistently predictive of TAD and loop boundaries (Figure 4C; Supplementary Figures 4C). We further optimized our model by only considering these top informative TFBS when building the random forest, thereby decreasing computational burden while maintaining high predictive performance. In summary, our model was able to yield the top molecular drivers of the 3D architecture of the human genome, components of the popular loop extrusion model [@hansen2018recent;@sanborn2015chromatin; @fudenberg2016formation; @alipour2012self; @Mirny:2019aa; @Davidson:2019aa]. 

## Training in one cell line accurately predicts boundary regions in other cell lines 

Having demonstrated the optimal performance in our framework's ability to predict boundary regions, we next attempted to predict boundary regions in one cell line when trained on another (See Methods). We found that training and testing using Arrowhead ground truth TAD boundaries and genomic elements from GM12878 resulted in an average AUC=0.845 (Figure 4D). However when training on GM12878 and testing on K562, our model's average AUC only marginally decreased to 0.803. Likewise, training and testing on Peakachu resulted in an average AUC=0.885 and decreased only slightly to 0.868 when trained on GM12878 and tested on K562. Results were consistent when training on K562 and testing on GM12878 (Supplementary Figure 4D). These results suggest that a classifier trained on the data from one cell line performs well when using data from another cell line.

![**Binding of SMC3, RAD21, CTCF, and ZNF143 transcription factors accurately predicts TAD and loop boundaries in GM12878 cell line.** (A) Barplots comparing performances of TAD (Arrowhead) and loop (Peakachu) boundary prediction models using histone modifications (HM), cis-regulated elements (BroadHMM), transcription factor binding sites (TFBS), in addition to a model containing all three classes (ALL). (B) Recursive feature elimination (RFE) analysis used to select the optimal number of predictors. Error bars represent 1 standard deviation from the cross-validated accuracy across each holdout chromosome. (C) Clustered heatmap of the predictive importance for the union of the top 8 most predictive chromosome-specific TFBSs. The columns represent the holdout chromosome excluded from the training data. (D) Receiver operating characteristic (ROC) curves and subsequent area under the curves (AUCs) when training and testing on the same cell line versus across cell lines. The curves represent the average sensitivities and specificities across each holdout chromosome.](Figures/figure4.png)


## *preciseTAD* identifies precise and biologically relevant locations of TAD boundaries

Using our optimally built random forest model trained on Arrowhead/Peakachu boundaries, we attempted to predict TAD/loop boundaries at base resolution. Intuitively, the predictor-response space was built on a base-resolution level, and the model trained on a bin-level space was used to predict the probability of each base being a boundary. Our method, referred to as *preciseTAD*, uses density-based hierarchical clustering (DBSCAN) and a scalable partitioning around mediods clustering (CLARA) to predict the boundary location at base resolution (see Methods).

When built using Arrowhead and Peakachu ground truth boundaries at 5 kb and 10 kb respectively, *preciseTAD* was able to predict a total of 7,846 TAD and 15,838 chromatin loop boundaries on GM12878, as well as 8,883 TAD and 11,016 chromatin loop boundaries on K562. Noticeably, we reported less predicted TAD boundaries at 5 kb than Arrowhead on both cell lines (Table 1, Supplementary Table 3). This can be attributed to Arrowhead's inflation of called TADs, especially at higher resolutions, that often do not correspond to intra-chromosomal contacts featured on a contact matrix. Additionally, *preciseTAD* boundaries were visually more likely to demarcate high concentrations of intra-chromosomal contacts as can be seen on a Hi-C contact matrix (Figure 5A). *preciseTAD* also predicted fewer chromatin loop boundaries than Peakachu (Table 1, Supplementary Table 3). This can be contributed to the fact that *preciseTAD* leverages four known drivers of 3D chromatin including CTCF, RAD21, SMC3, and ZNF143, contrasting Peakachu which considers only CTCF sites as true positives. These results suggest that genome annotation-driven predicted boundaries may be more biologically relevant, in contrast to boundaries detected from Hi-C data alone.

Next, we evaluated the biological significance of *preciseTAD* boundaries. Peak regions associated with known molecular drivers of 3D chromatin including CTCF, RAD21, SMC3, and ZNF143 were found to co-localize more frequently around *preciseTAD* predicted boundaries (Figure 5B). Furthermore, signal heatmaps indicated that our predicted boundaries were more enriched for the same genomic annotations than either Arrowhead or Peakachu boundaries alone (Figure 5 C-D; Supplementary Figures 5-6). When compared to called TAD or loop boundaries, *preciseTAD* boundaries were statistically significantly closer to the same cell line-specific top ranked TFBS (p<0.001 and p<0.001 versus Arrowhead and Peakachu respectively) (Figure 5E, Supplementary Figure 7). These results indicate that *preciseTAD*-predicted boundaries better reflected the known biology of boundary formation.

## *preciseTAD* boundaries are more conserved across cell lines

TAD boundaries are known to be relatively conserved across cell lines [@dixon2012topological; @krefting2018evolutionary; @nora2012spatial; @sexton2012three]. To assess conservation, we evaluated the overlap between cell line-specific boundaries detected by *preciseTAD*, Arrowhead, and Peakachu after flanking them by the resolution of the Hi-C data. Boundaries involving Arrowhead were flanked by 5 kb, while boundaries involving Peakachu were flanked by 10 kb. We found that 35% of *preciseTAD* predicted TAD boundaries and 55% of *preciseTAD* predicted loop boundaries were were conserved between cell lines, compared to only 26% and 55% for Arrowhead and Peakachu alone (Figure 5F). The better conservation of *preciseTAD*-predicted boundaries further support the notion of their higher biological relevance.


![**preciseTAD boundaries are enriched for known molecular drivers of 3D chromatin.** (A) The location of preciseTAD (green) vs. ARROWHEAD (blue) boundaries on GM12878 (chr 14: 50085000-50800000). The black line represents the predicted probability of each base being a TAD boundary. (B) Genome browse visualization of bigwig files for known molecular drivers of 3D chromatin including CTCF, RAD21, SMC3, and ZNF143. (C) Enriched heatmaps comparing the enrichment of CTCF signal around ARROWHEAD TAD boundaries and preciseTAD boundaries, and (D) around Peakachu loop boundaries and preciseTAD for the entire genome. (E) Log 2 genomic distances from boundaries to nearest CTCF sites. (F) Venn diagrams illustrating the levels of conservation between TAD/loop boundaries for GM12878 (red) and K562 (blue) identified by preciseTAD (using ARROWHEAD; top left), ARROWHEAD (top right), preciseTAD (using PEAKACHU; bottom left), and PEAKACHU (bottom right). The median Jaccard Index (IQR) across each chromosome is provided below each venn diagram](Figures/figure5.png) 

## *preciseTAD* predicted TAD boundaries are nested within predicted loop boundaries

The majority of boundaries predicted by the Arrowhead-trained *preciseTAD* model represented a subset of the larger group of boundaries predicted by the Peakachu-trained model. We found that 95.6% and 95.7% of our predicted TAD boundaries were overlapped by predicted loop boundaries for GM12878 and K562, respectively (Supplementary Figure 8). This is expected as loop boundaries (Peakachu) are more abundant, while TAD boundaries (Arrowhead) represent the narrower level of the 3D chromatin organization. Together with the results on Figure 4, the high overlap between boundaries predicted by Arrowhead- and Peakachu-trained models suggests that TAD and loop boundaries may be driven by similar protein complexes.


# Discussion

We present *preciseTAD*, a machine learning approach for precise prediction of TAD or loop boundaries from functional genomic annotations. *preciseTAD* leverages a random forest (RF) classification model built on ground truth TAD/loop boundaries obtained from TAD/loop calling tools and genomic annotations as the predictor space. We benchmarked *preciseTAD* against two boundary calling tools, Arrowhead [@durand2016juicebox], an established TAD-caller, and Peakachu [@salameh2019supervised] a recently published algorithm for predicting chromatin loops. We applied *preciseTAD* on cell line-specific boundaries and annotation data from both GM12878 and K562. We first optimized our RF model by systematically comparing different combinations of genome binning (resolution), feature engineering procedures, and re-sampling techniques. We show that genomic distance between high-resolution boundary regions and transcription factor binding sites, coupled with random under-sampling results in best model performance. We demonstrate that binding of four transcription factors (SMC3, RAD21, CTCF, ZNF143) are sufficient for accurate TAD and loop boundary predictions. Applied to each annotated base, the model predicts probability for each base to be a boundary, and, coupled with spatial clustering (DBSCAN) and scalable partition around medoids clustering (CLARA), allows for the detection of boundaries at the base-level resolution. *preciseTAD* is implemented as an R package, and allows for precise prediction of boundaries for any cell type for which the corresponding genomic annotations are available.

Our machine learning framework yielded several interesting observations. First, we demonstrated that RF models built using *distance*-type predictors outperformed models built on previously published feature engineering techniques including signal strength, overlap counts and overlap percents [@mourad2016computational; @Moore:2015aa; @Al-Bkhetan:2018aa; @kai2018predicting]. We further demonstrated that class imbalance hinders boundary prediction, but can be effectively addressed by a simple random under-sampling technique, an aspect of boundary prediction that was unaddressed in previous studies [@mourad2016computational; @Moore:2015aa; @Al-Bkhetan:2018aa]. We showed that information about only four transcription factors (SMC3, RAD21, CTCF, ZNF143) is necessary and sufficient for accurate boundary prediction, outperforming histone modification- and BroadHMM-built models [@Moore:2015aa; @Al-Bkhetan:2018aa]. These are known components of the loop extrusion model, an established theory for how loops are made by a ring-shaped adenosine triphosphatase complex called cohesin [@hansen2018recent; @sanborn2015chromatin; @fudenberg2016formation; @alipour2012self; @Mirny:2019aa; @Davidson:2019aa]. Interestingly, the same transcription factors accurately predicted both TAD and loop boundaries, suggesting similarity of the mechanisms of TAD and loop formation. This suggested that the random forest model, when tuned and feature engineered correctly, is highly effective in predicting biologically relevant domain boundaries. 

Besides balanced accuracy (BA), we investigated five other performance metrics, including accuracy, area under the receiver operating characteristic curve (AUROC), precision, F1-score, and area under the precision-recall curve (AUPRC). Our aim was to have a balanced metric sensitive to class imbalance such that it would not favor one component of the confusion matrix. The accuracy metric can be artificially inflated by true negatives (TN), the set of genomic bins correctly predicted as not containing a ground truth boundary. AUROC captures how a model generally performs across different thresholds. However, it doesn't place more emphasis on one class over the other, so it does not reflect the minority class well. Precision indicates the rate at which positive predictions are correct and can be artificially deflated by low proportions of true positives (TP), the set of genomic bins correctly predicted as containing a ground truth boundary. While F1-score is a composite metric, it can be very sensitive towards disparate values for precision and recall. Lastly, AUPRC is insensitive toward class imbalance, preventing us to investigate its effect, and also omits from its calculation TN values. All of these are important considerations to make when choosing a performance metric. For these reasons we opted to report the balanced accuracy (BA). The BA benefits from incorporating all components of the confusion matrix, while also being sensitive to class imbalance, a necessary characteristic when comparing performances to models built using no data re-sampling. [???refer to supp tables 4 and 5 here when ready]

Guided by both low-resolution Hi-C data and high-resolution genome annotation data, *preciseTAD* predicts base-level resolution boundaries, alleviating resolution limitations of Hi-C data. We show that, unlike Arrowhead, *preciseTAD* does not inflate the number of predicted boundaries, instead, providing only the most biologically meaningful boundaries that demarcate regions of high inter-chromosomal interactions. *preciseTAD* TAD and loop boundaries were enriched for known architectural transcription factors including CTCF, RAD21, SMC3, and ZNF143, supporting recent observations that TADs and loops are dynamic structures generated by similar mechanisms [@hansen2018recent]. Likewise, *preciseTAD* boundaries were found to be more conserved between GM12878 and K562 cell lines, a known feature among the 3D architecture of the human genome, further highlighting their biological significance.

*preciseTAD* offers flexibility in controlling both the number of TAD(loop) boundaries as well as the distance between them. The two primary parameters are the probability threshold (*t*) and $\epsilon$ (refered to as eps; parameter of DBSCAN). The combination of these two quantities change the resulting number of predicted boundaries from *preciseTAD*. A lower threshold and $\epsilon$ will result in more clusters of bases, and therefore, more boundaries. As a heuristic, we evaluated the pairwise combination of 4 different thresholds ($t=(0.95,0.975,0.99,1.0)$) and 4 different $eps$ values ($\epsilon=(r, 2r, 3r, 4r, 5r)$)), where $r=5000$, the optimal resolution. We found that the normalized overlaps--calculated as the total number of ChIP-seq peaks that overlapped within a given flanked boundary, divided by the number of boundaries --between top TFBS sites and flanked preciseTAD boundaries were maximized for combinations of t=1.0 and $eps=r$ (Supplementary Figure 9). 

A limitation of *preciseTAD* is that it is dependent on "ground truth" boundaries provided by a user-defined domain caller. Given the wide variety of domain callers and their variable performance [@Zufferey:2018aa; @forcato2017comparison], it is hard to define "ground truth" TAD boundaries. <!--Ideally we would benchmark *preciseTAD* against simulated boundaries. While methods for simulating Hi-C data sets with boundary annotations exist [@forcato2017comparison; @zheng2020freehi], methods for simulating the associated genomic annotations (the main component guiding *preciseTAD* predictions) are lacking.-->Ultra-deep Hi-C sequencing [@bonev2017multiscale] and newer technologies for precise mapping of chromatin interactions (e.g., Micro-C [@Krietenstein:2020aa]), coupled with more precise technologies for genomic annotation profiling (e.g., CUT&RUN for precise mapping of transcription factor binding sites) will help to refine the location and the genomic signatures of "ground truth" TAD boundaries. In the current work, given *preciseTAD* can be trained on the whole-genome data, we feel that the total number of TAD/loop boundaries is sufficient to guide learning of the association between genomic annotations and boundaries for precise boundary predictions.

Another limitation of *preciseTAD* is that it does not distinguish boundary types. The hierarchical nature of TAD boundaries [@weinreb2016identification; @fraser2015hierarchical; @Cresswell:2020ab] is not considered by *preciseTAD* due to the lack of gold standard of TAD hierarchy. Recent research distinguishes CTCF-associated boundaries, CTCF-negative YY1-enriched boundaries, CTCF- and YY1- depleted promoter boundaries, and a fourth class of weak boundaries largely depleted of all three features [@Krietenstein:2020aa]. Furthermore, actively transcribed regions can serve as TAD boundaries themselves, independently of CTCF binding [@Harrold:2020aa]. This may lead to some TAD boundaries being undetected by *preciseTAD* despite being detected by domain callers. Our future work will include defining separate models trained on different TAD boundary types, and predicting precise locations for each of them.

In summary, we demonstrate that TAD/loop boundary prediction is a multi-faceted problem requiring consideration of multiple statistical and biological properties of genomic data. Simply considering the properties of Hi-C contact matrices ignores the fundamental roles of known molecular drivers of the 3D chromatin structures. Instead, we propose *preciseTAD*, a supervised machine learning framework that leverages both Hi-C contact matrix information and genomic annotations. Our method introduces three concepts - *shifted binning*, *distance*-type predictors, and *random undersampling* - which we use to build random forest classification models for predicting boundary regions. Our method is able to bridge the resolution gap between 1D ChIP-seq annotations and 3D Hi-C sequencing data for more precise and biologically meaningful boundary identification. We introduce *preciseTAD*, an open source R package for leveraging random forests to predict TAD/loop boundaries at base resolution. We hope that *preciseTAD* will serve as an efficient and easy-to-use tool to further explore the 3D organization of the genome.


<!--
The two proteins differ in that YY1 preferentially
occupies interacting enhancers and promoters, whereas CTCF
preferentially occupies sites distal from these regulatory elements that tend to form larger loops and participate in insulation.
Deletion of YY1 binding sites or depletion of YY1 can disrupt
enhancer-promoter contacts and normal gene expression.
Thus, YY1-mediated structuring of enhancer-promoter loops is
analogous to CTCF-mediated structuring of TADs, CTCF contact domains, and insulated neighborhoods. This model of
YY1-mediated structuring of enhancer-promoter loops accounts
for diverse functions reported previously for YY1, including contributions to both gene activation and repression and to gene
dysregulation in cancer. ["YY1 Is a Structural Regulator
of Enhancer-Promoter Loops"]
-->





# Supplementary Legends

**Additional File 1**. An example script for applying the Arrowhead tool to in situ Hi-C data (HIC001-HIC018) for resolutions 5kb, 10kb, 25kb, 50kb, 100kb to obtain chromosome-specific TAD boundaries on the GM12878 cell line. <!--Prior to implementing SpectralTAD, the dump command offered by juicer [@durand2016juicer] was used to extract the contacts in sparse data matrix format.-->

**Supplementary Table 1.** Data source for Hi-C matrices used to call topologically associated domains with Arrowhead, as well as loop boundaries obtained by Peakachu.

**Supplementary Table 2.** A complete list of the genomic annotations used to build the predictor space for all downstream models. The GRCh37/hg19 human genome assembly was used. “Genomic Class” - broad category of genomic features, “Element” - names of genomic features, “Cell Type-Specific Source” - download URL specific to the cell line (not all annotations were provided by the same institutions).

**Supplementary Table 3.** TAD/loop boundary data and class imbalance summaries across resolutions for Arrowhead and Peakachu on the K562 cell line. Class imbalance is calculated as the ratio of total number of unique boundaries to the total number of genomic bins.

**Supplementary Table 4.** Additional performance metrics when implementing a random forest using
Arrowhead ground truth TAD boundaries. Performances are averaged across each holdout chromosome that was reserved for testing.

**Supplementary Table 5.** Additional performance metrics when implementing a random forest using
Peakachu ground truth chromatin loop boundaries. Performances are averaged across each holdout chromosome that was reserved for testing.

**Supplementary Figure 1.** Density curves illustrate the normalization of distances before (red) and after (blue) performing a $log_{2}$ transformation across 10 kb, 25 kb, 50 kb, and 100 kb data resolutions for both the (A) GM12878 and (B) K562 cell lines, respectively. Distances are measured as the number of bases from the center of a genomic bin to the center of the nearest genomic annotation. All genomic annotations were used. Each density curve represents an individual genomic annotation.

**Supplementary Figure 2. Distance-type predictors combined with random under-sampling at 5 kb resolution genomic bins yielded optimal TAD boundary region predictive performance for K562 Arrowhead TAD boundaries.** Averaged balanced accuracies are compared across resolution, within each predictor-type: (A) Signal, (B) OC, (C) OP, and (D) Distance, and across re-sampling techniques: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Error bars indicate 1 standard deviation away from the mean performance across each holdout chromosome used for testing.

**Supplementary Figure 3. Distance-type predictors combined with random under-sampling or SMOTE re-sampling yielded optimal predictive performance Peakachu loop boundaries at 10 kb resolution.** Averaged balanced accuracies are compared for A) GM12878 and B) K562  within each predictor-type: Signal, OC, OP, and Distance, and across resampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Error bars indicate 1 standard deviation away from the mean performance across each holdout chromosome used for testing.

**Supplementary Figure 4. Binding of SMC3, RAD21, CTCF, and ZNF143 transcription factors accurately predicts TAD and loop boundaries in K562 cell line.** (A) Barplots comparing performances of TAD (Arrowhead) and loop (Peakachu) boundary prediction models using histone modifications (HM), cis-regulated elements (BroadHMM), transcription factor binding sites (TFBS), in addition to a model containing all three classes (ALL). (B) Recursive feature elimination (RFE) analysis used to select the optimal number of predictors. Error bars represent 1 standard deviation from the cross-validated accuracy across each holdout chromosome. (C) Clustered heatmap of the predictive importance for the union of the top 8 most predictive chromosome-specific TFBSs. The columns represent the holdout chromosome excluded from the training data. (D) Receiver operating characteristic (ROC) curves and subsequent area under the curves (AUCs) when training and testing on the same cell line versus across cell lines. The curves represent the average sensitivities and specificities across each holdout chromosome.

**Supplementary Figure 5. preciseTAD boundaries are more enriched for known molecular drivers of 3D chromatin.** Enriched heatmaps comparing the enrichment of preciseTAD (green) boundaries with Arrowhead called TAD boundaries (blue) around key (A) GM12878-specific and (B) K562-specific transcription factor binding sites known to contribute to the 3D chromatin architecture.

**Supplementary Figure 6. preciseTAD boundaries are more enriched for known molecular drivers of 3D chromatin.** Enriched heatmaps comparing the enrichment of preciseTAD (green) boundaries with Peakachu predicted chromatin loop boundaries (blue) around key (A) GM12878-specific and (B) K562-specific transcription factor binding sites known to contribute to the 3D chromatin architecture.

**Supplementary Figure 7. preciseTAD boundaries are spatially closer to known molecular drivers of 3D chromatin.** Boxplots comparing the $log_{2}$ genomic distances from predicted and called boundaries to the nearest (A) GM12878-specific and (B) K562-specific CTCF, RAD21, SMC3, and ZNF143 transcription factor binding sites. 

**Supplementary Figure 8. preciseTAD predicted TAD boundaries using Arrowhead are nested within preciseTAD predicted chromatin loop boundaries using Peakachu.** Venn diagrams illustrating the agreement between boundaries predicted using Arrowhead- and Peakachu-trained models. TAD/loop boundaries were flanked by resolution (5 kb for models using Arrowhead and 10 kb for Peakachu).

**Supplementary Figure 9.** Linecharts illustrating the normalized overlap between known molecular drivers of 3D chromatin and flanked predicted TAD boundaries identified by preciseTAD (using Arrowhead ground truth boundaries) for different combinations of thresholds (t) and epsilon parameter values (eps) on GM12878 and for chromosome 22. Normalized overlap was calculated as the total number of ChIP-seq peaks that overlapped within a given flanked boundary, divided by the number of boundaries that were predicted (on chr22). Boundaries were flanked by 5 kb resolution. 


# Abbreviations


# Acknowledgements

_Conflict of Interest._ None.

# Funding

# References

