---
title: "Legends for Figures"
author: "Spiro Stilianoudakis"
date: "November 19, 2018"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Figure 1. Illustrating the resolution-specific data construction and feature engineering used for downstream random forest modeling.** (A) The linear genome was subsequently binned into resolution-specific intervals. The response vector **Y** was defined as a genomic bin that either overlapped with a TAD boundary ($Y = 1$) or did not ($Y = 0$). (B) Four types of associations between bins and genomic annotations (green shapes) were then used to build the feature space including _Peak Signal_ (Signal), _Overlap Counts_ (OC), _Overlap Percent_ (OP), and ($log_{2}$) _Distance_.

**Figure 2. A diagram of the ensemble framework detailing the steps for combining different inputs used in the random forest model building pipeline.** Step 1 is to establish the rowise (genomic bins) and colwise (genomic annotations) dimensions of the data matrix, **A**, by combining a resolution (5 kb, 10 kb, 25 kb, 50 kb, or 100 kb) and feature engineering procedure (overlap count (OC), overlap percent (OP), $log_{2}$ distance (Distance)). Step 2 is to split **A** into a training and testing set according to a 7:3 ratio. Step 3 is to apply a re-sampling technique to the training set (no re-sampling, random over-sampling, random under-sampling, or SMOTE re-sampling). Step 4 is to implement the random forest algorithm and perform 5-fold cross-validation to tune for the *mtry* parameter. Finally, step 5 is to validate the model on the separate test data for unbiased model evaluation.

**Figure 3. Models built using distance-type predictors combined with random under-sampling yielded optimal predictive performance for GM12878.** Line plots comparing the performances of random forest classifiers built using different combinations of data level characteristics. Performances are compared across resolution, within each predictor-type: (A) Signal (B) OC, (C) OP, and (D) Distance, and between each re-sampling technique: no re-sampling (None; red), random over-sampling (ROS; green), random under-sampling (RUS; blue), and synthetic minority over-sampling (SMOTE; purple). Performances were evaluated using MCCs and aggregated by averaging across each chromosome, with error bars representing 1 standard deviation away from the mean.

**Figure 4. Assessing top ranked transcription factors in TAD boundary prediction for GM12878.** (A) Barplots comparing performances of models using specific classes of functional genomic elements as the feature space including histone modifications (HM; red), cis-regulated elements (CRE; green), transcription factor binding sites (TFBS; blue), in addition to a model with a feature space containing all three classes (Combined; purple). (B) Recursive feature elimination analysis used to reduce the number of TFBS to include in the optimally build random forest classifier. Error bars represent standard deviation from the mean MCC across each chromosome. (C) Clustered heatmap of predictive importances for the union of the top 8 most predictive TFBS across chromosomes.

**Figure 5. Predicted TAD boundaries are enriched in known TAD boundary-forming TFBSs.** (A) Visualizing locations of predicted and called TAD boundaries (chr 22: 17389000-18011000; GM12878). The upper triangular contact map shows the normalized number of contacts within the specified window using the JUICER toolbox.The blue vertical lines represent called boundaries identified by ARROWHEAD, with the corresponding called TADs outlined in blue on the upper diagonal contact matrix above. The black line represents the predicted probability of each base pair being a TAD boundary when validated by *TADRF*. Collections of red vertical lines represent base pairs whose predictive probability exceed the ROC threshold, labeled as *predicted TAD boundary areas* (PTBA). The midpoints of spatially similar PTBA were clustered together into *predicted TAD boundary regions* (PTBR), denoted by the yellow regions along the x-axis. The green vertical lines represent the *predicted TAD boundary points* (PTBA) identified by *preciseTAD*, with the corresponding predicted TADs outlined above in green. <!--(B) Enrichment heatmaps illustrate the average signal strength of ChIP-seq peak regions defined by SMC3, ZNF143, CTCF, and RAD21 around the centers of 10 kb flanked predicted and called TAD boundaries.-->
